{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "varied-microphone",
   "metadata": {
    "tags": [
     "snakemake-job-properties"
    ]
   },
   "outputs": [],
   "source": [
    "\n",
    "######## snakemake preamble start (automatically inserted, do not edit) ########\n",
    "import sys; sys.path.extend(['/home/ktmeaton/miniconda3/envs/plague-phylogeography/lib/python3.7/site-packages', '/mnt/c/Users/ktmea/Projects/plague-phylogeography/workflow/notebooks']); import pickle; snakemake = pickle.loads(b'\\x80\\x03csnakemake.script\\nSnakemake\\nq\\x00)\\x81q\\x01}q\\x02(X\\x05\\x00\\x00\\x00inputq\\x03csnakemake.io\\nInputFiles\\nq\\x04)\\x81q\\x05(Xl\\x00\\x00\\x00/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/iqtree/all/chromosome/filter50/iqtree_post.cf.treeq\\x06XS\\x00\\x00\\x00/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/metadata/all/metadata.tsvq\\x07e}q\\x08(X\\x06\\x00\\x00\\x00_namesq\\t}q\\n(X\\x04\\x00\\x00\\x00treeq\\x0bK\\x00N\\x86q\\x0cX\\x03\\x00\\x00\\x00tsvq\\rK\\x01N\\x86q\\x0euX\\x12\\x00\\x00\\x00_allowed_overridesq\\x0f]q\\x10(X\\x05\\x00\\x00\\x00indexq\\x11X\\x04\\x00\\x00\\x00sortq\\x12eh\\x11cfunctools\\npartial\\nq\\x13cbuiltins\\ngetattr\\nq\\x14csnakemake.io\\nNamedlist\\nq\\x15X\\x0f\\x00\\x00\\x00_used_attributeq\\x16\\x86q\\x17Rq\\x18\\x85q\\x19Rq\\x1a(h\\x18)}q\\x1bX\\x05\\x00\\x00\\x00_nameq\\x1ch\\x11sNtq\\x1dbh\\x12h\\x13h\\x18\\x85q\\x1eRq\\x1f(h\\x18)}q h\\x1ch\\x12sNtq!bh\\x0bh\\x06h\\rh\\x07ubX\\x06\\x00\\x00\\x00outputq\"csnakemake.io\\nOutputFiles\\nq#)\\x81q$(Xk\\x00\\x00\\x00/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/parse_tree/all/chromosome/filter50/parse_tree.tsvq%Xk\\x00\\x00\\x00/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/parse_tree/all/chromosome/filter50/parse_tree.nwkq&e}q\\'(h\\t}q((h\\rK\\x00N\\x86q)h\\x0bK\\x01N\\x86q*uh\\x0f]q+(h\\x11h\\x12eh\\x11h\\x13h\\x18\\x85q,Rq-(h\\x18)}q.h\\x1ch\\x11sNtq/bh\\x12h\\x13h\\x18\\x85q0Rq1(h\\x18)}q2h\\x1ch\\x12sNtq3bh\\rh%h\\x0bh&ubX\\x06\\x00\\x00\\x00paramsq4csnakemake.io\\nParams\\nq5)\\x81q6}q7(h\\t}q8h\\x0f]q9(h\\x11h\\x12eh\\x11h\\x13h\\x18\\x85q:Rq;(h\\x18)}q<h\\x1ch\\x11sNtq=bh\\x12h\\x13h\\x18\\x85q>Rq?(h\\x18)}q@h\\x1ch\\x12sNtqAbubX\\t\\x00\\x00\\x00wildcardsqBcsnakemake.io\\nWildcards\\nqC)\\x81qD(X\\x03\\x00\\x00\\x00allqEX\\n\\x00\\x00\\x00chromosomeqFX\\x02\\x00\\x00\\x0050qGe}qH(h\\t}qI(X\\x0c\\x00\\x00\\x00reads_originqJK\\x00N\\x86qKX\\n\\x00\\x00\\x00locus_nameqLK\\x01N\\x86qMX\\x0c\\x00\\x00\\x00missing_dataqNK\\x02N\\x86qOuh\\x0f]qP(h\\x11h\\x12eh\\x11h\\x13h\\x18\\x85qQRqR(h\\x18)}qSh\\x1ch\\x11sNtqTbh\\x12h\\x13h\\x18\\x85qURqV(h\\x18)}qWh\\x1ch\\x12sNtqXbX\\x0c\\x00\\x00\\x00reads_originqYhEX\\n\\x00\\x00\\x00locus_nameqZhFX\\x0c\\x00\\x00\\x00missing_dataq[hGubX\\x07\\x00\\x00\\x00threadsq\\\\K\\x01X\\t\\x00\\x00\\x00resourcesq]csnakemake.io\\nResources\\nq^)\\x81q_(K\\x01K\\x01K\\x01M\\xb8\\x0bK<Kde}q`(h\\t}qa(X\\x06\\x00\\x00\\x00_coresqbK\\x00N\\x86qcX\\x06\\x00\\x00\\x00_nodesqdK\\x01N\\x86qeX\\x04\\x00\\x00\\x00cpusqfK\\x02N\\x86qgX\\x06\\x00\\x00\\x00mem_mbqhK\\x03N\\x86qiX\\x08\\x00\\x00\\x00time_minqjK\\x04N\\x86qkX\\x04\\x00\\x00\\x00loadqlK\\x05N\\x86qmuh\\x0f]qn(h\\x11h\\x12eh\\x11h\\x13h\\x18\\x85qoRqp(h\\x18)}qqh\\x1ch\\x11sNtqrbh\\x12h\\x13h\\x18\\x85qsRqt(h\\x18)}quh\\x1ch\\x12sNtqvbhbK\\x01hdK\\x01hfK\\x01X\\x06\\x00\\x00\\x00mem_mbqwM\\xb8\\x0bhjK<hlKdubX\\x03\\x00\\x00\\x00logqxcsnakemake.io\\nLog\\nqy)\\x81qzXz\\x00\\x00\\x00/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/parse_tree/all/chromosome/filter50/parse_tree_processed.py.ipynbq{a}q|(h\\t}q}X\\x08\\x00\\x00\\x00notebookq~K\\x00N\\x86q\\x7fsh\\x0f]q\\x80(h\\x11h\\x12eh\\x11h\\x13h\\x18\\x85q\\x81Rq\\x82(h\\x18)}q\\x83h\\x1ch\\x11sNtq\\x84bh\\x12h\\x13h\\x18\\x85q\\x85Rq\\x86(h\\x18)}q\\x87h\\x1ch\\x12sNtq\\x88bh~h{ubX\\x06\\x00\\x00\\x00configq\\x89}q\\x8a(X\\x0b\\x00\\x00\\x00results_dirq\\x8bX\\x07\\x00\\x00\\x00resultsq\\x8cX\\t\\x00\\x00\\x00sqlite_dbq\\x8dX\\x0b\\x00\\x00\\x00test.sqliteq\\x8eX\\x19\\x00\\x00\\x00sqlite_select_command_asmq\\x8fX\\xc8\\x00\\x00\\x00SELECT AssemblyFTPGenbank FROM BioSample LEFT JOIN Assembly ON AssemblyBioSampleAccession = BioSampleAccession WHERE (BioSampleComment LIKE \\'%KEEP%Assembly%Modern%\\' AND length(AssemblyFTPGenbank) > 0)q\\x90X\\x19\\x00\\x00\\x00sqlite_select_command_sraq\\x91X\\xbe\\x00\\x00\\x00SELECT BioSampleAccession, SRARunAccession FROM BioSample LEFT JOIN SRA ON SRABioSampleAccession = BioSampleAccession WHERE (SRARunAccession = \\'SRR1048902\\' OR SRARunAccession = \\'SRR1048905\\')q\\x92X\\x1b\\x00\\x00\\x00sqlite_select_command_localq\\x93XU\\x00\\x00\\x00SELECT BioSampleAccession FROM BioSample WHERE (BioSampleComment LIKE \\'%Local%Test%\\')q\\x94X\\x19\\x00\\x00\\x00sqlite_select_command_refq\\x95X\\xaa\\x00\\x00\\x00SELECT AssemblyFTPGenbank FROM BioSample LEFT JOIN Assembly ON AssemblyBioSampleAccession = BioSampleAccession WHERE (BioSampleComment LIKE \\'%Assembly%Modern%Reference%\\')q\\x96X\\x15\\x00\\x00\\x00max_datasets_assemblyq\\x97K\\x03X\\x10\\x00\\x00\\x00max_datasets_sraq\\x98K\\x02X\\x0c\\x00\\x00\\x00reads_originq\\x99]q\\x9a(X\\x08\\x00\\x00\\x00assemblyq\\x9bX\\x03\\x00\\x00\\x00sraq\\x9cX\\x05\\x00\\x00\\x00localq\\x9deX\\x18\\x00\\x00\\x00detect_repeats_thresholdq\\x9eKZX\\x15\\x00\\x00\\x00detect_repeats_lengthq\\x9fK2X\\x0f\\x00\\x00\\x00reference_locusq\\xa0X\\x08\\x00\\x00\\x00AL590842q\\xa1X\\x14\\x00\\x00\\x00reference_locus_nameq\\xa2X\\n\\x00\\x00\\x00chromosomeq\\xa3X\\x15\\x00\\x00\\x00reference_locus_startq\\xa4X\\x01\\x00\\x00\\x000q\\xa5X\\x13\\x00\\x00\\x00reference_locus_endq\\xa6X\\x07\\x00\\x00\\x004653728q\\xa7X\\t\\x00\\x00\\x00eager_revq\\xa8X\\x05\\x00\\x00\\x002.2.1q\\xa9X\\x15\\x00\\x00\\x00eager_clip_readlengthq\\xaaK#X\\r\\x00\\x00\\x00eager_bwaalnnq\\xabG?\\x84z\\xe1G\\xae\\x14{X\\r\\x00\\x00\\x00eager_bwaalnlq\\xacK\\x10X\\x0b\\x00\\x00\\x00eager_otherq\\xadX\\x0c\\x00\\x00\\x00--mergedonlyq\\xaeX\\x08\\x00\\x00\\x00organismq\\xafX\\x0f\\x00\\x00\\x00Yersinia pestisq\\xb0X\\x15\\x00\\x00\\x00eager_forward_adapterq\\xb1X@\\x00\\x00\\x00AGATCGGAAGAGCACACGTCTGAACTCCAGTCACNNNNNNATCTCGTATGCCGTCTTCTGCTTGq\\xb2X\\x15\\x00\\x00\\x00eager_reverse_adapterq\\xb3X:\\x00\\x00\\x00AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATTq\\xb4X\\x10\\x00\\x00\\x00snippy_ctg_depthq\\xb5K\\nX\\x10\\x00\\x00\\x00snippy_bam_depthq\\xb6K\\x03X\\x10\\x00\\x00\\x00snippy_base_qualq\\xb7K\\x14X\\x0f\\x00\\x00\\x00snippy_map_qualq\\xb8K\\x1eX\\x0f\\x00\\x00\\x00snippy_min_fracq\\xb9G?\\xec\\xcc\\xcc\\xcc\\xcc\\xcc\\xcdX\\x10\\x00\\x00\\x00snippy_mask_charq\\xbaX\\x01\\x00\\x00\\x00Xq\\xbbX\\x13\\x00\\x00\\x00snippy_missing_dataq\\xbcK2X\\x12\\x00\\x00\\x00snippy_snp_densityq\\xbdK\\nX\\x15\\x00\\x00\\x00snippy_keep_singletonq\\xbeX\\x10\\x00\\x00\\x00--keep-singletonq\\xbfX\\x1e\\x00\\x00\\x00snippy_multi_plot_missing_dataq\\xc0]q\\xc1(K\\x00K\\x19K2KKKdeX\\x0c\\x00\\x00\\x00iqtree_modelq\\xc2X\\x06\\x00\\x00\\x00-m MFPq\\xc3X\\x0b\\x00\\x00\\x00iqtree_seedq\\xc4X\\x08\\x00\\x00\\x0047321424q\\xc5X\\x0f\\x00\\x00\\x00iqtree_outgroupq\\xc6X\\t\\x00\\x00\\x00Referenceq\\xc7X\\x0c\\x00\\x00\\x00iqtree_otherq\\xc8X\\r\\x00\\x00\\x00--ufboot 1000q\\xc9X\\x0b\\x00\\x00\\x00iqtree_runsq\\xcaK\\x01uX\\x04\\x00\\x00\\x00ruleq\\xcbX\\n\\x00\\x00\\x00parse_treeq\\xccX\\x0f\\x00\\x00\\x00bench_iterationq\\xcdNX\\t\\x00\\x00\\x00scriptdirq\\xceXD\\x00\\x00\\x00/mnt/c/Users/ktmea/Projects/plague-phylogeography/workflow/notebooksq\\xcfub.'); from snakemake.logging import logger; logger.printshellcmds = True; import os; os.chdir('/mnt/c/Users/ktmea/Projects/plague-phylogeography');\n",
    "######## snakemake preamble end #########\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Objectives\n",
    "\n",
    "1. Convert newick tree to PhyloXML format.\n",
    "1. Fixes internal node names from an IQTREE tree. \n",
    "   - internal nodes by default get named the confidence (ex. \"99/100\")\n",
    "   - this notebook instead gives them the name NODEi where i is an integer starting from 0.\n",
    "   - NODE0 therefore becomes the root.\n",
    "1. Convert tree to data frame.\n",
    "1. Combine metadata with tree data frame.\n",
    "1. Export:\n",
    "   - Tree (newick, nexus, phyloxml)\n",
    "   - Tree Dataframe (tsv)\n",
    "   - Color mapping (tsv)\n",
    "   - Lat Lon Mapping for Auspice (tsv)\n",
    "   - Lat Lon Mapping for BEAST (tsv)   \n",
    "   - Augur JSON\n",
    "   - Auspice JSON"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Module Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# stats and plotting\n",
    "import pandas as pd\n",
    "from matplotlib import colors, gridspec\n",
    "import math\n",
    "\n",
    "# Phylogenetics\n",
    "from Bio import Phylo\n",
    "import treetime # Clock filter tips\n",
    "\n",
    "# System misc\n",
    "\n",
    "import time # Augur export\n",
    "import os\n",
    "import copy\n",
    "from augur import utils, export_v2\n",
    "import json\n",
    "\n",
    "# Logging to file\n",
    "import sys\n",
    "import io\n",
    "\n",
    "# Writing python objects to file\n",
    "import dill"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from config import *\n",
    "\n",
    "# Custom script variables\n",
    "SCRIPT_NAME = \"parse_tree\"\n",
    "\n",
    "try:\n",
    "    WILDCARDS = snakemake.wildcards\n",
    "    project_dir = os.getcwd()\n",
    "except NameError:\n",
    "    WILDCARDS = [\"all\", \"chromosome\", \"50\"]\n",
    "    #WILDCARDS = [\"all\", \"chromosome\", \"5\"]\n",
    "    project_dir = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "    \n",
    "READS_ORIGIN = WILDCARDS[0]\n",
    "LOCUS_NAME = WILDCARDS[1]\n",
    "MISSING_DATA = WILDCARDS[2]\n",
    "\n",
    "\n",
    "NAME_COL = \"sample\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Input File Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tree_path: /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/iqtree/all/chromosome/filter50/iqtree_post.cf.rerofot.nwk\n",
      "metadata_path: /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/metadata/all/metadata.tsv\n",
      "aln path: /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/snippy_multi/all/chromosome/filter50/snippy-multi.snps.aln\n",
      "auspice_config_path: /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json\n",
      "auspice_remote_dir_path: /mnt/c/Users/ktmea/Projects/plague-phylogeography/auspice/\n",
      "outdir: /mnt/c/Users/ktmea/Projects/plague-phylogeography/results/parse_tree/all/chromosome/filter50\n"
     ]
    }
   ],
   "source": [
    "config_dir = os.path.join(project_dir, \"results\", \"config\")\n",
    "results_dir = os.path.join(project_dir, \"results\")\n",
    "outdir = os.path.join(results_dir, \n",
    "                      SCRIPT_NAME,\n",
    "                      READS_ORIGIN,\n",
    "                      LOCUS_NAME,\n",
    "                      \"filter{}\".format(MISSING_DATA),\n",
    "                     )\n",
    "\n",
    "tree_path = os.path.join(\n",
    "    results_dir, \n",
    "    \"iqtree\", \n",
    "    READS_ORIGIN, \n",
    "    LOCUS_NAME,    \n",
    "    \"filter{}\".format(MISSING_DATA),    \n",
    "    \"iqtree_post.cf.rerofot.nwk\",\n",
    ")\n",
    "\n",
    "tree_alt_path = os.path.join(\n",
    "    results_dir, \n",
    "    \"iqtree\", \n",
    "    READS_ORIGIN, \n",
    "    LOCUS_NAME,    \n",
    "    \"filter{}\".format(MISSING_DATA),    \n",
    "    \"iqtree.treefile\",\n",
    ")\n",
    "\n",
    "metadata_path = os.path.join(\n",
    "    results_dir,\n",
    "    \"metadata\",\n",
    "    READS_ORIGIN,  \n",
    "    \"metadata.tsv\"\n",
    ")\n",
    "\n",
    "aln_path = os.path.join(\n",
    "    results_dir,\n",
    "    \"snippy_multi\",\n",
    "    READS_ORIGIN,\n",
    "    LOCUS_NAME,\n",
    "    \"filter{}\".format(MISSING_DATA),       \n",
    "    \"snippy-multi.snps.aln\",\n",
    ")\n",
    "\n",
    "auspice_config_path = os.path.join(config_dir, \"auspice_config.json\")\n",
    "auspice_remote_dir_path = os.path.join(project_dir, \"auspice/\")\n",
    "\n",
    "print(\"tree_path:\", tree_path)\n",
    "print(\"metadata_path:\", metadata_path)\n",
    "print(\"aln path:\", aln_path)\n",
    "print(\"auspice_config_path:\", auspice_config_path)\n",
    "print(\"auspice_remote_dir_path:\", auspice_remote_dir_path)\n",
    "print(\"outdir:\", outdir)\n",
    "# Create output directory if it doesn't exist\n",
    "if not os.path.exists(outdir):\n",
    "    os.makedirs(outdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Metadata Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>42126</td>\n",
       "      <td>2006</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>42123</td>\n",
       "      <td>2005</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>9_10</td>\n",
       "      <td>1923</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN02442721</th>\n",
       "      <td>Justinian</td>\n",
       "      <td>[500:800]</td>\n",
       "      <td>[-1521:-1221]</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>51.0834</td>\n",
       "      <td>10.4234</td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT4</td>\n",
       "      <td>SAMN02442721</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>0</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN02442718</th>\n",
       "      <td>Justinian</td>\n",
       "      <td>[500:800]</td>\n",
       "      <td>[-1521:-1221]</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>51.0834</td>\n",
       "      <td>10.4234</td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT4</td>\n",
       "      <td>SAMN02442718</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>0</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>testlocal1</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>testlocal1</td>\n",
       "      <td>KEEP: Local Test</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>testlocal2</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>testlocal2</td>\n",
       "      <td>KEEP: Local Test</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        strain       date        date_bp  \\\n",
       "sample                                                                     \n",
       "GCA_009669545.1_ASM966954v1_genomic      42126       2006            -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic      42123       2005            -16   \n",
       "GCA_009909635.1_ASM990963v1_genomic       9_10       1923            -98   \n",
       "SAMN02442721                         Justinian  [500:800]  [-1521:-1221]   \n",
       "SAMN02442718                         Justinian  [500:800]  [-1521:-1221]   \n",
       "testlocal1                                  NA         NA             NA   \n",
       "testlocal2                                  NA         NA             NA   \n",
       "\n",
       "                                     country       province country_lat  \\\n",
       "sample                                                                    \n",
       "GCA_009669545.1_ASM966954v1_genomic    China       Xinjiang     35.0001   \n",
       "GCA_009669555.1_ASM966955v1_genomic    China       Xinjiang     35.0001   \n",
       "GCA_009909635.1_ASM990963v1_genomic   Russia  Rostov Oblast     64.6863   \n",
       "SAMN02442721                         Germany        Bavaria     51.0834   \n",
       "SAMN02442718                         Germany        Bavaria     51.0834   \n",
       "testlocal1                                NA             NA          NA   \n",
       "testlocal2                                NA             NA          NA   \n",
       "\n",
       "                                    country_lon province_lat province_lon  \\\n",
       "sample                                                                      \n",
       "GCA_009669545.1_ASM966954v1_genomic         105      42.4805      85.4633   \n",
       "GCA_009669555.1_ASM966955v1_genomic         105      42.4805      85.4633   \n",
       "GCA_009909635.1_ASM990963v1_genomic     97.7453      47.6222      40.7958   \n",
       "SAMN02442721                            10.4234      48.9468      11.4039   \n",
       "SAMN02442718                            10.4234      48.9468      11.4039   \n",
       "testlocal1                                   NA           NA           NA   \n",
       "testlocal2                                   NA           NA           NA   \n",
       "\n",
       "                                         biovar branch_major branch_minor  \\\n",
       "sample                                                                      \n",
       "GCA_009669545.1_ASM966954v1_genomic     Antiqua        0.ANT       0.ANT1   \n",
       "GCA_009669555.1_ASM966955v1_genomic     Antiqua        0.ANT       0.ANT1   \n",
       "GCA_009909635.1_ASM990963v1_genomic  Medievalis        2.MED       2.MED1   \n",
       "SAMN02442721                         Orientalis        0.ANT       0.ANT4   \n",
       "SAMN02442718                         Orientalis        0.ANT       0.ANT4   \n",
       "testlocal1                                   NA           NA           NA   \n",
       "testlocal2                                   NA           NA           NA   \n",
       "\n",
       "                                    biosample_accession  \\\n",
       "sample                                                    \n",
       "GCA_009669545.1_ASM966954v1_genomic        SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic        SAMN07722924   \n",
       "GCA_009909635.1_ASM990963v1_genomic        SAMN13632815   \n",
       "SAMN02442721                               SAMN02442721   \n",
       "SAMN02442718                               SAMN02442718   \n",
       "testlocal1                                   testlocal1   \n",
       "testlocal2                                   testlocal2   \n",
       "\n",
       "                                         biosample_comment branch_number  \\\n",
       "sample                                                                     \n",
       "GCA_009669545.1_ASM966954v1_genomic  KEEP: Assembly Modern             0   \n",
       "GCA_009669555.1_ASM966955v1_genomic  KEEP: Assembly Modern             0   \n",
       "GCA_009909635.1_ASM990963v1_genomic  KEEP: Assembly Modern             2   \n",
       "SAMN02442721                             KEEP: SRA Ancient             0   \n",
       "SAMN02442718                             KEEP: SRA Ancient             0   \n",
       "testlocal1                                KEEP: Local Test            NA   \n",
       "testlocal2                                KEEP: Local Test            NA   \n",
       "\n",
       "                                    continent  \n",
       "sample                                         \n",
       "GCA_009669545.1_ASM966954v1_genomic      Asia  \n",
       "GCA_009669555.1_ASM966955v1_genomic      Asia  \n",
       "GCA_009909635.1_ASM990963v1_genomic    Europe  \n",
       "SAMN02442721                           Europe  \n",
       "SAMN02442718                           Europe  \n",
       "testlocal1                                 NA  \n",
       "testlocal2                                 NA  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Parse the metadata\n",
    "metadata_df = pd.read_csv(metadata_path, sep='\\t')\n",
    "\n",
    "# Fix the problem with multiple forms of NA in the table\n",
    "# Consolidate missing data to the NO_DATA_CHAR\n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "\n",
    "# set the metadata index\n",
    "metadata_df.set_index(NAME_COL, inplace=True)\n",
    "\n",
    "metadata_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 1. Convert tree to PhyloXML format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp_xml = os.path.join(outdir, \"temp.xml\")\n",
    "try:\n",
    "    Phylo.convert(tree_path, \"newick\", tmp_xml, \"phyloxml\")\n",
    "except FileNotFoundError:\n",
    "    Phylo.convert(tree_alt_path, \"newick\", tmp_xml, \"phyloxml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 2. Fix internal node names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in XML tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the new tree\n",
    "tree = Phylo.read(tmp_xml, \"phyloxml\")\n",
    "\n",
    "# Deepest nodes first\n",
    "tree.ladderize(reverse=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fix node names and confidences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Counter for integer node names\n",
    "node_i = 0\n",
    "\n",
    "# Iterate through the nodes in the tree\n",
    "for c in tree.find_clades():\n",
    "    # Check if the name has 'accidentally' become the confidence vals\n",
    "    ufboot_val = NO_DATA_CHAR    \n",
    "    scf_val = NO_DATA_CHAR\n",
    "    # If it's a named node, parse it\n",
    "    if c.name:\n",
    "        name_split = c.name.split(\"/\")\n",
    "        \n",
    "        # Check if the name has 'accidentally' become the confidence vals\n",
    "        if len(name_split) > 1:\n",
    "            # Name the internal node\n",
    "            c.name = \"NODE\" + str(node_i)    \n",
    "\n",
    "            # Assign confidence values     \n",
    "            ufboot_val = float(name_split[0])\n",
    "            scf_val = float(name_split[1])\n",
    "    # If it's not a named node, give it a name\n",
    "    else:\n",
    "        c.name = \"NODE\" + str(node_i) \n",
    "        # The one and only confidence is \"unknown\" and if UFboot\n",
    "        if len(c.confidences) > 0:\n",
    "            ufboot_val = c.confidences[0].value\n",
    "        c.confidences = []\n",
    "        \n",
    "    \n",
    "    ufboot_conf = Phylo.PhyloXML.Confidence(ufboot_val, type=\"ufboot\")\n",
    "    scf_conf = Phylo.PhyloXML.Confidence(scf_val, type=\"scf\")\n",
    "    \n",
    "    c.confidences.append(ufboot_conf)\n",
    "    c.confidences.append(scf_conf)\n",
    "        \n",
    "    # Increment node counter if not a terminal\n",
    "    if not c.is_terminal():\n",
    "        node_i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAACB8AAAWMCAYAAABsp7w2AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAD2EAAA9hAHVrK90AAB4XUlEQVR4nOzde7Sld13f8c/3MATEXAQjOjAgN7WhEwXLxWAlArUXDeqySlUUInLtKu1i2iraVlrU1WrraS1eY9EkVq0oRcZWxUshXhItUl1SEWm4KDmMUEAIEDTA/PrH2aezM8xkLt99zpP9nNdrrb328+z927/zDfzDIu/1/GqMEQAAAAAAAACA87Ux9QAAAAAAAAAAwHoTHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALQcmHoA9q+qOpjkqpM+fkuS2yYYBwAAAAAAAGCV7pXkISd99t/GGMemGGa3iQ+Y0lVJrpl6CAAAAAAAAIA98uwkPzL1ELvBsQsAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC0Hph6Afe0tJ3/wwz/8w7n88sunmAUAAAAAAABgZV7/+tfnOc95zskff9y/I50L8QFTuu3kDy6//PJcccUVU8wCAAAAAAAAsNs+7t+RzoVjFwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPlgzVXXfqrqqql5cVb9YVe+uqrF4XTv1fAAAAAAAAADsPwemHoBz9s6pBwAAAAAAAACAZZ58sN7enuSXpx4CAAAAAAAAgP3Nkw/Wz4uTvDbJa8cY76yqByV567QjAQAAAAAAALCfiQ/WzBjjRVPPAAAAAAAAAADLHLsAAAAAAAAAALSIDwAAAAAAAACAFscucN6q6ormFodXMggAAAAAAAAAkxIf0HHj1AMAAAAAAAAAMD3xAdzFbW5uZnNzc+oxAAAAAPbMkSNHcuTIkanHAAAAzoH4AO7CNjc3c/To0WxtbU09CgAAAMCeufXWW6ceAQAAOEfiA7gL29zczNbWVjY2NnLw4MGpxwEAAADYExdffPHUIwAAAOdIfEDH45q/P5zkmlUMMncHDx7MLbfcMvUYAAAAAAAAAKckPuC8jTFu6vy+qlY1CgAAAAAAAAAT2ph6AAAAAAAAAABgvYkPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaDkw9AOemqv56koctfXTp0vXDqurq5fVjjGv3YCwAAAAAAAAA9jHxwfp5ZpKnn+a7z1+8ll27q9MAAAAAAAAAsO85dgEAAAAAAAAAaBEfrJkxxtVjjDrb19TzAgAAAAAAADB/4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaDkw9QCcm6r63CR/O8kXJDmc5L5JPpLkHUluTPLSMcZvTDchAAAAAAAAAPuN+GCNVNUNSR5/iq8uSPIZi9fTq+rHkzxzjHH7Xs4HAAAAAAAAwP4kPlgv91+8vyPJzyT5jSR/muRuSa5I8o8Xa74+2//dfu0EMwIAAAAAAACwz4gP1ssbk3xrkpePMT520ne/vXjiwW8l+cwkX1NVP+gIBgAAAAAAAAB228bUA3D2xhhXjTFedorwYOf7d2f76Qc7vnJvJgMAAAAAAABgPxMfzM9rlq4fOtUQAAAAAAAAAOwf4oP5uWDp+vhkUwAAAAAAAACwbxyYegBW7sql6zfu5h+qqiuaWxxeySAAAAAAAAAATEp8MCNVtZHkhUsfvWyX/+SNu7w/AAAAAAAAAGtAfDAvL0jymMX1K8YYvzvlMKzOsWPHcujQoanHAAAAAPbQkSNHcuTIkanHAAAAOCvig5moqiuT/JvF7buSPG/CcVix48ePZ2tra+oxAAAAgD106623Tj0CAADAWRMfzEBV/dUkr8j2f59/meQpY4x3TjsVq7SxsZGDBw9OPQYAAACwhy6++OKpRwAAADhr4oM1V1UPTvLLSe6d5GNJvmaMccMe/fnHNX9/OMk1qxhk7g4ePJhbbrll6jEAAAAAAAAATkl8sMaq6n5JfjXJ/ZKMJM8YY7xir/7+GOOmzu+ralWjAAAAAAAAADChjakH4PxU1aVJfiXJQxYfPX+Mcf2EIwEAAAAAAACwT4kP1lBVXZLkVUkevvjohWOM759wJAAAAAAAAAD2MfHBmqmqeyX570k+d/HRd44xvmvCkQAAAAAAAADY58QHa6SqLkjyiiSfv/joe8cY/3zCkQAAAAAAAAAgB6YegHPyU0n+5uL6fyR5aVUdvpP1t48x3rT7YwEAAAAAAACwn4kP1stXLF0/MckfnGH9nyR50K5NAwAAAAAAAABx7AIAAAAAAAAA0OTJB2tkjFFTzwAAAAAAAAAAJ/PkAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQMuBqQfg7FXVxUm+OMmjkzwqyf2TfEqST0jyviRvSPILSV46xnjPRGMCAAAAAAAAsM+ID9bLY5L81Gm++5QkVy5e/7Sqvm6M8ao9mwwAAAAAAACAfUt8sH7enuTVSV63uD6W7eMzDiX5yiRfkeTSJEer6tFjjD+YalAAAAAAAAAA9gfxwXp59RjjgXfy/cuq6suTvCLJBUlelOTv7sVgAAAAAAAAAOxfG1MPwNkbY3zsLNb8XJI3Lm4fv6sDAQAAAAAAAEDEB3P1ocX7PSedAgAAAAAAAIB9QXwwM1V1WZJHLG7feCdLAQAAAAAAAGAlDkw9AH1Vda8k90/y5CTflORui6++d5f/7hXNLQ6vZBAAAAAAAAAAJiU+WFNVdXWSH7uTJf8uyU/s8hg37vL+AAAAAAAAAKwB8cH8/H6S544xfmfqQQAAAAAAzmRzczNJcuTIkYknAQCgY2PqAThvP5fk8sXrMUm+JskrkjwiyU9U1VWTTQYAAAAAcBY2Nzdz9OjRvOpVr5p6FAAAmjz5YE2NMd6X5H1LH702yX+pqq9Pcl2SV1bVN44xrt376QAAAAAAzmxzczNbW1u55JJLph4FAIAm8cHMjDF+fPHUg6ck+b6qeuUY48936c89rvn7w0muWcUgAAAAAMD6uvDCC6ceAQCAJvHBPL0y2/HBJyb5O0l+cjf+yBjjps7vq2pVowAAAAAAAAAwoY2pB2BX/N+l60+fbAoAAAAAAAAA9gXxwTzdf+n6g5NNAQAAAAAAAMC+ID6Yp69aun79ZFMAAAAAAAAAsC+ID9ZIVV1dVfc8w5oXJPnixe3bkvzmbs8FAAAAAAAAwP52YOoBOCf/Msn3VNXLsx0VvDnbxypclOTyJE9N8vmLtbcnedYY46MTzAkAAAAAAADAPiI+WD/3SfKsxet0bknyjDHGr+7NSAAAAAAAAADsZ+KD9fKkJH8jyROSXJbkU5N8cpK/SPLOJL+f5L8ledkY47aJZgQAAAAAAABgnxEfrJExxpuzfdTCD089CwAAAAAAAADs2Jh6AAAAAAAAAABgvYkPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAA2HObm5s5duzY1GMAALAi4gMAAAAAAPbU5uZmjh49muPHj089CgAAK3Jg6gEAAAAAANhfNjc3s7W1NfUYAACskCcfAAAAAAAwiY0N/xc1AMBc+F92AAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRH8xIVX13VY2l1xdOPRMAAAAAAAAA8yc+mImq+pwkL5h6DgAAAAAAAAD2H/HBDFTVRpIfSXIgybsmHgcAAAAAAACAfUZ8MA//MMmjk7wxyUsnngUAAAAAAACAfUZ8sOaq6gFJvn1x+7wkt084DgAAAAAAAAD7kPhg/f1AkguTXDfGeM3EswAAAAAAAACwD4kP1lhVPSXJVUnem+SfTjwOAAAAAAAAAPvUgakH4PxU1Scl+d7F7TePMf7vBDNc0dzi8EoGAQAAAAAAAGBS4oP19d1JPi3JjUleOtEMN070dwEAAAAAAAC4CxEfrKGq+utJnpnko0meO8YYE48EAAAAAHDejh07lkOHDp1x3ZEjR3LkyJE9mAgAgHMlPlgzVXVBkmuSVJJ/P8Z4/cQjAQAAAACclwsvvDCPfOQjc8MNN2Rra+uM62+99dY9mAoAgPMhPlg/35rksiR/muRfTTwLAAAAAMB5u+iii/KlX/qlufnmm89q/cUXX7zLEwEAcL7EB2ukqv5Kkm9Z3D5/jPGhKedJ8rjm7w9n+ykOAAAAAMA+5SgFAIB5EB+slxckuSDJW5Lcq6q++hRrDi9dP7GqPm1x/fOrjhXGGDd1fl9VqxoFAAAAAAAAgAmJD9bLPRbvD0nyU2ex/l8sXT84ydRPSgAAAAAAAABghjamHgAAAAAAAAAAWG/igzUyxrh6jFF39kryr5Z+8oSl79420dgAAAAAAAAAzJz4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSID2ZmjPEvxxi1eL1m6nkAAAAAAAAAmD/xAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAlgNTD7CbqupvJ3likscmeWCST0pyUZI6x63GGGPW/1kBAAAAAAAAwPma5b9Qr6qvTvLiJA9d/niicQAAAAAAAABg1mYXH1TVDyV5Vk7EBuOk93Pesj0UAAAAAAAAAMzYrOKDqnpxkmcvbke2w4HleGAnQDhdUHCm7wEAAAAAAACAk2xMPcCqVNXhJP8s2wHBTnjwmiRfnOQ+SZ6XEzHCGGNsJLkoyYOTPCXJjyf5yM73SW5IcmiMsTHGuNue/sMAAAAAAAAAwBqZTXyQ5IW54xMLrhljPHGM8UtjjPed6gdjjA+NMf5kjPGzY4ynJ/mMbEcHleTxSW6qqgfv9uAAAAAAAAAAsM5mER9U1d2SfFlOPPHg/yR5/rnuM8Z4e5InJTm62OcBSX6+qi5Y3bQAAAAAAAAAMC+ziA+SPDLJJy6uR5LvG2N85Hw2GmMcT/K1SbYWH12W5AXtCQEAAAAAAABgpuYSH3zW4n3n2IVfPtMPquq0/+xjjNuS/NulPf9BazoAAAAAAAAAmLG5xAf3Wbr+2BjjTadYc/yk+3ueYc+fW7q+X1V99vkMBgAAAAAAAABzN5f44KKl61tPs+ZDJ91ffGcbjjH+NMnt2T7GIUnEBwAAAAAAAABwCnOJD25bur7badZ84KT7+5/Fvh/IiaMcPu1chwIAAAAAAACA/WAu8cF7lq4vOs2aY4v3nScZXHZnG1bVgSSXLK2/+3lPBwAAAAAAAAAzNpf44I+XrquqHniKNX+Y5PjS/ZVn2PPzkhxYun/f+Y0GAAAAAAAAAPM2l/jgDdkOC3aeUnD45AVjjA9nO0BIto9SeEpV3ftO9vwnS2uT5M0rmBMAAAAAAAAAZmcW8cEY44NJfj8nQoHTPdXgZxZrRpILk/x0VV2yvKC2fUeSL82JmOEjSX5rxWMDAAAAAAAAwCzMIj5Y+LWl6y85zZofTXLb0v2Tkrylqq6vqu+oqu9L8qYk37L4fidU+IkxxodWPTAAAAAAAAAAzMGc4oOfXbxXksuq6rEnLxhjvCPJd+bEExIqyb2TPDXbwcHzkjw0J6KDJHlPkm/bvbEBAAAAAAAAYL0dmHqAVRljvLaqXpZk5xiFRyX5nVOs+9dV9dAkz8iJwOAOSxbvleR9Sf7uGGNr9RMDAAAAAAAAwDzMJj5IkjHGV5/lumdW1a8neXGSB55qSZJXJPmmMcabVzgiAAAAAAAAAMzOrOKDczHGuD7J9VX115J8dpL7Jrk9yS1JbhhjvGvK+QAAAAAAAABgXezb+GDHGON1SV439RwAAAAAAAAAsK42ph4AAAAAAAAAAFhv4gMAAAAAAAAAoEV8cBaq6qFTzwAAAAAAAAAAd1WziQ+q6rqq+oRd2PfvJXndqvcFAAAAAAAAgLmYTXyQ5OuTvLaqLlvFZlV1QVX9QJKfTHLRKvYEAAAAAAAAgDmaU3yQJJdlO0B4WmeTxTELNyV5TpJaxWAAAAAAAAAAMFdziw+S5F5JfqyqfvR8jmFYOmbhEREeAAAAAAAAAMAZzSk+eP/ifWQ7Gnh6kt+pqs86mx8vjln4wWwfs3Dx0l5J8tJVDgoAAAAAAAAAczKn+OBzk/yvbIcHOwHC4SS/W1Vfd2c/rKqHJfntJM8+6fe3JXnaGOPZuzg3AAAAAAAAAKy12cQHY4y3Jnlcku/PiYBgJPnEJNdV1Y9U1T1P/l1VfXW2j1n4nNwxPPjfSR41xvjPe/NPAAAAAAAAAADraTbxQZKMMT4yxnh+kqck+cDOx9mOCZ6R7WMYPjP5/8cs/FCSn0hy0UlrfyzJY8YYf7yX8wMAAAAAAADAOppVfLBjjPGzSR6V5A9yx6cZXJ7tYxj+UZLfSfKsnPqYhW8cY/zFFLMDAAAAAAAAwLqZZXyQJGOMm5M8NsmP5ERgkCQXJtnMxx+z8IdJHu2YBQAAAAAAAAA4N7OND5JkjHH7GOM5SZ6a5EPZDg12YoOxtPTabB+z8MY9HxIAAAAAAAAA1tys44MlNyV5a7ajg+SO4cFbknzzGOPDez4VAAAAAAAAAMzA7OODqvqyJK9Lcjh3jA6yuH9Ikt+rqsfv9WwAAAAAAAAAMAezjQ+q6m5VtZnkvya5d0489aCS/PHSfZLcL8mvVdW37u2UAAAAAAAAALD+ZhkfVNUDkvxmkn+U7chg54kHtyV52hjj4Um+OcnHFp+PJHdL8u1V9UtVdekejwwAAAAAAAAAa2t28UFVXZXk95I8JifCg0ryhiSPGWP85yQZY/zbJFcm2Tpp3Rcl+f2q+oK9nx4AAAAAAAAA1s9s4oPFMQvfneSVSe6z+HgnKLg+yaPHGH+0/Jsxxk1JPifJL+SOT0jYOYbhW/ZidgAAAAAAAABYZ7OJD5LckOQf545PMfiLJM8cY1w9xvjwqX40xvjzMcZVSV6YOx7DcCDJd1TVLzqGAQAAAAAAAABOb07xweMW7zvhwZuSfN4Y40fP5sdjjO9O8oX5+GMY/la2j3EAAAAAAAAAAE5hTvHBjkryX5I8aozx+nP54RjjxiSPSPKL+fhjGAAAAAAAAACAU5hbfHB7kr8/xvjaMcYHz2eDMcZ7xxhfkuRbcuIYBgAAAAAAAADgNOYUH7w528cs/NAqNhtjfFeSJ2T7GAYAAAAAAAAA4DQOTD3ACv21Mcatq9xwjPFbVfWIJNevcl8AAAAAAAAAmJPZxAerDg+W9n1Pki/Zjb0BAAAAAAAAYA7mdOwCAAAAAAAAADAB8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQMuBqQfYbVX18CSPTfLAJJ+U5KKce3QxxhjfuOLRAAAAAAAAAGAWZhkfVNW9k/yDJM9L8qnd7ZKMJOIDAAAAAAAAADiF2cUHVfX4JD+Z5GC2w4HTGcs/O4s1AAAAAAAAAMApzCo+qKrHJfmVbP9z7TyxYDksuLPg4OTQoE6xBgAAAAAAAAA4yWzig6q6OMnLk9w9J0KC25O8MslvJ3lEkqctPh9JviHJhUkuTfLoJI9f3O/89jVJrtv9yQEAAAAAAABgvc0mPkjyvCSfmhPxwM1JrhpjvClJquo5OREfZIxxh7Cgqi5K8veTfFuSeya5MslbkzxrjHF816cHAAAAAAAAgDW1MfUAK/TcnDhm4S+S/J2d8OBsjDE+MMb4riSPSfLOxT5XJ3nJ6kcFAAAAAAAAgPmYRXxQVQ9I8umL25HkP40x3nw+e40x/jDJk5N8NNsBwnOr6kkrGRQAAAAAAAAAZmgW8UGSRy/ea/H+053NxhivS/KjSx/9i85+AAAAAAAAADBnc4kPLj3p/vdOsWYs31TVPc6w53U7S5N8QVXd9zxnAwAAAAAAAIBZm0t8cJ+l6w+OMT58ijV/edL9J5xhz/+Z7WBhJ1p47HnOBgAAAAAAAACzNpf44GNL1ydHBjs+cNL9wTvbcIzxsSR/nhNHOTz4/EYDAAAAAAAAgHmbS3zw/qXri0+z5s9Pun/QWex7z5x48sGF5zgTAAAAAAAAAOwLc4kP3rZ0ffeqOlWA8EeL97M6RqGqHpTkXksf3Xa+wwEAAAAAAADAnM0lPnjDSfeHT14wxvizJO9e3FaSp5xhz69bWpskf3be0wEAAAAAAADAjM0iPhhj3JLkHUsfPeo0S4/mREzwWVX1/FMtqqpHJnlhTjwlIUlu6s4JAAAAAAAAAHM0i/hg4dVL1198mjXXL95HtiOE/1BVL62qJ1bVw6rqUVX1oiSvyfaRC7VY+7tjjD/ZpbkBAAAAAAAAYK0dmHqAFfr5JE9dXF9ZVZeOMd69vGCM8etV9QvZjhN2AoSrF69ltfR9krxol2YGAAAAAAAAgLU3pycf/HySD2c7GLhHkueeZt2zk7wtJwKDLK6XX8vHLfy7McYv7cK8AAAAAAAAADALs3nywRjjw1X1gCR3X3z0l6dZ946qujLJtUmesPzV4n0nQLgtyYvGGN+zOxMDAAAAAAAAwDzMJj5IkjHGe89y3duTPKmqnpDkK5J8dpL7Jrk9yS1JXpPk+jHGO3dpVAAAAAAAAACYjVnFB+dqjPHqJK+eeg4AAAAAAAAAWGcbUw8AAAAAAAAAAKy32Tz5oKqetnR70xjj/zT3+4wkV+zcjzGu7+wHAAAAAAAAAHM1m/ggybVJxuL6eUla8UGSJyb5gaV78QEAAAAAAAAAnMLcjl2oXdhv1XsCAAAAAAAAwKzMLT4YZ14CAAAAAAAAAKzS3OIDAAAAAAAAAGCPiQ9O7xOWrj882RQAAAAAAAAAcBcnPji9hy5df2CyKQAAAAAAAADgLu7A1APcFVXVvZJ8WZKx+OjNE44DAAAAAAAAAHdpaxMfVNUDkzzoLJd/ZlU9/hz/xN2SXJzksiRPS3Jo8flI8rpz3AsAAAAAAAAA9o21iQ+SfEOSbzvDmlq8v2DxOl+V7ehgZ7+fbuwFAAAAAAAAALO2TvFBciIGWNW60xlL168cY9zY3A8AAAAAAAAAZmtj6gHOwzjzkradJx9cm+Tr9uDvAQAAAAAAAMDaWqcnH7wtyQ138v2VOREm3JzkHee4/0eTfCDJO5P8XpJfGGPcco57AAAAAAAAAMC+szbxwRjjuiTXne77qjq+dPs9Y4xrdn8qAAAAAAAAAGAdj124MzX1AAAAAAAAAACw36zNkw/Owq/nxLEL53rkAgAAAAAAAABwnmYTH4wxvnDqGQAAAAAAAABgP5rbsQsAAAAAAAAAwB4THwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER+smaoaZ/l6zdSzAgAAAAAAALA/iA8AAAAAAAAAgJYDUw/AefvBJD9wJ99/aK8GAQAAAAAAAGB/Ex+sr3eNMf731EMAAAAAAAAAgGMXAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaDkw9AOftq6rqa5I8MMlHk/xZkhuTXDvGePVeDFBVVzS3OLySQQAAAAAAAACYlPhgfT38pPuHLV5Pq6qfS3L1GOP9uzzDjbu8PwAAAAAAAABrQHywfm5LcjTJryV5Y5IPJvmUJFcmeW6ST07y5UleWVVfNMb4yERzskLHjh3LoUOHph4DAAAAAFbi2LFjU48Aa2tzczNJcuTIkYknAYA7Eh+sn/uPMd53is9/papekuQXkzwy2zHC85L8xz2cjRU7cuRIjh49mhtuuCFbW1tTjwMAAAAAwIQ2Nzdz9OjR3OMe9xAfAHCXIz5YM6cJD3a+e2dVfWWSP0pyQZLnR3yw1nb+x+PNN9888SQAAAAAsDrHjh3L8ePHpx4D1s7m5ma2trZyySWXTD0KAHwc8cHMjDHeUlW/kuRLkjysqu43xnjHLv25xzV/fzjJNasYZM6OHDmiYAUAAABgVg4dOuRJn9Bw4YUXTj0CAHwc8cE8vSHb8UGS3D/JrsQHY4ybOr+vqlWNAgAAAAAAAMCENqYegF3h3+oDAAAAAAAAsGfEB/P08KXr3TpyAQAAAAAAAACSiA9mp6oekuSLFrdvGWM4OA0AAAAAAACAXSU+WCNV9eSqOnAn339qkp9NcvfFR9+/J4MBAAAAAAAAsK+d9l9kc5f0kiR3r6qXJ7kpyduSfDjJpUm+MMlzk3zyYu1vRnwAAAAAAAAAwB4QH6yf+yV5/uJ1Oi9P8swxxl/uzUgAAAAAAAAA7Gfig/Xy9CRXJrkiyUOy/cSDi5N8MMnbk9yY5Loxxk2TTQgAAAAAAADAviM+WCNjjBuS3DD1HAAAAAAAAACwbGPqAQAAAAAAAACA9SY+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANByYOoBAAAAAAAA5mZzczObm5sr3fPYsWMr3Q8AVkl8AAAAAAAAsEKbm5s5evRotra2ph4FAPaM+AAAAAAAAGCFNjc3s7W1lY2NjRw8eHBl+x47dizHjx9f2X4AsEriAwAAAAAAgF1w8ODB3HLLLSvb79ChQ56mAMBd1sbUAwAAAAAAAAAA6018AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtIgPAAAAAAAAAIAW8QEAAAAAAAAA0CI+AAAAAAAAAABaxAcAAAAAAAAAQIv4AAAAAAAAAABoER8AAAAAAAAAAC3iAwAAAAAAAACgRXwAAAAAAAAAALSIDwAAAAAAAACAFvEBAAAAAAAAANAiPgAAAAAAAAAAWsQHAAAAAAAAAECL+AAAAAAAAAAAaBEfAAAAAAAAAAAt4gMAAAAAAAAAoEV8AAAAAAAAAAC0iA8AAAAAAAAAgBbxAQAAAAAAAADQIj4AAAAAAAAAAFrEBwAAAAAAAABAi/gAAAAAAAAAAGgRHwAAAAAAAAAALeIDAAAAAAAAAKBFfAAAAAAAAAAAtByYegDOX1VdmuQZSb4syUOT3DvJe5K8PcmvJ/mvY4ybppsQAAAAAAAAgP1AfLCmquqrkvxgkk8+6auDi9djknxGki/f28kAAAAAAAAA2G/EB2uoqp6W5MeyfWzGu7IdIfxmkvcm+bRsPwXhyUk+MtWMAAAAAAAAAOwf4oM1U1WXJbkm2+HBbyR58hjj/adY+pKqumBPhwMAAAAAAABgX9qYegDO2UuS3CPJu5N8xWnCgyTJGOP2PZsKAAAAAAAAgH1LfLBGquqvJHnS4vb7xhjvnnIeAAAAAAAAAEgcu7Buvmrp+md2Lqrq3kkuTfLeMcZ79nwqAAAAAAAAAPY18cF6+bzF+/uT/FFVPTXJNyX57J0FVfXWJNcl+Z4xxgd3c5iquqK5xeGVDALA/2vvzsNku8p68X/fkIkkQIAACmGQQQMCCogyKIMgyqAM4gAKBEHB6QJHVHAixJ8X9adxunJlUpCrqKiEyAwSIjILIiAociGKISAzJAgxyXv/qDqeOpXuPl29uqtOn/58nqee2nvX3nu91afXWd29v7U2AAAAAADASgkf7C63mD6fn+R3kvzoGvt8RZIzkjy4qr61uz+yg/W8cQfPDQAAAAAAAMAucdSqC2Ah15g+n5ZJ8OAzSR6b5NpJjk9y+yQvn+5zyyQvrCr/xgAAAAAAcAS56KIdnfgYALbEzAe7y4nT5+OSXJbk3t395pnX/66q7pfkJUnuneROSR6U5M+XWiUAAAAAALDt9u3bl3POOSfHHXfcqksBgCsQPthdvpgDAYQXzgUPkiTdfXlV/WQm4YMkeUiEDwAAAAAAYNfbt2/fqksAgHUJH+wun8+B8MHL19upu/+xqi5Icr1MbsWwU+40ePwtkzxzOwoBAAAAAIC9QAABgMOV8MHu8uEkXzZd/vdN7Hu9JNfeqWK6+00jx1fVdpUCAAAAAAAAwAodteoCWMg/zixf6RD77n/90h2qBQAAAAAAAACSCB/sNn8zs3yTQ+x74+nzBTtUCwAAAAAAAAAkET7Ybc5J8l/T5Qett1NV3TXJNaerr9/pogAAAAAAAADY24QPdpHu/mSSZ09Xv6Wqvnd+n6q6SpLfnNn0jCWUBgAAAAAAAMAeJnyw+zwlyb9Nl59fVb9TVXevqttV1elJ3prka6ev/+/uftsKagQAAAAAAABgDzl61QWwmO7+eFV9Wya3YLhpkh+bPub9fpLHLbM2AAAAAAAAAPYmMx/sQt39vkxmN/jJJG9J8qkklyT59yR/muSbu/tR3f1fKysSAAAAAAAAgD3DzAe7VHdfnOTXpg8AAAAAAAAAWBkzHwAAAAAAAAAAQ4QPAAAAAAAAAIAhwgcAAAAAAAAAwBDhAwAAAAAAAABgiPABAAAAAAAAADBE+AAAAAAAAAAAGCJ8AAAAAAAAAAAMET4AAAAAAAAAAIYIHwAAAAAAAAAAQ4QPAAAAAAAAAIAhwgcAAAAAAAAAwBDhAwAAAAAAAABgiPABAAAAAAAAADBE+AAAAAAAAAAAGCJ8AAAAAAAAAAAMET4AAAAAAAAAAIYIHwAAAAAAAAAAQ4QPAAAAAAAAAIAhwgcAAAAAAAAAwBDhAwAAAAAAAABgiPABAAAAAAAAADBE+AAAAAAAAAAAGCJ8AAAAAAAAAAAMET4AAAAAAAAAAIYIHwAAAAAAAAAAQ4QPAAAAAAAAAIAhwgcAAAAAAAAAwBDhAwAAAAAAAABgiPABAAAAAAAAADBE+AAAAAAAAAAAGCJ8AAAAAAAAAAAMET4AAAAAAAAAAIYIHwAAAAAAAAAAQ4QPAAAAAAAAAIAhwgcAAAAAAAAAwBDhAwAAAAAAAABgiPABAAAAAAAAADBE+AAAAAAAAAAAGCJ8AAAAAAAAAAAMET4AAAAAAAAAAIYIHwAAAAAAAAAAQ4QPAAAAAAAAAIAhwgcAAAAAAAAAwBDhAwAAAAAAAABgiPABAAAAAAAAADBE+AAAAAAAAAAAGCJ8AAAAAAAAAAAMET4AAAAAAAAAAIYIHwAAAAAAAAAAQ4QPAAAAAAAAAIAhwgcAAAAAAAAAwBDhAwAAAAAAAABgiPABAAAAAAAAADBE+AAAAAAAAAAAGCJ8AAAAAAAAAAAMET4AAAAAAAAAAIYIHwAAAAAAAAAAQ45edQEAAAAAAOxNF154YU499dRVlwHb7sILL1x1CQCwdMIHAAAAAAAs1b59+3LOOefkvPPOywUXXLDqcgAA2AbCBwAAAAAALNW+ffuSJB/4wAdWXAnsjAsvvDCXX375qssAgKUSPgAAAAAAYOn27dv33yEEONKceuqpZvUAYM85atUFAAAAAAAAAAC7m/ABAAAAAAAAADBE+AAAAAAAAAAAGCJ8AAAAAAAAAAAMET4AAAAAAAAAAIYIHwAAAAAAAAAAQ4QPAAAAAAAAAIAhwgcAAAAAAAAAwBDhAwAAAAAAAABgiPABAAAAAAAAADBE+AAAAAAAAAAAGCJ8AAAAAAAAAAAMET4AAAAAAAAAAIYIHwAAAAAAAAAAQ4QPAAAAAAAAAIAhwgcAAAAAAAAAwBDhAwAAAAAAAABgiPABAAAAAAAAADBE+AAAAAAAAAAAGCJ8AAAAAAAAAAAMET4AAAAAAAAAAIYIHwAAAAAAAAAAQ4QPAAAAAAAAAIAhwgcAAAAAAAAAwBDhAwAAAAAAAABgiPABAAAAAAAAADBE+AAAAAAAAAAAGCJ8AAAAAAAAAAAMET4AAAAAAAAAAIYIHwAAAAAAAAAAQ4QPAAAAAAAAAIAhwgcAAAAAAAAAwBDhAwAAAAAAAABgiPABAAAAAAAAADBE+AAAAAAAAAAAGCJ8AAAAAAAAAAAMET4AAAAAAAAAAIYIHwAAAAAAAAAAQ4QPAAAAAAAAAIAhwgcAAAAAAAAAwBDhAwAAAAAAAABgiPABAAAAAAAAADBE+AAAAAAAAAAAGCJ8AAAAAAAAAAAMET4AAAAAAAAAAIYIHwAAAAAAAAAAQ4QPAAAAAAAAAIAhwgcAAAAAAAAAwBDhAwAAAAAAAABgiPABAAAAAAAAADBE+AAAAAAAAAAAGCJ8AAAAAAAAAAAMET4AAAAAAAAAAIYIHwAAAAAAAAAAQ4QPAAAAAAAAAIAhwgcAAAAAAAAAwBDhAwAAAAAAAABgiPDBLlJVr6uqXvBxt1XXDQAAAAAAAMCRTfjgyHZ5kn9ZdREAAAAAAAAAHNmOXnUBLOSRSU48xD63SPKn0+W/7u4LdrYkAAAAAAAAAPY64YNdpLs/dKh9quphM6t/uIPlAAAAAAAAAEASt104olTVUUm+b7p6UZK/XGE5AAAAAAAAAOwRwgdHlnskud50+c+7+wurLAYAAAAAAACAvUH44Mjy8Jllt1wAAAAAAAAAYCmOXnUBbI+qOinJA6er/5bkdUto846Dp7jlthQCAAAAAAAAwEoJHxw5vjPJidPl53d3L6HNNy6hDQAAAAAAAAAOc267cORwywUAAAAAACBnnXVWzjrrrFWXAcAeI3xwBKiqU5Pcbbr65u5+/wrLAQAAAAAAVuSss87KOeeck1e+8pWrLgWAPcZtF44M358DQZLnrbIQAAAAAABgdc4666xccMEFudrVrrbqUgDYY4QPjgwPmz5/KcmfLrHdOw0ef8skz9yOQgAAAAAAgANOOumkVZcAwB4jfLDLVdXXJbnFdPUl3f3pZbXd3W8aOb6qtqsUAAAAAAAAAFboqEPvwmHu4TPLbrkAAAAAAAAAwNIJH+xiVXVMku+drn48yctXWA4AAAAAAAAAe5Twwe527yTXmi7/cXdfuspiAAAAAAAAANibhA92t9lbLvzhyqoAAAAAAAAAYE8TPtilqurqSe43XX1Pd79jlfUAAAAAAAAAsHcJH+xe35PkuOmyWQ8AAAAAAAAAWBnhg93rYdPny5L80SoLAQAAAAAAAGBvO3rVBbA13X3nVdcAAAAAAAAAAImZDwAAAAAAAACAQcIHAAAAAAAAAMAQ4QMAAAAAAAAAYIjwAQAAAAAAAAAwRPgAAAAAAAAAABgifAAAAAAAAAAADBE+AAAAAAAAAACGCB8AAAAAAAAAAEOEDwAAAAAAAACAIcIHAAAAAAAAAMAQ4QMAAAAAAAAAYIjwAQAAAAAAAAAwRPgAAAAAAAAAABgifAAAAAAAAAAADBE+AAAAAAAAAACGCB8AAAAAAAAAAEOEDwAAAAAAAACAIcIHAAAAAAAAAMAQ4QMAAAAAAAAAYIjwAQAAAAAAAAAwRPgAAAAAAAAAABgifAAAAAAAAAAADBE+AAAAAAAAAACGCB8AAAAAAAAAAEOEDwAAAAAAAACAIcIHAAAAAAAAAMAQ4QMAAAAAAAAAYIjwAQAAAAAAAAAwRPgAAAAAAAAAABgifAAAAAAAAAAADBE+AAAAAAAAAACGCB8AAAAAAAAAAEOEDwAAAAAAAACAIcIHAAAAAAAAAMAQ4QMAAAAAAAAAYIjwAQAAAAAAAAAwRPgAAAAAAAAAABgifAAAAAAAAAAADBE+AAAAAAAAAACGCB8AAAAAAAAAAEOEDwAAAAAAAACAIcIHAAAAAAAAAMAQ4QMAAAAAAAAAYIjwAQAAAAAAAAAwRPgAAAAAAAAAABgifAAAAAAAAAAADBE+AAAAAAAAAACGCB8AAAAAAAAAAEOEDwAAAAAAAACAIcIHAAAAAAAAAMAQ4QMAAAAAAAAAYIjwAQAAAAAAAAAwRPgAAAAAAAAAABgifAAAAAAAAAAADBE+AAAAAAAAAACGCB8AAAAAAAAAAEOEDwAAAAAAAACAIUevugAAAAAAAIAj0YUXXphTTz116W0CwCoIHwAAAAAAAGyjffv25Zxzzsl5552XCy64YNXlAMBSCB8AAAAAAABso3379iVJPvCBDyy97QsvvDCXX3750tsFAOEDAAAAAACAbbZv377/DiEs06mnnmq2BQBW4qhVFwAAAAAAAAAA7G7CBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOGDXaqqjq2qR1XVK6rqwqr6UlVdVFX/XFW/X1V3WHWNAAAAAAAAAOwNR6+6ABZXVddP8tIkt5p76dgkXzl9PLKqfiPJT3R3L7lEAAAAAAAAAPYQMx/sMlV1dA4OHrwryelJ7pjkXknOTHLx9LUnJHnikksEAAAAAAAAYI8x88Huc/8cCB68Kck3dfdlM6+/uqrOmb52TJInV9VvdPelS64TAAAAAAAAgD3CzAe7z51nlp82FzxIknT325O8ZLp69SSnLaMwAAAAAAAAAPYm4YPd59iZ5Q9usN//nVk+bodqAQAAAAAAAADhg13o/TPLN95gv5tMnzvJv+xcOQAAAAAAAADsdUevugAW9oIkv5jkqkl+uqpeNn/rhaq6TZL7Tlf/pLs/txOFVNUdB09xy20pBAAAAAAAAICVEj7YZbr741V1epI/SnLnJG+rqt/MZEaEk6bbfiKT2zO8M8m+HSznjTt4bgAAAAAAAAB2CeGDXai7X1RVX5dJsOAHkjxvbpePJXlKkmd298XLrg8AAAAAAFitiy66aNUlALDHCB/sQlV1TJKHJvn2JLXGLtdJ8pBMZkN46RJLAwAAAAAAVmjfvn0555xzctxxx626FAD2GOGDXaaqTkzysiR3SXJZkl9N8gdJPpjk+CTfkOQXknxjkr+qqid092+tqFwAAAAAAGCJ9u3bybsxA8D6hA92n6dmEjxIkkd19+wtFy5J8uqqOjfJq5LcPclZVXVud79rB2q50+Dxt0zyzO0oBAAAAAAAmBBAAGAVhA92kaqqJI+crr5/Lnjw37r70qr6+SR/m+So6TFP2O56uvtNI8dP3g4AAAAAAAAAu91Rqy6AhVwnyTWmy39/iH3fPrN82s6UAwAAAAAAAADCB7vNpTPLh5q14ph1jgMAAAAAAACAbSV8sLt8Ksnnpst3rKqNAgh3nVn+0M6VBAAAAAAAAMBeJ3ywi3T35UleOl29bpKfXWu/qrp6kl+Z2fSSHS4NAAAAAAAAgD3sUFP3c/g5M8n9k5yQ5Iyqul2S5yX5YJLjk9whyeOT3GC6/19396tWUCcAAAAAAAAAe4TwwS7T3f9UVfdP8oIkpyT59uljLa9N8l3Lqg0AAAAAAACAvUn4YBfq7tdU1WlJHpXk3km+OsnJSS5N8tEkb0vyx0nO6e5eVZ0AAAAAAAAA7A3CB7tUd38yya9OHwAAAAAAAACwMketugAAAAAAAAAAYHcTPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAw5etUFsKedML/h3e9+9yrqAAAAAAAAANhW61z7vMI10iNFdfeqa2CPqqofTPLMVdcBAAAAAAAAsCQ/1N3PWnURO8FtFwAAAAAAAACAIcIHAAAAAAAAAMAQ4QMAAAAAAAAAYEh196prYI+qqi9Pcr+5zR9M8oUVlHO4uWWSZ85t+6Ek71lBLbCX6HuwGvoerIa+B6uh78Fq6HuwGvoerIa+B6uxVt/7H0m+OLftJd194XJKWq6jV10Ae9e0Uz1r1XUcjqpqrc3v6e43LbsW2Ev0PVgNfQ9WQ9+D1dD3YDX0PVgNfQ9WQ9+D1Vin7/3dXup7brsAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMET4AAAAAAAAAAAYInwAAAAAAAAAAAwRPgAAAAAAAAAAhggfAAAAAAAAAABDhA8AAAAAAAAAgCHCBwAAAAAAAADAEOEDAAAAAAAAAGCI8AEAAAAAAAAAMKS6e9U1AAAAAAAAAAC7mJkPAAAAAAAAAIAhwgcAAAAAAAAAwBDhAwAAAAAAAABgiPABAAAAAAAAADBE+AAAAAAAAAAAGCJ8AAAAAAAAAAAMET4AAAAAAAAAAIYIHwAAAAAAAAAAQ4QPAAAAAAAAAIAhwgcAAAAAAAAAwBDhAwAAAAAAAABgiPABAAAAAAAAADBE+AAAAAAAAAAAGCJ8AAAAAAAAAAAMET4AAAAAAAAAAIYIHwAAAAAAAAAAQ4QPAAAAAAAAAIAhwgcAAAAAAAAAwBDhAwAAAAAAAABgiPABAAAAAAAAADBE+AB2WFXdoKp+rareV1UXV9WnquqtVfXEqjphG9v53qp6ZVVdWFVfrKrzq+r5VXWH7WoDdpOd7HtVdUZV9SYfd9uedwSHr6q6dlXdr6rOrKqXV9UnZvrAc3eoTeMee96y+p5xDw5WVbetqp+Z9rsPV9WXquqiqnp/VT23qr5pB9o07rHnLavvGffggKq66nQM+vWqOq+qPlBVn62qS6rqP6rqdVX1U1V1zW1s05jHnresvmfMg8VU1a/uRL840sa+o1ddABzJquq+Sf4oydVmNp+Q5PbTx6Or6j7d/cGBNo5P8sIk95t76YbTx0Or6ozu/sWttgG7zTL6HnCQjy2rIeMeHGRpfQ+YqKrzktxljZeOTXKz6eMRVfX8JI/u7ksG2zPuQZbf94D/9vVJXrDOa9dKctfp4yer6vu7+5VbbciYBwdZWt8DNqeqvibJE7b5nEfk2Cd8ADtk+h/Rn2VywfOiJE9Lcm6SKyf53iQ/mOSrkry0qm7f3Rdtsann5MB/TOcm+a0kH0lyqyQ/k+QmSc6sqgu7+9lbbAN2jSX2vf1udYjXPzR4fthtPpzkfUnutUPnN+7B2na67+1n3GOvu970+SOZ/JHo9Un+LcmVktwxyU9M93lYJn9zeehge8Y9mFh239vPuAeTnzPPTfL26fKFmcyofGqSByd5UJJTkpwz/TvLu7bYjjEPDrasvrefMQ/WUVVHJXlWJj9n/keSa2/TqY/Isa+6e9U1wBGpqs5Ncrcklya5S3e/ae71n0zyq9PVp3T3mVto465JXjdd/askD+zuy2ZePyWTH05ukOTTSW7c3Z9ZtB3YTZbU985I8pQk6e4aqReOBFX11CRvS/K27v5YVd0oB34pfV53n75N7Rj3YMYS+94ZMe5BkqSqXpLkD5P8xewYNPP6KUnekOQrp5vu0t2v32Jbxj2YWnLfOyPGPUiSVNWV1upzc/s8IMmLpqt/2d3fuYV2jHkwY4l974wY8+CQqurxSX4jyT9l0u+ePH3p7t39ui2e84gd+45adQFwJKqq22dy8TNJnjN/8XPq1zP5dFqSPL6qjtlCUz81fb4syY/M/0DS3Z9I8tPT1asnedQW2oBdY4l9D5jR3U/p7pd0905PAW/cgxlL7HvAVHffr7v/bL0/Bk/Hop+Y2fTggeaMezC15L4HTB3q4ud0n7MzuRiTrH17lM0w5sGMJfY94BCq6vpJ9t/24IeTbNftvY7YsU/4AHbGA2aW/2CtHbr78kxS+8nkP467LdJAVZ2U5B7T1Vd397+vs+tfJvncdPlBi7QBu9ADZpZ3pO8Bq2HcA2AXed3M8k22cgLjHmzJ62aWt9T3gC27ePp8/KIHGvNgyJb7HrBpT09yUiYzTL5uO054pI99wgewM75p+nxxJtOirOe8meVvXLCNr09y3BrnOUh3X5LkzfuP8SlvjnDL6HvAahj3ANgtjp1ZvnyL5zDuweK2o+8BC6qqmyf52unqP22w63qMebAF29D3gEOoqu9Ocr8kn0ryk9t46iN67BM+gJ1x8+nzB7r70g32m/2h4Obr7rVxG/Pn2aido5PcbMF2YDdZRt87SFW9uqo+WVWXVNV/VNXrqupJVXX1kfMCV2Dcg8OAcQ825a4zy1v9Q7BxDxa3HX3vIMY9WFtVnVBVN6uqfUnOTXKl6Uu/tYXTGfNgk7a5782f25gHM6rq5BzoWz/d3R/fxtMf0WOf8AFss6o6Pskp09X1pkpJknT3p3NgaqTrL9jU7P4btpPkw+scB0eMJfa9efdMco0kxyS5ViZ/8Hpakg9W1f0Hzw0cYNyDw4NxDzZQVUcledLMpj/b4qmMe7CAbex784x7MFVVp1dVV1Vn8jeV9yf59STXme7ya0n+aAunNubBBnaw780z5sHBfjXJlyV5Y5LnbPO5j+ix7+hVFwBHoKvMLF+0if0vTnJiJveM2al2Lp5ZXrQd2C2W1ff2e3eSs5O8NclHMvnB/KuSfF+SeyU5OclfVNW3d/fLt9gGcIBxD1bLuAeb84RMptBMkhd1999t8TzGPVjMdvW9/Yx7sHnvTPLY7n7LFo835sHWvDNjfW8/Yx7MqapvTPLoJJdm0s96m5s4osc+4QPYfsfPLF+yif2/NH2+8g6286WZ5UXbgd1iWX0vSX6zu89YY/tbkvxhVT0mye9lMvXZs6vqpt39n1toBzjAuAerY9yDTaiquyb55enqfyT54YHTGfdgk7a57yXGPVjP2Un2B3uunOQmSb47yQOT/FFVPb67X7KF8xrzYGNnZ2f6XmLMgyuoqmOTPDNJJfmN7n73DjRzRI99brsA2++LM8vHbmL/46bPiw7ai7Rz3MyyHw44Ui2r76W7P3OI15+R5NnT1esmedCibQBXYNyDFTHuwaFV1VcneVEmH/L4UpLv7u6PDZzSuAebsAN9z7gH6+juz3T3e6aPt3X3n3T3g5I8PMmNk7y4qk7fwqmNebCBHex7xjxY288kuXmSf0vy1B1q44ge+4QPYPt9fmZ5M1OgnDh93sw08Vtt58SZ5UXbgd1iWX1vs54xs3zXHWoD9hLjHhzejHvsWVX1FUleleTqSS5L8pDuPm/wtMY9OIQd6nubZdyDqe5+fpIXZnKt4X9V1dUXPIUxD7ZgG/reZhnz2DOq6rQkT56u/nh3X7zR/gOO6LFP+AC2WXd/McknpqunbrTv9AeC/f9xfHjBpv59ZnnDdpJcf2Z50XZgV1hi39us984sX2+H2oC9xLgHhzfjHntSVV03yWsy+SRYJ/mB7n7RNpzauAcb2MG+t1nGPTjYi6fPJya594LHGvNg60b63mYZ89hLnpDJTAQfTHJCVX3v/CPJLWf2/+aZ105c84xrO6LHvqNXXQAcod6X5JuS3LSqju7uS9fZ77S5YxYxO+iftu5eB79+aZIPLNgO7CbL6HubVTt0XtirjHtweDPusedU1SlJXp3JdLfJ5JMxf7hNpzfuwTp2uO9tuowltweHu4/PLN9wwWONebB1I31vs4x57CX7b3Fw4yQv2MT+Pz+z/BVJNjtTwhE99pn5AHbG306fT0xyuw32m52m6A0LtvG2JJescZ6DVNWxSe6w/5juvmS9feEIsIy+t1m3mFn+yA61AXuJcQ8Ob8Y99pSqulqSV+bA9/6Tuvt3t7EJ4x6sYQl9b7OMe3Cw2U9DLzoltDEPtm6k722WMQ+23xE99gkfwM44e2b5kWvtUFVHJXn4dPUzSc5dpIHu/nySv56u3rOq1pua5UFJrjpdXuYUhLAKZ88s70jfW8BjZpaXdd9ROGIZ9+CwZ9xjz6iqE5K8NMltp5t+qbt/ZTvbMO7BFS2j7y3AuAcH+66Z5XcvcqAxD4Zsue8twJjHntHdp3d3bfRI8tSZQ+4+89r5C7RzRI99wgewA7r7rUleP119VFXdcY3dfiLJzafLv9Xd/zX7YlWdXlU9fZyxTlO/Nn0+OsnvVtWV5s5xSpL9v4h/JsmzF3ojsMsso+9V1a2q6qYb1VFVj0nyqOnqR7OLfjCAVTHuwWoY92Ax00+evCjJnaebfqu7f24L5zHuwQKW1feMe3CwaZ85/hD7PCHJfaar5+fArJSz5zDmwQKW0feMebBz9vrYd/SqC4Aj2OMymc79ykleVVX/M5NPWF85yfcm+aHpfu9P8utbaaC7X1tVfzI933ckeXVV/WYm0x/dKsnPJrnBdPcndfent/ZWYFfZ6b53uyTPrqpzk7w8k1TxJzMZU09L8v1JvmW672VJHtPdm73XE+xKVfWNSWZ/YT1lZvmmVXX67P7d/dyttGPcg4Mtqe8Z9+BgL0hyr+nya5M8p6puucH+l3T3+7fSkHEPDrKsvmfcg4OdkeTXq+ovMrmw+X8zmdr9KpmMRd+XA6GgS5L8YHdfumgjxjy4gjOy833PmAcrdCSPfcIHsEO6+++r6nuS/J9MpkX5n2vs9v4k951OsbJVPzA9/32S3H36mHV5kl/s7mcMtAG7xpL63pWS3HP6WM8nkzyqu8/ZYhuwmzw6ySPWee3OOfAL8X7PHWjLuAcHLKvvGffggAfNLH9zkncdYv9/TXKjgfaMezCxzL5n3IODXSPJD04f6/n3JD/Q3a8ZaMeYBwdbRt8z5sFqHZFjn/AB7KDu/ququnUmn8S+b5JTM0kifiDJC5P8r+7+wmAb/5nkvlX10CSnJ/maJCcn+Vgm08//r+5+00gbsNvscN97WSbTjd0xyW2SXCfJNZNUkk8l+Yckr0jy3O7+3Mj7AK7IuAdLZ9yDFTLuwdIZ9+Bg98jkouTdM7mF5f4+8cVMxqJ3JnlJkj/zN07YVsvoe8Y8WLEjdeyr7l51DQAAAAAAAADALnbUqgsAAAAAAAAAAHY34QMAAAAAAAAAYIjwAQAAAAAAAAAwRPgAAAAAAAAAABgifAAAAAAAAAAADBE+AAAAAAAAAACGCB8AAAAAAAAAAEOEDwAAAAAAAACAIcIHAAAAAAAAAMAQ4QMAAAAAAAAAYIjwAQAAAAAAAAAwRPgAAAAAAAAAABgifAAAAAAAAAAADBE+AAAAAAAAAACGCB8AAAAAAAAAAEOEDwAAAAAAAACAIcIHAAAAAAAAAMAQ4QMAAAAAAAAAYIjwAQAAAAAAAAAwRPgAAAAAAAAAABgifAAAAAAAAAAADBE+AAAAAAAAAACGCB8AAAAAAAAAAEOEDwAAAAAAAADYsqq6dlXdr6rOrKqXV9Unqqqnj+euur7tMvOeDvV43aprXYWjV10AAAAAAAAAALvax1ZdAKsnfAAAAAAAAADAdvlwkvcludeqC9lB/zvJ0zd4/eJlFXI4ET4AAAAAAAAAYMSZSd6W5G3d/bGqulGSD622pB31H939nlUXcbg5atUFAAAAwHrm75m46npYjqo6f+7f/karrmktVXXGXJ1nrLomFlNVN5r7Nzx/1TUBAMBu1N1P6e6XdLfbL+xhwgcAAAAAAAAAHBaq6uiqelRVvayqPlJVX6qqT1TV31TV46vq+FXXyNrcdgEAAACAXaeqvjbJA2Y2vbO7z15JMQAAwLaoqpskOSfJLeZeumaSb5o+fqSq7tvd/7Ls+tiY8AEAAAAAu9HXJnnKzPrzkpy9kkoAAIBhVfXlSd6Q5DpJPp/kmUlek+RjSa6W5F5JHpfkZkleUVW37e7Prqjc76qqhyS5QZJLk3w0yRuTPLe7z11RTSsnfAAAAAAAAADAqj0zk+DBh5Pcrbs/OPf666rqhUlen+TGSZ6Y5OeXW+J/m5+Z4abTx8Or6uwkp68wGLEyR626AAAAAAAAAAD2rqq6ZZL7TVd/bI3gQZKku/8+ye9OV39gGbXN+UKSP0nyg5ncAuI2mczI8EtJPjnd5wFJXlxVx6ygvpUy8wEAAAAAAAAAq3T/6fMXkrz0EPv+TZKfSnLdqrp+d394Rys72PW6+zNrbH91Vf1OkpdnEki4a5IfTvLbS6xt5cx8AAAAAAAAAMAqfd30+YQkl1ZVr/dI8pKZ475s9iRVdc+Njl3gcaO1ilwneLD/tY8leXCSS6abfnyLX4tdS/gAAAAAAAAAgFW69haPO2Fbqxg0vV3Eq6erN62q666ynmVz2wUAAAB2rao6Psmdktw8yclJPpvkgiRv6O7/2IH2KsnXJrlVJp+uOCbJp5K8uLs/cojjbjyt8wZJrpqkp8d+LMlbpp+Q2BFVdZ1Mvk5fkeT4abvnJ3l9d1+8je0cm+TrM3mPpyS5SpKLk3w0yXuTvLe7L92mtr4myW2TXCfJZUk+nuQfkryzu3s72limqjopyR2TXDfJtZJUJu/pQ0ne1N2XbHD4Vts8Ksntk9x62uYXp22+rbv/aZvaOC7JXZLcMJM/Jn4xyb9m8p7W7TOHk+l9Wu+Y5KuTXCOT7+mPJXljd//rKmsDAIAjyJWmzx9K8h0LHPehufU3ZvK796gLBo59b5L7Tpevl2RX/O6zHYQPAAAA2HWq6ppJzkjyyCQnrrHLZVV1bpIndffbFzjv+ZlcJN3vK7r7/Ko6IckTM7lf45etceiFSc6eO9fJSR6UyR8c7prkmodo+31JfifJH3T3Fxeo+YwkT5nZ9NTuPmP62m2T/GKSb8vasx9+qar+OMnPd/eW/7BSVfdK8vhM3udGnzr5XFX9dZI/TPKSRYMI04vlj07y05mEOdZyQVX9UpJnbVfQYSdV1f2TPC7JnZMcu85uF1XVOUnO6O5/WeDcB4Uwurum249N8oRM/s3W+n5OVb0/yS90959utr2546+d5JeSfE8mIZR5Pe2jv9Ddb5ge89wkj5jZ55Hd/dy5856e5A/WafYRVfWIdV5LkvO6+26bqX/a1lWS/EySx2YSblprn7cneXJ3v3qt1wEAgE375PT5Okn+aau/z3X3F5JsS5h6QK24/ZVx2wUAAAB2lar6uiTvSfJjWTt4kEw+MXHPJG+pqqcOtvdVSd6Z5KlZ50LtGsfcOpNPRj8nkwDChsGDqZsneXqSd1XV8Kc0quqJSd6a5D5Z//f/4zIJcLxzGlRYtI0bV9Ubkrwyyb1z6Okur5rkgUlelOT7FmzrGplMXfmMrB88SCafKnl6krOr6sqLtLFMVXWTqnpTJqGVu2f94EGSnJTkoUneW1U/O9juDTP5vvjlbPz9/JVJ/qSqnjUNfSzSxn0z+WPfo7N28CCZ/DHum5O8vqp+fpHzL8N0Zo13J3lS1gkeTN0uyasOx/cAAAC7zN9Pn0/IJJy9m91iZnnPzHqQCB8AAACwi1TVzZK8Kle8aPrZJP+1xiFXSvILVfUrW2zyekn+OsnN5rZ/IclFGxx3Qta/mHxJJp/o+MI6r98syZur6rQF6jxIVf1ckv8/B6atTJJLM7ndwmVrHHJKkldW1bUWaOPuSd6Wye0c1nLZtL0vrXeKzbaVyQXs12RysXrWF5J8bp1j7pvk9xZoY2mq6k5J3pzkDuvsclHWfl9HJ/n/qmpL76uqrpfkvCRfM/fS5zO5lcBaHp2DZ9Y4VBv3S/KXSa6+xsuXJ/l0Dv4erCRnTsMyh4WqumWSc3PwLCjJ5P+Z/1znsDOr6pE7WhgAABzZXjyz/FMrq2JQVd04ybdMVz84MsvgbiR8AAAAwG7y/By4qPm2JA9OclJ3n5zJp/i/KpPbDMxfIPypqvruLbT3zEwCCEnyjiQPSXLN7j6xu6+SySeiH5or3mNyvy8kOSeTadu/IclVuvu47j6lu0+cvpd7ZXIbgtkLsldN8sLp9PiLuleSM6fLn0rys5l86uLY7r5mJl+nu2QS4ph1SpJf20wDVfXVSf4qyTXmXvr7TC5W3yjJMd19ze4+PpOZH74lya9m/a/VRn4vyW2my2/NZCr//f8OV5vW8ahMbn8x6+FVNR9YWKmqukmSl2fy9d7vi5l8r90tyYndfZXp+7pKJvc6PW/uNI+pqsduofk/y4EL6q9K8u1JrtrdV+3uk5J8eZJ9mVxkn/XkqvrKQ528qm6Q5I9zcPCmM7lNwp2SHNfd18jke/A2SX4zk1BMkjwtya028R7elMntT344k34z680zr631OGsT579yJjNzXH1a+59kMovKlbv75O4+IZOZN87M5N9t1llVtVboAgAAOITuflsO/J56n0PNYlhVN6qqh+x8ZQe1+e1VdfQGr18nyZ8nOWa66XeXUthhpLr70HsBAADACszfs37G7yZ5XHev9Sn+TG9b8NocPEPCfyS5eXd/aoP2zs8VP+2cTC7K/3R3X77Jur8qk9sQ/H53r/fJ/Pljvj7JS3PwRenTu/t5hzjujKz9yfQ3J/mO7v74OsdVkt9PcvrM5kuSXK+7P7FBeydkchuK2dkgLkvyuCRP70P8oWHa7v2TfKK7/3adfc7P2v8OT+nuM9fYvv+46yd5Y5JTZza/qLsftFFNW7XG1/6p3X3GBvsfm8nF89lbXPxjkgd394b3JK2qJ2VygX6/Lyb5yu7+8AbHrPVvcWmSx3T3729w3K2T/G0OvmXCb3T3vkPU+LJMvu/3+1KSB3b3yzc45usz+QPj1dZ4+ZHd/dwNjj09k2DDfs/r7tM3qnGNc9woawdiLkryPd39sg2OvWeSV+TgGUYe192/vUgNAABwJKiqb0xy05lNp2QyI1+SvCHJs2f3X+tn/aq6bpK/yyQYnSRvyeT31ndn8jvQNZPcOsm3ZTIz3tnd/eBtexOHMP1d9Zgkf5HJ73bnZ/Lhh1MyCZM/Ngduu/i3Se7Z3evNBnhEWjeZAQAAAIeplyb58Y0ucnf3+6b3nX9rDlwYvHaSH8rkXveL+NPu/slFDujuf07yzwse89bp7Ayvndn8o0k2DB+s41+T3Ke7P71Be11VP57JxeLrTDcfm0kw4DkbnPsxueJtKB7R3X+0mcKm/25nb2bfOb+9UfBgeu4PV9UTkrxwZvP9qur47p7/lPoqPCwHBw/+Nck9uvtjhzqwu395+oe4H59uOj6TwMeityv4qY2CB9O23lVVZ+bAHwqT5LsymRVhTVV1qxwcPEiSx24UPJi29daq+q5ccSaOVXvERsGDJOnu10xvgfGjM5u/K4nwAQAAe9GjkzxindfuPH3Meu78Tt39kaq6Yya/090+kxkEv2GDNjcV9t9m+38v+/EN9vmLJI/ea8GDxG0XAAAA2F0uzSGCB/t19zsyma5/1mOqapHfhS/J5ALvUnT3uZkEJva7XVVddQun+qmNggcz7V2UZD408HXr7V9VxyT5ibnNf7zZ4MGATyZ58ib3fVEms1zsd0ySr9n2ihY0nfFhPijw+M0ED2b8Qg6+JcKjqupK6+28hvdmcquDzfj9HLglQpKcOp1CdD2PmVt/y0azFszq7ldn8se5w8XLu/svN7nvM+bWb7Pg/zEAAMCM7v7XTAIHD8zkNmgfyuSWhv+V5OOZzHb360numsnt95bpEZnMfveKJO/P5FaHlyb5TCazMzwjyZ26+8Hd/Zkl13ZY8MsQAAAAu8kru3utKdLX8/S59Rsl+eoFjv+rBS8Ob4c3zywflcmnPRbxySx2IfcNc+unbbDvHZJcb27bLy3Q1lY9v7u/sJkdp7fieMvc5o3e07LcNgfX8W9JXrzICaZ/vHrFzKaTM5lydLOeuZngzrStTyV539zmjb6O3zq3/qwF6kqueBF/lRap5T05+NNWJya5/vaWAwAAh7/uPr27a7OPQ5yru/vs7n5Id9+4u0/s7mO7+9rdfefufmJ3/81mf7/ZLt19Xnef2d337u6v6u5rdvcx3X317r51dz+2u9+0zJoON267AAAAwG6y6MXa91bVB3LwfSe/IZNPJGzGuYu0dyhVdcMkX5nJ/e2vmsltDubdZG79Bgs287fTC/Cb9cG59ZM32Pduc+vv6O73LtDWVp234P6LvKdluevc+qu2+IeydyT5npn1Oyb5+00eu5Wv461m1k9ea6equmYO7mNJsuHtFtZwbib3cD1+weO2Wyf5m03vPLl9yYdy8OwaJ2dySw0AAIA9RfgAAACA3eQdWzxm9sLoIlPwbzaksKbpVPv3S/J9Se6dSeBgUScvuP/5C+7/+bn1jWq87dz6Gxdsa6vOX3D/Rd7Tsszfp/SUqnrsFs4zP9PBly9w7PkLtrXZr+P8jAgf7+6PLNJQd19aVe/JBrf9WJLPbeaWJXMOx+83AACApRM+AAAAYDfZyqeJ54+55gLHfmoL7SVJqurmSZ6TySfTR1xlwf0/u+D+87MkXGmDfa81t/5/F2xrq3byPS3Ll82tP2D6GHWNBfbdqa/j1efWL1ywnf0+usXjttOiX6Pk8Px+AwAAWDrhAwAAAHaTzx16lyuYv5g4f6F0Ixdtob1U1dcmeU0WCzqs56gF99/Je17Ov5/P7GBbs5Z6H88dskhIYBEnbHbHHbwf6tXm1udnAtisrfTv7XYkfK8BAACsxKJ/wAAAAIDdppbaWNUxSV6QK16o/7skP5fkW5PcPJMQxJWTHNXdtf+R5KnLrHeQC7Wbd+wOnXep39/r+NLc+lbf6059jQAAAFgCMx8AAACwm1w1ySe2cMysRe/nvqiHJTltZv2SJA/v7j/d5PEnbX9J2+aTc+uLzCKx183fwuNh3f1/VlLJ9vvM3Pr8TAibtdXjAAAAOAyY+QAAAIDd5IbbcMz8BfTt9qC59actEDxIklO2s5ht9vG59ZuupIrdaf5rd+OVVLEzPjq3/hVVtZVZDE479C4AAAAcroQPAAAA2E1uuw3HvGs7CtnArefWF/10++23q5Ad8Pa59TutpIrd6Z1z63dZRRE75H1JLp5ZPybJbRY5QVV9eZLrb2dRAAAALJfwAQAAALvJ/RfZuapukSt+Ov8t21fOmq49t/5vmz2wqm6S5ObbW862et3c+m2mX2MO7dVz63epquuspJJt1t2XJXnr3OaHLHiah26h6Uvn1q+0hXMAAACwTYQPAAAA2E2+taq+YoH9f2Ru/V+TvGcb61nLJXPrJy9w7BOT1PaVsu3enOTDc9t+dhWF7EJvTnLBzPoxSX5+RbXshD+eW/+BqrruZg6sqpOSPGELbX5+bv2qWzgHAAAA20T4AAAAgN3k6CS/U1WHvEBfVbdN8ti5zc/s7st3pLID/n1u/b6bOaiq7pHkMdtfzvbp7kuT/Nrc5odW1fetop7dpLv/K8nT5jb/cFU9cKvn3Ew/WKI/TvKZmfWrJHluVR270UHT9/C/k1xvC21+bG79cJ41BAAA4IgnfAAAAMBuc98kv11V606xXlWnJXlJDp6G/RNJnrHDtSVXvDXBL1XVDTc6oKq+NcnZObxnPdjvWUn+eW7b86rqRzYZCqmqun9VfePOlHdYe1YOnnnjqCQvqKofWyRIUFWnVtXTkjxvuwvcqu7+QpInz23+liSvqKobr3XM9LYTL0zy/dNNX1yw2XclmQ0T3Wwa4gEAAGAFhA8AAADYTd4yff6xJG+sqgdW1QnJf1/UvllVnZnkHUm+fO7Y/9Hdn1xCjc9I0jPrX57k7VX1P2anoa+q46vqnlX1J0lekeSk6XGvX0KNW9bd/5nkO5NcPLP5Skl+N8nfVdUPVNX1Zy+mV9U1quoeVfXLST6QSdDipkss+7DQ3Zck+Y5MgjD7HZfkd5K8ZxpC+OrZYM30+/o60++VJ1fVG5P8W5InJbnGMuvfhGckeeXctrsn+aeqenlVPbWqfryqfqGqXpTkQ5l8LyXJPyT5y7ljOxuYBh7+em7zy6rqz6vqZ6aBmMfOPL5ja28LAACAzTh61QUAAADAAh6W5M2ZXHT9+kwvVlbVZ5KckGS9Kd5/vbtfsIwCu/sfqurpSX50ZvM1k/xWkt+qqouSXJrk5DUOP3P6/E07WuSg7v7H6YXcP09y9ZmXbpvkOdPlS6vqs0lOTHL8kks8bHX3h6YzXbwoyQ1mXrpFJiGEJOnp1y5Jrppd8uGR7u6q+s5MZh2528xLxyT5tuljLf+e5AFJnjq3fTMzITwtyT1y4Gt0bCaBhu9cY9/zkpyziXMCAACwBbvil1cAAABIku7+lyT3SvLRuZdOztrBg8uT/FJ3P3GHS5v3+CTrhR1OyhWDB5cn+YXuPmPnStpe3f3aJHdI8nfr7HJ0JqGL9YIHl6+z/YjX3e9IcrtMZoBYS2XyPXJy1v/bzaU5+BYOh4XuvjjJfZL8SpL/2sQhr01yh+4+P5OgxazPbKK9c5M8Nsl/LlQoAAAA2074AAAAgF2lu9+e5JaZTPN/8Tq7XZ7JdOx36O6fW1Zt+3X3pd390CQPT/LPG+x6eZKXJbljd//iUorbRt39/u6+fZIHJXldkksOccink/xpknsnef7OVnd46+5PdPcDk9wmyf/JwbdiWM/FSV6e5HFJTu3uJ+1giVvW3f85re20JD+f5E1JPpJJGOGiJO9N8uwk39zd9+juC6aHXnvuVJ/eZHvPyuQ2Hj+dSX86P8nns4cDLgAAAKtQ3RvePg8AAAAOW1V15SR3yuQi58mZXHC8IMnfdvfHVljaQarq5pncJuLamczQ8LkkH0jy5u7e1AXW3aCqTkxyxyTXS3KtTN7rRUkuTPK+JO/r7stWV+Hhq6oqk1DNaUlOyeR2Fpdm8j39kUxCLB/o7ktXVuQOqqqjk3w2k9unJEknuXp3f3b9owAAADicCB8AAAAAsFJVde9MZi3Y75+6++arqgcAAIDFue0CAAAAACtTVUcl+Zm5za9ZRS0AAABsnfABAAAAANuiqo7dwmH/M8k3zm37vW0oBwAAgCUSPgAAAABgu/xwVb26qh5QVVfeaMeq+sqq+oskPz330tnd/Y87VyIAAAA74ehVFwAAAADAEaOS3HP6+EJVvTnJu5J8NMlFSU5Kct0kd0xyu1zxgzEfTfJDS6sWAACAbSN8AAAAAMBOOCHJN08fm3F+km/v7o/vWEUAAADsGLddAAAAAGC7fDiTGQ4W8aUkz0jy9d39nu0vCQAAgGWo7l51DQAAAAAcIarqyknukeSbknxtkhsluXaSE6e7fDrJJ5P8Q5K/SfLi7v7I0gsFAABgWwkfAAAAAAAAAABD3HYBAAAAAAAAABgifAAAAAAAAAAADBE+AAAAAAAAAACGCB8AAAAAAAAAAEOEDwAAAAAAAACAIcIHAAAAAAAAAMAQ4QMAAAAAAAAAYIjwAQAAAAAAAAAwRPgAAAAAAAAAABgifAAAAAAAAAAADBE+AAAAAAAAAACGCB8AAAAAAAAAAEOEDwAAAAAAAACAIcIHAAAAAAAAAMAQ4QMAAAAAAAAAYIjwAQAAAAAAAAAwRPgAAAAAAAAAABgifAAAAAAAAAAADBE+AAAAAAAAAACGCB8AAAAAAAAAAEOEDwAAAAAAAACAIcIHAAAAAAAAAMAQ4QMAAAAAAAAAYIjwAQAAAAAAAAAwRPgAAAAAAAAAABgifAAAAAAAAAAADBE+AAAAAAAAAACGCB8AAAAAAAAAAEOEDwAAAAAAAACAIcIHAAAAAAAAAMAQ4QMAAAAAAAAAYIjwAQAAAAAAAAAwRPgAAAAAAAAAABgifAAAAAAAAAAADBE+AAAAAAAAAACGCB8AAAAAAAAAAEOEDwAAAAAAAACAIcIHAAAAAAAAAMAQ4QMAAAAAAAAAYMj/AwXSAIMmrvzcAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 2400x1600 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax1 = plt.subplots(1, dpi=400)\n",
    "\n",
    "Phylo.draw(tree,\n",
    "           axes=ax1, \n",
    "           show_confidence=False, \n",
    "           label_func = lambda x:'', \n",
    "           do_show=False,\n",
    "           )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 3. Convert tree to dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialize the tree dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ufboot</th>\n",
       "      <th>scf</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>node_type</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.447080e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.307300e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>91</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.721210e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>57</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.630000e-08</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN02442718</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>testlocal1</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE5</th>\n",
       "      <td>94</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.052000e-07</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN02442721</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>testlocal2</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    ufboot scf  branch_length node_type  \\\n",
       "name                                                                      \n",
       "NODE0                                   NA  NA   0.000000e+00        NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic     NA  NA   2.447080e-05        NA   \n",
       "Reference                               NA  NA   2.307300e-05        NA   \n",
       "NODE1                                   91  NA   2.721210e-05        NA   \n",
       "NODE2                                   57  NA   5.630000e-08        NA   \n",
       "...                                    ...  ..            ...       ...   \n",
       "SAMN02442718                            NA  NA   2.330000e-08        NA   \n",
       "testlocal1                              NA  NA   2.330000e-08        NA   \n",
       "NODE5                                   94  NA   7.052000e-07        NA   \n",
       "SAMN02442721                            NA  NA   2.330000e-08        NA   \n",
       "testlocal2                              NA  NA   2.330000e-08        NA   \n",
       "\n",
       "                                    coord_x coord_y  \n",
       "name                                                 \n",
       "NODE0                                    NA      NA  \n",
       "GCA_009909635.1_ASM990963v1_genomic      NA      NA  \n",
       "Reference                                NA      NA  \n",
       "NODE1                                    NA      NA  \n",
       "NODE2                                    NA      NA  \n",
       "...                                     ...     ...  \n",
       "SAMN02442718                             NA      NA  \n",
       "testlocal1                               NA      NA  \n",
       "NODE5                                    NA      NA  \n",
       "SAMN02442721                             NA      NA  \n",
       "testlocal2                               NA      NA  \n",
       "\n",
       "[14 rows x 6 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# initialize empty dataframe\n",
    "tree_df = pd.DataFrame(columns = [\n",
    "    \"name\",\n",
    "    \"ufboot\", \n",
    "    \"scf\", \n",
    "    \"branch_length\", \n",
    "    \"node_type\",\n",
    "    \"coord_x\", \n",
    "    \"coord_y\",\n",
    "]\n",
    ")\n",
    "\n",
    "for c in tree.find_clades():\n",
    "    if not c.branch_length: c.branch_length = 0.0\n",
    "    \n",
    "    node_data = {\n",
    "        \"name\" : c.name,\n",
    "        \"ufboot\" : [conf.value for conf in c.confidences if conf.type==\"ufboot\"][0],\n",
    "        \"scf\" :  [conf.value for conf in c.confidences if conf.type==\"scf\"][0],\n",
    "        \"node_type\" :  NO_DATA_CHAR,\n",
    "        \"branch_length\" :  c.branch_length,\n",
    "        \"coord_x\" :  NO_DATA_CHAR,\n",
    "        \"coord_y\" :  NO_DATA_CHAR,\n",
    "    }\n",
    "    \n",
    "    tree_df = tree_df.append(node_data, ignore_index=True)\n",
    "\n",
    "# Set the index to the node name\n",
    "tree_df.set_index(\"name\", inplace=True)\n",
    "\n",
    "# Visualize data frame\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add plotting x and y coordinates of each node"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ufboot</th>\n",
       "      <th>scf</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>node_type</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.447080e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.44708e-05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.307300e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.3073e-05</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>91</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.721210e-05</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.72121e-05</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>57</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.630000e-08</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.72684e-05</td>\n",
       "      <td>3.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN02442718</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.18884e-05</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>testlocal1</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.18884e-05</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE5</th>\n",
       "      <td>94</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.052000e-07</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.20575e-05</td>\n",
       "      <td>7.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN02442721</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.20808e-05</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>testlocal2</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.20808e-05</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    ufboot scf  branch_length node_type  \\\n",
       "name                                                                      \n",
       "NODE0                                   NA  NA   0.000000e+00        NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic     NA  NA   2.447080e-05        NA   \n",
       "Reference                               NA  NA   2.307300e-05        NA   \n",
       "NODE1                                   91  NA   2.721210e-05        NA   \n",
       "NODE2                                   57  NA   5.630000e-08        NA   \n",
       "...                                    ...  ..            ...       ...   \n",
       "SAMN02442718                            NA  NA   2.330000e-08        NA   \n",
       "testlocal1                              NA  NA   2.330000e-08        NA   \n",
       "NODE5                                   94  NA   7.052000e-07        NA   \n",
       "SAMN02442721                            NA  NA   2.330000e-08        NA   \n",
       "testlocal2                              NA  NA   2.330000e-08        NA   \n",
       "\n",
       "                                         coord_x coord_y  \n",
       "name                                                      \n",
       "NODE0                                          0       3  \n",
       "GCA_009909635.1_ASM990963v1_genomic  2.44708e-05       1  \n",
       "Reference                             2.3073e-05       2  \n",
       "NODE1                                2.72121e-05       5  \n",
       "NODE2                                2.72684e-05     3.5  \n",
       "...                                          ...     ...  \n",
       "SAMN02442718                         3.18884e-05       5  \n",
       "testlocal1                           3.18884e-05       6  \n",
       "NODE5                                3.20575e-05     7.5  \n",
       "SAMN02442721                         3.20808e-05       7  \n",
       "testlocal2                           3.20808e-05       8  \n",
       "\n",
       "[14 rows x 6 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_posns = get_x_positions(tree)\n",
    "y_posns = get_y_positions(tree)\n",
    "\n",
    "# Add x and y coordinates as other attributes\n",
    "for c in tree.find_clades():\n",
    "    # x coordinates will be of branch length units\n",
    "    coord_x = [value for key,value in x_posns.items() if key.name == c.name][0]\n",
    "    # y coordinates will be arbitrary, based on number of tips\n",
    "    coord_y = [value for key,value in y_posns.items() if key.name == c.name][0]\n",
    "    # Add data to tree dataframe\n",
    "    tree_df.at[c.name, 'coord_x'] = coord_x\n",
    "    tree_df.at[c.name, 'coord_y'] = coord_y    \n",
    "\n",
    "# Visualize dataframe\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add node type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ufboot</th>\n",
       "      <th>scf</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>node_type</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>internal</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.447080e-05</td>\n",
       "      <td>terminal</td>\n",
       "      <td>2.44708e-05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.307300e-05</td>\n",
       "      <td>terminal</td>\n",
       "      <td>2.3073e-05</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>91</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.721210e-05</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.72121e-05</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>57</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.630000e-08</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.72684e-05</td>\n",
       "      <td>3.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN02442718</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>terminal</td>\n",
       "      <td>3.18884e-05</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>testlocal1</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>terminal</td>\n",
       "      <td>3.18884e-05</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE5</th>\n",
       "      <td>94</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.052000e-07</td>\n",
       "      <td>internal</td>\n",
       "      <td>3.20575e-05</td>\n",
       "      <td>7.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN02442721</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>terminal</td>\n",
       "      <td>3.20808e-05</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>testlocal2</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>terminal</td>\n",
       "      <td>3.20808e-05</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    ufboot scf  branch_length node_type  \\\n",
       "name                                                                      \n",
       "NODE0                                   NA  NA   0.000000e+00  internal   \n",
       "GCA_009909635.1_ASM990963v1_genomic     NA  NA   2.447080e-05  terminal   \n",
       "Reference                               NA  NA   2.307300e-05  terminal   \n",
       "NODE1                                   91  NA   2.721210e-05  internal   \n",
       "NODE2                                   57  NA   5.630000e-08  internal   \n",
       "...                                    ...  ..            ...       ...   \n",
       "SAMN02442718                            NA  NA   2.330000e-08  terminal   \n",
       "testlocal1                              NA  NA   2.330000e-08  terminal   \n",
       "NODE5                                   94  NA   7.052000e-07  internal   \n",
       "SAMN02442721                            NA  NA   2.330000e-08  terminal   \n",
       "testlocal2                              NA  NA   2.330000e-08  terminal   \n",
       "\n",
       "                                         coord_x coord_y  \n",
       "name                                                      \n",
       "NODE0                                          0       3  \n",
       "GCA_009909635.1_ASM990963v1_genomic  2.44708e-05       1  \n",
       "Reference                             2.3073e-05       2  \n",
       "NODE1                                2.72121e-05       5  \n",
       "NODE2                                2.72684e-05     3.5  \n",
       "...                                          ...     ...  \n",
       "SAMN02442718                         3.18884e-05       5  \n",
       "testlocal1                           3.18884e-05       6  \n",
       "NODE5                                3.20575e-05     7.5  \n",
       "SAMN02442721                         3.20808e-05       7  \n",
       "testlocal2                           3.20808e-05       8  \n",
       "\n",
       "[14 rows x 6 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for c in tree.find_clades():   \n",
    "    # Default Color\n",
    "    node_type = \"internal\"\n",
    "    # Terminal branches will be grey\n",
    "    if c.is_terminal():\n",
    "        node_type = \"terminal\"\n",
    "    tree_df.at[c.name, \"node_type\"] = node_type          \n",
    "\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 5. Combine metadata into tree dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse metadata into df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_009669545.1_ASM966954v1_genomic</th>\n",
       "      <td>42126</td>\n",
       "      <td>2006</td>\n",
       "      <td>-15</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722925</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009669555.1_ASM966955v1_genomic</th>\n",
       "      <td>42123</td>\n",
       "      <td>2005</td>\n",
       "      <td>-16</td>\n",
       "      <td>China</td>\n",
       "      <td>Xinjiang</td>\n",
       "      <td>35.0001</td>\n",
       "      <td>105</td>\n",
       "      <td>42.4805</td>\n",
       "      <td>85.4633</td>\n",
       "      <td>Antiqua</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT1</td>\n",
       "      <td>SAMN07722924</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>0</td>\n",
       "      <td>Asia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>9_10</td>\n",
       "      <td>1923</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN02442721</th>\n",
       "      <td>Justinian</td>\n",
       "      <td>[500:800]</td>\n",
       "      <td>[-1521:-1221]</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>51.0834</td>\n",
       "      <td>10.4234</td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT4</td>\n",
       "      <td>SAMN02442721</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>0</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN02442718</th>\n",
       "      <td>Justinian</td>\n",
       "      <td>[500:800]</td>\n",
       "      <td>[-1521:-1221]</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>51.0834</td>\n",
       "      <td>10.4234</td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT4</td>\n",
       "      <td>SAMN02442718</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>0</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>testlocal1</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>testlocal1</td>\n",
       "      <td>KEEP: Local Test</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>testlocal2</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>testlocal2</td>\n",
       "      <td>KEEP: Local Test</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        strain       date        date_bp  \\\n",
       "sample                                                                     \n",
       "GCA_009669545.1_ASM966954v1_genomic      42126       2006            -15   \n",
       "GCA_009669555.1_ASM966955v1_genomic      42123       2005            -16   \n",
       "GCA_009909635.1_ASM990963v1_genomic       9_10       1923            -98   \n",
       "SAMN02442721                         Justinian  [500:800]  [-1521:-1221]   \n",
       "SAMN02442718                         Justinian  [500:800]  [-1521:-1221]   \n",
       "testlocal1                                  NA         NA             NA   \n",
       "testlocal2                                  NA         NA             NA   \n",
       "\n",
       "                                     country       province country_lat  \\\n",
       "sample                                                                    \n",
       "GCA_009669545.1_ASM966954v1_genomic    China       Xinjiang     35.0001   \n",
       "GCA_009669555.1_ASM966955v1_genomic    China       Xinjiang     35.0001   \n",
       "GCA_009909635.1_ASM990963v1_genomic   Russia  Rostov Oblast     64.6863   \n",
       "SAMN02442721                         Germany        Bavaria     51.0834   \n",
       "SAMN02442718                         Germany        Bavaria     51.0834   \n",
       "testlocal1                                NA             NA          NA   \n",
       "testlocal2                                NA             NA          NA   \n",
       "\n",
       "                                    country_lon province_lat province_lon  \\\n",
       "sample                                                                      \n",
       "GCA_009669545.1_ASM966954v1_genomic         105      42.4805      85.4633   \n",
       "GCA_009669555.1_ASM966955v1_genomic         105      42.4805      85.4633   \n",
       "GCA_009909635.1_ASM990963v1_genomic     97.7453      47.6222      40.7958   \n",
       "SAMN02442721                            10.4234      48.9468      11.4039   \n",
       "SAMN02442718                            10.4234      48.9468      11.4039   \n",
       "testlocal1                                   NA           NA           NA   \n",
       "testlocal2                                   NA           NA           NA   \n",
       "\n",
       "                                         biovar branch_major branch_minor  \\\n",
       "sample                                                                      \n",
       "GCA_009669545.1_ASM966954v1_genomic     Antiqua        0.ANT       0.ANT1   \n",
       "GCA_009669555.1_ASM966955v1_genomic     Antiqua        0.ANT       0.ANT1   \n",
       "GCA_009909635.1_ASM990963v1_genomic  Medievalis        2.MED       2.MED1   \n",
       "SAMN02442721                         Orientalis        0.ANT       0.ANT4   \n",
       "SAMN02442718                         Orientalis        0.ANT       0.ANT4   \n",
       "testlocal1                                   NA           NA           NA   \n",
       "testlocal2                                   NA           NA           NA   \n",
       "\n",
       "                                    biosample_accession  \\\n",
       "sample                                                    \n",
       "GCA_009669545.1_ASM966954v1_genomic        SAMN07722925   \n",
       "GCA_009669555.1_ASM966955v1_genomic        SAMN07722924   \n",
       "GCA_009909635.1_ASM990963v1_genomic        SAMN13632815   \n",
       "SAMN02442721                               SAMN02442721   \n",
       "SAMN02442718                               SAMN02442718   \n",
       "testlocal1                                   testlocal1   \n",
       "testlocal2                                   testlocal2   \n",
       "\n",
       "                                         biosample_comment branch_number  \\\n",
       "sample                                                                     \n",
       "GCA_009669545.1_ASM966954v1_genomic  KEEP: Assembly Modern             0   \n",
       "GCA_009669555.1_ASM966955v1_genomic  KEEP: Assembly Modern             0   \n",
       "GCA_009909635.1_ASM990963v1_genomic  KEEP: Assembly Modern             2   \n",
       "SAMN02442721                             KEEP: SRA Ancient             0   \n",
       "SAMN02442718                             KEEP: SRA Ancient             0   \n",
       "testlocal1                                KEEP: Local Test            NA   \n",
       "testlocal2                                KEEP: Local Test            NA   \n",
       "\n",
       "                                    continent  \n",
       "sample                                         \n",
       "GCA_009669545.1_ASM966954v1_genomic      Asia  \n",
       "GCA_009669555.1_ASM966955v1_genomic      Asia  \n",
       "GCA_009909635.1_ASM990963v1_genomic    Europe  \n",
       "SAMN02442721                           Europe  \n",
       "SAMN02442718                           Europe  \n",
       "testlocal1                                 NA  \n",
       "testlocal2                                 NA  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Parse the metadata\n",
    "metadata_df = pd.read_csv(metadata_path, sep='\\t')\n",
    "\n",
    "# Fix the problem with multiple forms of NA in the table\n",
    "# Consolidate missing data to the NO_DATA_CHAR\n",
    "metadata_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "\n",
    "# set the metadata index\n",
    "metadata_df.set_index(NAME_COL, inplace=True)\n",
    "\n",
    "metadata_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add metadata to tree and tree dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ufboot</th>\n",
       "      <th>scf</th>\n",
       "      <th>branch_length</th>\n",
       "      <th>node_type</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "      <th>date_bp</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>country_lat</th>\n",
       "      <th>country_lon</th>\n",
       "      <th>province_lat</th>\n",
       "      <th>province_lon</th>\n",
       "      <th>biovar</th>\n",
       "      <th>branch_major</th>\n",
       "      <th>branch_minor</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>biosample_comment</th>\n",
       "      <th>branch_number</th>\n",
       "      <th>continent</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>internal</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_009909635.1_ASM990963v1_genomic</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.447080e-05</td>\n",
       "      <td>terminal</td>\n",
       "      <td>2.44708e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>9_10</td>\n",
       "      <td>1923</td>\n",
       "      <td>-98</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Rostov Oblast</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>47.6222</td>\n",
       "      <td>40.7958</td>\n",
       "      <td>Medievalis</td>\n",
       "      <td>2.MED</td>\n",
       "      <td>2.MED1</td>\n",
       "      <td>SAMN13632815</td>\n",
       "      <td>KEEP: Assembly Modern</td>\n",
       "      <td>2</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.307300e-05</td>\n",
       "      <td>terminal</td>\n",
       "      <td>2.3073e-05</td>\n",
       "      <td>2</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>91</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.721210e-05</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.72121e-05</td>\n",
       "      <td>5</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>57</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.630000e-08</td>\n",
       "      <td>internal</td>\n",
       "      <td>2.72684e-05</td>\n",
       "      <td>3.5</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN02442718</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>terminal</td>\n",
       "      <td>3.18884e-05</td>\n",
       "      <td>5</td>\n",
       "      <td>Justinian</td>\n",
       "      <td>[500:800]</td>\n",
       "      <td>[-1521:-1221]</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>51.0834</td>\n",
       "      <td>10.4234</td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT4</td>\n",
       "      <td>SAMN02442718</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>0</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>testlocal1</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>terminal</td>\n",
       "      <td>3.18884e-05</td>\n",
       "      <td>6</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>testlocal1</td>\n",
       "      <td>KEEP: Local Test</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE5</th>\n",
       "      <td>94</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.052000e-07</td>\n",
       "      <td>internal</td>\n",
       "      <td>3.20575e-05</td>\n",
       "      <td>7.5</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN02442721</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>terminal</td>\n",
       "      <td>3.20808e-05</td>\n",
       "      <td>7</td>\n",
       "      <td>Justinian</td>\n",
       "      <td>[500:800]</td>\n",
       "      <td>[-1521:-1221]</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Bavaria</td>\n",
       "      <td>51.0834</td>\n",
       "      <td>10.4234</td>\n",
       "      <td>48.9468</td>\n",
       "      <td>11.4039</td>\n",
       "      <td>Orientalis</td>\n",
       "      <td>0.ANT</td>\n",
       "      <td>0.ANT4</td>\n",
       "      <td>SAMN02442721</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>0</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>testlocal2</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.330000e-08</td>\n",
       "      <td>terminal</td>\n",
       "      <td>3.20808e-05</td>\n",
       "      <td>8</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>testlocal2</td>\n",
       "      <td>KEEP: Local Test</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    ufboot scf  branch_length node_type  \\\n",
       "name                                                                      \n",
       "NODE0                                   NA  NA   0.000000e+00  internal   \n",
       "GCA_009909635.1_ASM990963v1_genomic     NA  NA   2.447080e-05  terminal   \n",
       "Reference                               NA  NA   2.307300e-05  terminal   \n",
       "NODE1                                   91  NA   2.721210e-05  internal   \n",
       "NODE2                                   57  NA   5.630000e-08  internal   \n",
       "...                                    ...  ..            ...       ...   \n",
       "SAMN02442718                            NA  NA   2.330000e-08  terminal   \n",
       "testlocal1                              NA  NA   2.330000e-08  terminal   \n",
       "NODE5                                   94  NA   7.052000e-07  internal   \n",
       "SAMN02442721                            NA  NA   2.330000e-08  terminal   \n",
       "testlocal2                              NA  NA   2.330000e-08  terminal   \n",
       "\n",
       "                                         coord_x coord_y     strain  \\\n",
       "name                                                                  \n",
       "NODE0                                          0       3         NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic  2.44708e-05       1       9_10   \n",
       "Reference                             2.3073e-05       2         NA   \n",
       "NODE1                                2.72121e-05       5         NA   \n",
       "NODE2                                2.72684e-05     3.5         NA   \n",
       "...                                          ...     ...        ...   \n",
       "SAMN02442718                         3.18884e-05       5  Justinian   \n",
       "testlocal1                           3.18884e-05       6         NA   \n",
       "NODE5                                3.20575e-05     7.5         NA   \n",
       "SAMN02442721                         3.20808e-05       7  Justinian   \n",
       "testlocal2                           3.20808e-05       8         NA   \n",
       "\n",
       "                                          date        date_bp  country  \\\n",
       "name                                                                     \n",
       "NODE0                                       NA             NA       NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic       1923            -98   Russia   \n",
       "Reference                                   NA             NA       NA   \n",
       "NODE1                                       NA             NA       NA   \n",
       "NODE2                                       NA             NA       NA   \n",
       "...                                        ...            ...      ...   \n",
       "SAMN02442718                         [500:800]  [-1521:-1221]  Germany   \n",
       "testlocal1                                  NA             NA       NA   \n",
       "NODE5                                       NA             NA       NA   \n",
       "SAMN02442721                         [500:800]  [-1521:-1221]  Germany   \n",
       "testlocal2                                  NA             NA       NA   \n",
       "\n",
       "                                          province country_lat country_lon  \\\n",
       "name                                                                         \n",
       "NODE0                                           NA          NA          NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic  Rostov Oblast     64.6863     97.7453   \n",
       "Reference                                       NA          NA          NA   \n",
       "NODE1                                           NA          NA          NA   \n",
       "NODE2                                           NA          NA          NA   \n",
       "...                                            ...         ...         ...   \n",
       "SAMN02442718                               Bavaria     51.0834     10.4234   \n",
       "testlocal1                                      NA          NA          NA   \n",
       "NODE5                                           NA          NA          NA   \n",
       "SAMN02442721                               Bavaria     51.0834     10.4234   \n",
       "testlocal2                                      NA          NA          NA   \n",
       "\n",
       "                                    province_lat province_lon      biovar  \\\n",
       "name                                                                        \n",
       "NODE0                                         NA           NA          NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic      47.6222      40.7958  Medievalis   \n",
       "Reference                                     NA           NA          NA   \n",
       "NODE1                                         NA           NA          NA   \n",
       "NODE2                                         NA           NA          NA   \n",
       "...                                          ...          ...         ...   \n",
       "SAMN02442718                             48.9468      11.4039  Orientalis   \n",
       "testlocal1                                    NA           NA          NA   \n",
       "NODE5                                         NA           NA          NA   \n",
       "SAMN02442721                             48.9468      11.4039  Orientalis   \n",
       "testlocal2                                    NA           NA          NA   \n",
       "\n",
       "                                    branch_major branch_minor  \\\n",
       "name                                                            \n",
       "NODE0                                         NA           NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic        2.MED       2.MED1   \n",
       "Reference                                     NA           NA   \n",
       "NODE1                                         NA           NA   \n",
       "NODE2                                         NA           NA   \n",
       "...                                          ...          ...   \n",
       "SAMN02442718                               0.ANT       0.ANT4   \n",
       "testlocal1                                    NA           NA   \n",
       "NODE5                                         NA           NA   \n",
       "SAMN02442721                               0.ANT       0.ANT4   \n",
       "testlocal2                                    NA           NA   \n",
       "\n",
       "                                    biosample_accession  \\\n",
       "name                                                      \n",
       "NODE0                                                NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic        SAMN13632815   \n",
       "Reference                                            NA   \n",
       "NODE1                                                NA   \n",
       "NODE2                                                NA   \n",
       "...                                                 ...   \n",
       "SAMN02442718                               SAMN02442718   \n",
       "testlocal1                                   testlocal1   \n",
       "NODE5                                                NA   \n",
       "SAMN02442721                               SAMN02442721   \n",
       "testlocal2                                   testlocal2   \n",
       "\n",
       "                                         biosample_comment branch_number  \\\n",
       "name                                                                       \n",
       "NODE0                                                   NA            NA   \n",
       "GCA_009909635.1_ASM990963v1_genomic  KEEP: Assembly Modern             2   \n",
       "Reference                                               NA            NA   \n",
       "NODE1                                                   NA            NA   \n",
       "NODE2                                                   NA            NA   \n",
       "...                                                    ...           ...   \n",
       "SAMN02442718                             KEEP: SRA Ancient             0   \n",
       "testlocal1                                KEEP: Local Test            NA   \n",
       "NODE5                                                   NA            NA   \n",
       "SAMN02442721                             KEEP: SRA Ancient             0   \n",
       "testlocal2                                KEEP: Local Test            NA   \n",
       "\n",
       "                                    continent  \n",
       "name                                           \n",
       "NODE0                                      NA  \n",
       "GCA_009909635.1_ASM990963v1_genomic    Europe  \n",
       "Reference                                  NA  \n",
       "NODE1                                      NA  \n",
       "NODE2                                      NA  \n",
       "...                                       ...  \n",
       "SAMN02442718                           Europe  \n",
       "testlocal1                                 NA  \n",
       "NODE5                                      NA  \n",
       "SAMN02442721                           Europe  \n",
       "testlocal2                                 NA  \n",
       "\n",
       "[14 rows x 22 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Iterate through the different metadata attributes\n",
    "for attr in metadata_df.columns:\n",
    "    # Initialize an empty column for the attribute\n",
    "    tree_df[attr.lower()] = [NO_DATA_CHAR for row in range(0,len(tree_df))]\n",
    "    # Iterate over the nodes in the tree\n",
    "    for c in tree.find_clades():\n",
    "        # Initialize to empty\n",
    "        attr_val = NO_DATA_CHAR\n",
    "        # If the node shows up in the metadata\n",
    "        if c.name in metadata_df.index:\n",
    "            tree_df.at[c.name, attr.lower()] = metadata_df[attr][c.name]\n",
    "            attr_val = metadata_df[attr][c.name]         \n",
    "\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add Reference Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in tree_df.columns:\n",
    "    if col in REF_META:\n",
    "        tree_df.at[\"Reference\", col] = REF_META[col]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make NA values consistent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fix the problem with multiple forms of NA in the table\n",
    "# Consolidate missing data to the NO_DATA_CHAR\n",
    "tree_df.fillna(NO_DATA_CHAR, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ?. Add Metadata As Node Comments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in tree.find_clades():\n",
    "    c.comment = \"\"\n",
    "    for col in tree_df.columns:\n",
    "        col_val = tree_df[col][c.name]   \n",
    "        # Test for non ascii\n",
    "        if type(col_val) == str and not col_val.isascii():\n",
    "            col_val = col_val.encode(\"unicode_escape\")     \n",
    "        # Add comment\n",
    "        if not hasattr(c, \"comment\") or not c.comment:\n",
    "            c.comment = \"&{}={}\".format(col, col_val)\n",
    "        else:\n",
    "            c.comment += \",{}={}\".format(col, col_val)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 6. Export"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Objects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save tree and tree dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Dataframe\n",
    "out_path_df = os.path.join(outdir, SCRIPT_NAME + \".tsv\" )\n",
    "out_path_pickle_df = os.path.join(outdir,  SCRIPT_NAME + \".df.obj\" )\n",
    "\n",
    "tree_df.to_csv(out_path_df, sep=\"\\t\")\n",
    "with open(out_path_pickle_df,\"wb\") as outfile:\n",
    "    dill.dump(tree_df, outfile)\n",
    "\n",
    "# Nexus\n",
    "out_path_nexus = os.path.join(outdir, SCRIPT_NAME + \".nexus\" )\n",
    "Phylo.write(tree, out_path_nexus, 'nexus', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))\n",
    "\n",
    "# Dill object\n",
    "out_path_dill_tree = os.path.join(outdir,  SCRIPT_NAME + \".phylo.obj\" )\n",
    "with open(out_path_dill_tree,\"wb\") as outfile:\n",
    "    dill.dump(tree, outfile)\n",
    "    \n",
    "# Newick (remove comments)\n",
    "for c in tree.find_clades(): c.comment = None\n",
    "out_path_nwk = os.path.join(outdir, SCRIPT_NAME + \".nwk\" )\n",
    "Phylo.write(tree, out_path_nwk, 'newick', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save lat and lon\n",
    "\n",
    "Write an output tsv for auspice that is:  \n",
    "resolution   name   lat    lon\n",
    "\n",
    "Write an output tsv for beast that is: \n",
    "sample lat\n",
    "\n",
    "Write an output tsv for beast that is: \n",
    "sample lon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "latlon_dict_auspice = {\"country\": {}, \"province\": {}}\n",
    "latlon_dict_beast = {}\n",
    "\n",
    "out_path_lat_lon_auspice = os.path.join(outdir, SCRIPT_NAME + \"_lat-lon_auspice.tsv\")\n",
    "out_path_lat_beast = os.path.join(outdir, SCRIPT_NAME + \"_lat_beast.tsv\")\n",
    "out_path_lon_beast = os.path.join(outdir, SCRIPT_NAME + \"_lon_beast.tsv\")\n",
    "\n",
    "for c in tree.find_clades():\n",
    "    # For auspice\n",
    "    # Iterate through each geographic level\n",
    "    for level in latlon_dict_auspice:\n",
    "        # Store the location name\n",
    "        loc = tree_df[level][c.name]\n",
    "        # If the node has location data and is new\n",
    "        if loc != NO_DATA_CHAR and loc not in latlon_dict_auspice[level]:\n",
    "            latlon_dict_auspice[level][loc] = {}\n",
    "            latlon_dict_auspice[level][loc][\"lat\"] = tree_df[level + \"_lat\"][c.name]\n",
    "            latlon_dict_auspice[level][loc][\"lon\"] = tree_df[level + \"_lon\"][c.name]\n",
    "    \n",
    "    # For beast\n",
    "    # skip internal nodes\n",
    "    if not c.is_terminal(): continue\n",
    "    # Get latitude and longitude\n",
    "    lat = tree_df[\"province_lat\"][c.name]\n",
    "    if lat == NO_DATA_CHAR:\n",
    "        lat = tree_df[\"country_lat\"][c.name]\n",
    "    lon = tree_df[\"province_lon\"][c.name]\n",
    "    if lon == NO_DATA_CHAR:\n",
    "        lon = tree_df[\"country_lon\"][c.name] \n",
    "    \n",
    "    latlon_dict_beast[c.name] = {\"lat\": lat, \"lon\": lon}    \n",
    "\n",
    "with open(out_path_lat_lon_auspice, \"w\") as outfile:\n",
    "    for level in latlon_dict_auspice:\n",
    "        for loc in latlon_dict_auspice[level]:\n",
    "            outfile.write(level.lower() + \"\\t\"\n",
    "                         + loc + \"\\t\"\n",
    "                         + str(latlon_dict_auspice[level][loc][\"lat\"]) + \"\\t\"\n",
    "                         + str(latlon_dict_auspice[level][loc][\"lon\"]) + \"\\n\"\n",
    "                 )\n",
    "            \n",
    "with open(out_path_lat_beast, \"w\") as latfile:\n",
    "    with open(out_path_lon_beast, \"w\") as lonfile:          \n",
    "        for sample in latlon_dict_beast:\n",
    "            latfile.write(sample + \"\\t\" + str(latlon_dict_beast[sample][\"lat\"]) + \"\\n\")\n",
    "            lonfile.write(sample + \"\\t\" + str(latlon_dict_beast[sample][\"lon\"]) + \"\\n\")\n",
    "\n",
    "# For clarity with auspice        \n",
    "auspice_latlons_path = out_path_lat_lon_auspice        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save Dates\n",
    "\n",
    "Write an output tsv for beast that is:  \n",
    "sample    date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "dates_dict_beast = {}\n",
    "out_path_dates_beast = os.path.join(outdir, SCRIPT_NAME + \"_dates_beast.tsv\")\n",
    "\n",
    "for c in tree.find_clades():\n",
    "    if not c.is_terminal(): continue\n",
    "    date_bp = str(tree_df[\"date_bp\"][c.name])\n",
    "    if date_bp == NO_DATA_CHAR: continue\n",
    "    date_bp = date_bp.lstrip(\"[\").rstrip(\"]\")\n",
    "    date_bp_split = [float(d) for d in date_bp.split(\":\")]\n",
    "    # Calculate the mean\n",
    "    date_bp_mean = sum(date_bp_split) / len(date_bp_split)\n",
    "    # Convert to positive\n",
    "    date_bp_pos = 0 - date_bp_mean\n",
    "    dates_dict_beast[c.name] = date_bp_pos\n",
    "\n",
    "with open(out_path_dates_beast, \"w\") as datefile:    \n",
    "    for sample in dates_dict_beast:\n",
    "        datefile.write(sample + \"\\t\" + str(dates_dict_beast[sample]) + \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save colors\n",
    "\n",
    "Write an output tsv that is:\n",
    "attribute   attribute_val   hex_color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path_colors = os.path.join(outdir, SCRIPT_NAME + \"_colors.tsv\")\n",
    "file_colors = open(out_path_colors, \"w\")\n",
    "\n",
    "hex_dict = {}\n",
    "\n",
    "for attr in ATTRIBUTE_LIST:\n",
    "    # Create the color map\n",
    "    attr_key = attr.lower()\n",
    "    hex_dict[attr_key] = {}\n",
    "    for t in tree.get_terminals():\n",
    "        attr_val = tree_df[attr][t.name]\n",
    "        if attr_val not in hex_dict[attr_key] and attr_val != NO_DATA_CHAR:\n",
    "            hex_dict[attr_key][attr_val] = {}\n",
    "            \n",
    "    # Create the custom color map (pyplot)\n",
    "    cmap = plt.get_cmap(CONT_COLOR_PAL, len(hex_dict[attr_key]))\n",
    "    # Convert the color map to a list of RGB values\n",
    "    cmaplist = [cmap(i) for i in range(cmap.N)]\n",
    "    # Convert RGB values to hex colors\n",
    "    attr_hex = [colors.to_hex(col) for col in cmaplist]\n",
    "    \n",
    "    # Assign colors to value\n",
    "    for attr_val, attr_col in zip(hex_dict[attr_key], attr_hex):\n",
    "        hex_dict[attr_key][attr_val] = attr_col   \n",
    "\n",
    "for attr_key in hex_dict:\n",
    "    for attr_val in hex_dict[attr_key]:\n",
    "        file_colors.write(str(attr_key) + \"\\t\" + str(attr_val) + \"\\t\" + str(hex_dict[attr_key][attr_val]) + \"\\n\")\n",
    "\n",
    "file_colors.close()        \n",
    "\n",
    "# For clarity with auspice\n",
    "auspice_colors_path = out_path_colors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Augur JSON\n",
    "\n",
    "  - alignment (empty)\n",
    "  - input_tree (tree_path)\n",
    "  - nodes (node_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ufboot': 'NA', 'scf': 'NA', 'branch_length': 2.3073e-05, 'node_type': 'terminal', 'strain': 'CO92', 'date': 1992.0, 'date_bp': -29.0, 'country': 'United States of America', 'province': 'Colorado', 'biovar': 'Orientalis', 'branch_major': '1.ORI', 'branch_minor': '1.ORI1', 'biosample_accession': 'NA', 'biosample_comment': 'KEEP: Assembly Modern Reference', 'branch_number': '1', 'continent': 'North America'}\n"
     ]
    }
   ],
   "source": [
    "augur_dict = augur_export(\n",
    "    tree_path=out_path_nwk, \n",
    "    aln_path=aln_path, \n",
    "    tree=tree, \n",
    "    tree_df=tree_df, \n",
    "    color_keyword_exclude=[\"color\", \"coord\", \"lat\", \"lon\"],\n",
    "    type_convert = {\n",
    "        \"branch_number\" : (lambda x : str(x))\n",
    "    },\n",
    ")\n",
    "\n",
    "out_path_augur_json = os.path.join(outdir, SCRIPT_NAME + \"_augur.json\" )\n",
    "utils.write_json(data=augur_dict, file_name=out_path_augur_json, indent=JSON_INDENT)\n",
    "\n",
    "print(augur_dict[\"nodes\"][\"Reference\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Auspice JSON\n",
    "\n",
    "Manual edits of https://github.com/nextstrain/augur/blob/master/augur/export_v2.py\n",
    "\n",
    "This can then be used for auspice via:\n",
    "\n",
    "```\n",
    "HOST=\"localhost\" auspice view --datasetDir .\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'dict'>\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n",
      "Validating produced JSON\n",
      "Validating schema of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/parse_tree/all/chromosome/filter50/parse_tree_auspice.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/mnt/c/Users/ktmea/Projects/plague-phylogeography/results/parse_tree/all/chromosome/filter50/parse_tree_auspice.json' succeeded, but there were warnings you may want to resolve.\n",
      "Validation successful for local JSON.\n",
      "Validation successful for remote JSON.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: You asked for a color-by for trait 'biosample', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'biosamplecomment', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'datebp', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'rate_fold_change', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'raw_date', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'chrom_cov', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'pcd1_cov', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'pmt1_cov', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'ppcp1_cov', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'pla_cov', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'chrom_dep', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'pcd1_dep', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'pmt1_dep', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'ppcp1_dep', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'pla_dep', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'ppcp1_pla_dep_ratio', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'continent' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: [config file] Trait 'biosample_accession' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: [config file] Trait 'node_type' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: [config file] Trait 'date_bp' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: [config file] Trait 'biosample_comment' is missing type information. We've guessed 'categorical'.\n",
      "\n",
      "WARNING: These values for trait branch_major were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait branch_minor were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait branch_number were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait country were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: These values for trait province were not specified in your provided color scale: na. Auspice will create colors for them.\n",
      "\n",
      "WARNING: country->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "WARNING: province->NA did not have an associated lat/long value (matching performed in lower case). Auspice won't be able to display this location.\n",
      "\n",
      "\tWARNING:  The coloring \"scf\" does not appear as an attr on any tree nodes.\n",
      "\tWARNING:  The coloring \"scf\" does not appear as an attr on any tree nodes.\n"
     ]
    }
   ],
   "source": [
    "auspice_dict = auspice_export(\n",
    "    tree=tree, \n",
    "    augur_json_paths=[out_path_augur_json], \n",
    "    auspice_config_path=auspice_config_path, \n",
    "    auspice_colors_path=auspice_colors_path,\n",
    "    auspice_latlons_path=auspice_latlons_path, \n",
    "    )\n",
    "\n",
    "# Write outputs - For Local Rendering\n",
    "out_path_auspice_local_json = os.path.join(outdir, SCRIPT_NAME + \"_auspice.json\" )\n",
    "utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "print(\"Validation successful for local JSON.\")\n",
    "\n",
    "# Write outputs - For Remote Rendering\n",
    "#out_path_auspice_remote_json = os.path.join(auspice_remote_dir_path, AUSPICE_PREFIX + SCRIPT_NAME.replace(\"_\",\"-\") + \".json\" )\n",
    "#utils.write_json(data=auspice_dict, file_name=out_path_auspice_remote_json, indent=JSON_INDENT, include_version=False)\n",
    "#export_v2.validate_data_json(out_path_auspice_remote_json)\n",
    "print(\"Validation successful for remote JSON.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Cleanup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEPRECATED: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "os.remove(tmp_xml)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
