{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": [
     "snakemake-job-properties"
    ]
   },
   "outputs": [],
   "source": [
    "\n",
    "######## snakemake preamble start (automatically inserted, do not edit) ########\n",
    "import sys; sys.path.extend(['/home/keaton/miniconda3/envs/plague-phylogeography/lib/python3.7/site-packages', '/2/scratch/keaton/plague-phylogeography/workflow/notebooks']); import pickle; snakemake = pickle.loads(b'\\x80\\x03csnakemake.script\\nSnakemake\\nq\\x00)\\x81q\\x01}q\\x02(X\\x05\\x00\\x00\\x00inputq\\x03csnakemake.io\\nInputFiles\\nq\\x04)\\x81q\\x05(Xe\\x00\\x00\\x00/2/scratch/keaton/plague-phylogeography/results/clock/all/chromosome_filter5/clock_model_timetree.nwkq\\x06X\\\\\\x00\\x00\\x00/2/scratch/keaton/plague-phylogeography/results/clock/all/chromosome_filter5/clock_model.tsvq\\x07e}q\\x08(X\\x06\\x00\\x00\\x00_namesq\\t}q\\n(X\\x04\\x00\\x00\\x00treeq\\x0bK\\x00N\\x86q\\x0cX\\x03\\x00\\x00\\x00tsvq\\rK\\x01N\\x86q\\x0euX\\x12\\x00\\x00\\x00_allowed_overridesq\\x0f]q\\x10(X\\x05\\x00\\x00\\x00indexq\\x11X\\x04\\x00\\x00\\x00sortq\\x12eh\\x11cfunctools\\npartial\\nq\\x13cbuiltins\\ngetattr\\nq\\x14csnakemake.io\\nNamedlist\\nq\\x15X\\x0f\\x00\\x00\\x00_used_attributeq\\x16\\x86q\\x17Rq\\x18\\x85q\\x19Rq\\x1a(h\\x18)}q\\x1bX\\x05\\x00\\x00\\x00_nameq\\x1ch\\x11sNtq\\x1dbh\\x12h\\x13h\\x18\\x85q\\x1eRq\\x1f(h\\x18)}q h\\x1ch\\x12sNtq!bh\\x0bh\\x06h\\rh\\x07ubX\\x06\\x00\\x00\\x00outputq\"csnakemake.io\\nOutputFiles\\nq#)\\x81q$(Xm\\x00\\x00\\x00/2/scratch/keaton/plague-phylogeography/results/mugration/all/chromosome_filter5/mugration_model_timetree.nwkq%Xd\\x00\\x00\\x00/2/scratch/keaton/plague-phylogeography/results/mugration/all/chromosome_filter5/mugration_model.tsvq&e}q\\'(h\\t}q((h\\x0bK\\x00N\\x86q)h\\rK\\x01N\\x86q*uh\\x0f]q+(h\\x11h\\x12eh\\x11h\\x13h\\x18\\x85q,Rq-(h\\x18)}q.h\\x1ch\\x11sNtq/bh\\x12h\\x13h\\x18\\x85q0Rq1(h\\x18)}q2h\\x1ch\\x12sNtq3bh\\x0bh%h\\rh&ubX\\x06\\x00\\x00\\x00paramsq4csnakemake.io\\nParams\\nq5)\\x81q6}q7(h\\t}q8h\\x0f]q9(h\\x11h\\x12eh\\x11h\\x13h\\x18\\x85q:Rq;(h\\x18)}q<h\\x1ch\\x11sNtq=bh\\x12h\\x13h\\x18\\x85q>Rq?(h\\x18)}q@h\\x1ch\\x12sNtqAbubX\\t\\x00\\x00\\x00wildcardsqBcsnakemake.io\\nWildcards\\nqC)\\x81qD(X\\x03\\x00\\x00\\x00allqEX\\n\\x00\\x00\\x00chromosomeqFX\\x01\\x00\\x00\\x005qGe}qH(h\\t}qI(X\\x0c\\x00\\x00\\x00reads_originqJK\\x00N\\x86qKX\\n\\x00\\x00\\x00locus_nameqLK\\x01N\\x86qMX\\x0c\\x00\\x00\\x00missing_dataqNK\\x02N\\x86qOuh\\x0f]qP(h\\x11h\\x12eh\\x11h\\x13h\\x18\\x85qQRqR(h\\x18)}qSh\\x1ch\\x11sNtqTbh\\x12h\\x13h\\x18\\x85qURqV(h\\x18)}qWh\\x1ch\\x12sNtqXbX\\x0c\\x00\\x00\\x00reads_originqYhEX\\n\\x00\\x00\\x00locus_nameqZhFX\\x0c\\x00\\x00\\x00missing_dataq[hGubX\\x07\\x00\\x00\\x00threadsq\\\\K\\x01X\\t\\x00\\x00\\x00resourcesq]csnakemake.io\\nResources\\nq^)\\x81q_(K\\x01K\\x01K\\x01M\\xa0\\x0fK<Kde}q`(h\\t}qa(X\\x06\\x00\\x00\\x00_coresqbK\\x00N\\x86qcX\\x06\\x00\\x00\\x00_nodesqdK\\x01N\\x86qeX\\x04\\x00\\x00\\x00cpusqfK\\x02N\\x86qgX\\x06\\x00\\x00\\x00mem_mbqhK\\x03N\\x86qiX\\x08\\x00\\x00\\x00time_minqjK\\x04N\\x86qkX\\x04\\x00\\x00\\x00loadqlK\\x05N\\x86qmuh\\x0f]qn(h\\x11h\\x12eh\\x11h\\x13h\\x18\\x85qoRqp(h\\x18)}qqh\\x1ch\\x11sNtqrbh\\x12h\\x13h\\x18\\x85qsRqt(h\\x18)}quh\\x1ch\\x12sNtqvbhbK\\x01hdK\\x01hfK\\x01X\\x06\\x00\\x00\\x00mem_mbqwM\\xa0\\x0fhjK<hlKdubX\\x03\\x00\\x00\\x00logqxcsnakemake.io\\nLog\\nqy)\\x81qzXs\\x00\\x00\\x00/2/scratch/keaton/plague-phylogeography/results/mugration/all/chromosome_filter5/mugration_model_processed.py.ipynbq{a}q|(h\\t}q}X\\x08\\x00\\x00\\x00notebookq~K\\x00N\\x86q\\x7fsh\\x0f]q\\x80(h\\x11h\\x12eh\\x11h\\x13h\\x18\\x85q\\x81Rq\\x82(h\\x18)}q\\x83h\\x1ch\\x11sNtq\\x84bh\\x12h\\x13h\\x18\\x85q\\x85Rq\\x86(h\\x18)}q\\x87h\\x1ch\\x12sNtq\\x88bh~h{ubX\\x06\\x00\\x00\\x00configq\\x89}q\\x8a(X\\x0b\\x00\\x00\\x00results_dirq\\x8bX\\x07\\x00\\x00\\x00resultsq\\x8cX\\t\\x00\\x00\\x00sqlite_dbq\\x8dX\\x19\\x00\\x00\\x00yersinia_pestis_db.sqliteq\\x8eX\\x19\\x00\\x00\\x00sqlite_select_command_asmq\\x8fX\\x1a\\x01\\x00\\x00SELECT AssemblyFTPGenbank FROM BioSample LEFT JOIN Assembly ON AssemblyBioSampleAccession = BioSampleAccession WHERE BioSampleComment LIKE \\'%KEEP%Assembly%Modern%\\' AND length(AssemblyFTPGenbank) > 0 AND length(BioSampleCollectionDate) > 0 AND length(BioSampleGeographicLocation) > 0q\\x90X\\x19\\x00\\x00\\x00sqlite_select_command_sraq\\x91X\\x1f\\x01\\x00\\x00SELECT BioSampleAccession, SRARunAccession FROM BioSample LEFT JOIN SRA ON SRABioSampleAccession = BioSampleAccession WHERE (BioSampleComment LIKE \\'%KEEP%SRA%Ancient%\\' AND SRAComment NOT LIKE \"%REMOVE%\") AND length(BioSampleCollectionDate) > 0 AND length(BioSampleGeographicLocation) > 0q\\x92X\\x1b\\x00\\x00\\x00sqlite_select_command_localq\\x93XW\\x00\\x00\\x00SELECT BioSampleAccession FROM BioSample WHERE (BioSampleComment LIKE \\'%Local%Modern%\\')q\\x94X\\x19\\x00\\x00\\x00sqlite_select_command_refq\\x95X\\xaa\\x00\\x00\\x00SELECT AssemblyFTPGenbank FROM BioSample LEFT JOIN Assembly ON AssemblyBioSampleAccession = BioSampleAccession WHERE (BioSampleComment LIKE \\'%Assembly%Modern%Reference%\\')q\\x96X\\x15\\x00\\x00\\x00max_datasets_assemblyq\\x97M\\xe8\\x03X\\x10\\x00\\x00\\x00max_datasets_sraq\\x98M\\xe8\\x03X\\x0c\\x00\\x00\\x00reads_originq\\x99]q\\x9a(X\\x08\\x00\\x00\\x00assemblyq\\x9bX\\x03\\x00\\x00\\x00sraq\\x9ceX\\x18\\x00\\x00\\x00detect_repeats_thresholdq\\x9dKZX\\x15\\x00\\x00\\x00detect_repeats_lengthq\\x9eK2X\\x0f\\x00\\x00\\x00reference_locusq\\x9fX\\x08\\x00\\x00\\x00AL590842q\\xa0X\\x14\\x00\\x00\\x00reference_locus_nameq\\xa1X\\n\\x00\\x00\\x00chromosomeq\\xa2X\\x15\\x00\\x00\\x00reference_locus_startq\\xa3X\\x01\\x00\\x00\\x000q\\xa4X\\x13\\x00\\x00\\x00reference_locus_endq\\xa5X\\x07\\x00\\x00\\x004653728q\\xa6X\\t\\x00\\x00\\x00eager_revq\\xa7X\\x05\\x00\\x00\\x002.2.1q\\xa8X\\x15\\x00\\x00\\x00eager_clip_readlengthq\\xa9K#X\\r\\x00\\x00\\x00eager_bwaalnnq\\xaaG?\\x84z\\xe1G\\xae\\x14{X\\r\\x00\\x00\\x00eager_bwaalnlq\\xabK\\x10X\\x0b\\x00\\x00\\x00eager_otherq\\xacX\\x0c\\x00\\x00\\x00--mergedonlyq\\xadX\\x08\\x00\\x00\\x00organismq\\xaeX\\x0f\\x00\\x00\\x00Yersinia pestisq\\xafX\\x15\\x00\\x00\\x00eager_forward_adapterq\\xb0X@\\x00\\x00\\x00AGATCGGAAGAGCACACGTCTGAACTCCAGTCACNNNNNNATCTCGTATGCCGTCTTCTGCTTGq\\xb1X\\x15\\x00\\x00\\x00eager_reverse_adapterq\\xb2X:\\x00\\x00\\x00AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATTq\\xb3X\\x10\\x00\\x00\\x00snippy_ctg_depthq\\xb4K\\nX\\x10\\x00\\x00\\x00snippy_bam_depthq\\xb5K\\x03X\\x10\\x00\\x00\\x00snippy_base_qualq\\xb6K\\x14X\\x0f\\x00\\x00\\x00snippy_map_qualq\\xb7K\\x1eX\\x0f\\x00\\x00\\x00snippy_min_fracq\\xb8G?\\xec\\xcc\\xcc\\xcc\\xcc\\xcc\\xcdX\\x10\\x00\\x00\\x00snippy_mask_charq\\xb9X\\x01\\x00\\x00\\x00Xq\\xbaX\\x13\\x00\\x00\\x00snippy_missing_dataq\\xbbK\\x05X\\x12\\x00\\x00\\x00snippy_snp_densityq\\xbcK\\nX\\x15\\x00\\x00\\x00snippy_keep_singletonq\\xbdX\\x00\\x00\\x00\\x00q\\xbeX\\x1e\\x00\\x00\\x00snippy_multi_plot_missing_dataq\\xbf]q\\xc0(K\\x00K\\x01K\\x02K\\x03K\\x04K\\x05K\\x06K\\x07K\\x08K\\tK\\neX\\x0c\\x00\\x00\\x00iqtree_modelq\\xc1X\\x0b\\x00\\x00\\x00-m K3Pu+F+Iq\\xc2X\\x0b\\x00\\x00\\x00iqtree_seedq\\xc3X\\x08\\x00\\x00\\x0047321424q\\xc4X\\x0f\\x00\\x00\\x00iqtree_outgroupq\\xc5X\\x19\\x00\\x00\\x00SAMEA3541826,SAMEA3541827q\\xc6X\\x0c\\x00\\x00\\x00iqtree_otherq\\xc7X\\r\\x00\\x00\\x00--ufboot 1000q\\xc8X\\x0b\\x00\\x00\\x00iqtree_runsq\\xc9K\\nuX\\x04\\x00\\x00\\x00ruleq\\xcaX\\x0f\\x00\\x00\\x00mugration_modelq\\xcbX\\x0f\\x00\\x00\\x00bench_iterationq\\xccNX\\t\\x00\\x00\\x00scriptdirq\\xcdX:\\x00\\x00\\x00/2/scratch/keaton/plague-phylogeography/workflow/notebooksq\\xceub.'); from snakemake.logging import logger; logger.printshellcmds = False; import os; os.chdir('/2/scratch/keaton/plague-phylogeography');\n",
    "######## snakemake preamble end #########\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Objectives\n",
    "\n",
    "1. Run mugration analysis on select traits.\n",
    "1. Add mugration results to dataframe.\n",
    "1. Plot confidence boxplots.\n",
    "1. Plot colored trees.\n",
    "1. Export:\n",
    "   - Tree Dataframe (tsv)\n",
    "   - Augur JSON\n",
    "   - Auspice JSON"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Module Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Phylo\n",
    "from treetime.utils import parse_dates\n",
    "from treetime import wrappers\n",
    "\n",
    "import copy\n",
    "import os\n",
    "import sys\n",
    "import IPython\n",
    "import io\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import math\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import colors, lines, patches, gridspec\n",
    "import seaborn as sns\n",
    "\n",
    "import dill"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from config import *\n",
    "\n",
    "# Custom script variables\n",
    "SCRIPT_NAME = \"mugration\"\n",
    "PREV_DIR_NAME = \"clock\"\n",
    "PREV_SCRIPT_NAME = \"clock_model\"\n",
    "\n",
    "try:\n",
    "    WILDCARDS = snakemake.wildcards\n",
    "    project_dir = os.getcwd()\n",
    "except NameError:\n",
    "    WILDCARDS = [\"all\", \"chromosome\", \"50\"]\n",
    "    project_dir = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "    \n",
    "READS_ORIGIN = WILDCARDS[0]\n",
    "LOCUS_NAME = WILDCARDS[1]\n",
    "MISSING_DATA = WILDCARDS[2]\n",
    "\n",
    "NAME_COL = \"Name\"\n",
    "\n",
    "plt.rc('legend', frameon=False) # legend frame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Input File Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tree_dill:\t /2/scratch/keaton/plague-phylogeography/results/clock/all/chromosome_filter5/clock_model_timetree.treetime.obj\n",
      "tree_df_dill:\t /2/scratch/keaton/plague-phylogeography/results/clock/all/chromosome_filter5/clock_model.df.obj\n",
      "aln path:\t /2/scratch/keaton/plague-phylogeography/results/snippy_multi/all/snippy-core_chromosome.snps.filter5.aln\n",
      "auspice_latlon_path: /2/scratch/keaton/plague-phylogeography/results/parse_tree/all/chromosome_filter5/parse_tree_latlon.tsv\n",
      "auspice_colors_path: /2/scratch/keaton/plague-phylogeography/results/parse_tree/all/chromosome_filter5/parse_tree_colors.tsv\n",
      "auspice_config_path: /2/scratch/keaton/plague-phylogeography/results/config/auspice_config.json\n",
      "auspice_remote_dir_path: /2/scratch/keaton/plague-phylogeography/auspice/\n",
      "outdir: /2/scratch/keaton/plague-phylogeography/results/mugration/all/chromosome_filter5\n"
     ]
    }
   ],
   "source": [
    "results_dir = os.path.join(project_dir, \"results\")\n",
    "config_dir = os.path.join(results_dir, \"config\")\n",
    "\n",
    "\n",
    "outdir       = os.path.join(results_dir, SCRIPT_NAME, READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA),)\n",
    "tree_dill    = os.path.join(results_dir, PREV_DIR_NAME, READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA),PREV_SCRIPT_NAME + \"_timetree.treetime.obj\")\n",
    "tree_df_dill = os.path.join(results_dir, PREV_DIR_NAME, READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA), PREV_SCRIPT_NAME + \".df.obj\")\n",
    "tree_df_path = os.path.join(results_dir, PREV_DIR_NAME, READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA), PREV_SCRIPT_NAME + \".tsv\")\n",
    "aln_path     = os.path.join(results_dir,\"snippy_multi\",READS_ORIGIN,\"snippy-core_{}.snps.filter{}.aln\".format(LOCUS_NAME, MISSING_DATA))\n",
    "\n",
    "# Auspice\n",
    "auspice_latlon_path = os.path.join(results_dir, \"parse_tree\", READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA), \"parse_tree\" + \"_latlon.tsv\")\n",
    "auspice_colors_path = os.path.join(results_dir, \"parse_tree\", READS_ORIGIN, LOCUS_NAME + \"_filter{}\".format(MISSING_DATA), \"parse_tree\" + \"_colors.tsv\")\n",
    "auspice_config_path = os.path.join(config_dir, \"auspice_config.json\")\n",
    "auspice_remote_dir_path = os.path.join(project_dir, \"auspice/\")\n",
    "\n",
    "print(\"tree_dill:\\t\", tree_dill)\n",
    "print(\"tree_df_dill:\\t\", tree_df_dill)\n",
    "print(\"aln path:\\t\", aln_path)\n",
    "print(\"auspice_latlon_path:\", auspice_latlon_path)\n",
    "print(\"auspice_colors_path:\", auspice_colors_path)\n",
    "print(\"auspice_config_path:\", auspice_config_path)\n",
    "print(\"auspice_remote_dir_path:\", auspice_remote_dir_path)\n",
    "print(\"outdir:\", outdir)\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "while not os.path.exists(outdir):\n",
    "    os.makedirs(outdir)   \n",
    "    \n",
    "SCRIPT_NAME = \"mugration_model\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tree_div = Phylo.read(tree_path, \"newick\")\n",
    "#tree_div.ladderize(reverse=False)\n",
    "\n",
    "with open(tree_dill, \"rb\") as infile:\n",
    "    tt = dill.load(infile)\n",
    "tt.tree.ladderize(reverse=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UFboot</th>\n",
       "      <th>sCF</th>\n",
       "      <th>Branch_Length</th>\n",
       "      <th>Branch_Support_Color</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Date</th>\n",
       "      <th>DateBP</th>\n",
       "      <th>Country</th>\n",
       "      <th>Province</th>\n",
       "      <th>CountryLat</th>\n",
       "      <th>CountryLon</th>\n",
       "      <th>ProvinceLat</th>\n",
       "      <th>ProvinceLon</th>\n",
       "      <th>Biovar</th>\n",
       "      <th>Branch_Major</th>\n",
       "      <th>Branch_Minor</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleComment</th>\n",
       "      <th>Branch_Number</th>\n",
       "      <th>timetree_rate</th>\n",
       "      <th>timetree_rate_fold_change</th>\n",
       "      <th>timetree_mutation_length</th>\n",
       "      <th>timetree_raw_date</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>timetree_num_date_bar</th>\n",
       "      <th>timetree_clock_length</th>\n",
       "      <th>timetree_coord_x</th>\n",
       "      <th>timetree_coord_y</th>\n",
       "      <th>timetree_reg_x</th>\n",
       "      <th>timetree_reg_y</th>\n",
       "      <th>timetree_reg_bad</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>100</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>black</td>\n",
       "      <td>0.001</td>\n",
       "      <td>15.542</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>9.056080e-09</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000e-03</td>\n",
       "      <td>-4001-04-08</td>\n",
       "      <td>-4000.733897</td>\n",
       "      <td>[-4028.4379593546546, -3773.0741326132847]</td>\n",
       "      <td>[6049.437959354655, 5794.074132613285]</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>15.541999</td>\n",
       "      <td>-4000.733897</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3541827</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.8408e-06</td>\n",
       "      <td>grey</td>\n",
       "      <td>0.00100322</td>\n",
       "      <td>1</td>\n",
       "      <td>RISE509</td>\n",
       "      <td>[-2876:-2677]</td>\n",
       "      <td>[-4897:-4698]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Krasnoyarsk Krai</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>63.3234</td>\n",
       "      <td>97.098</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.PRE2</td>\n",
       "      <td>SAMEA3541827</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>0</td>\n",
       "      <td>2.775061e-09</td>\n",
       "      <td>0.306431</td>\n",
       "      <td>3.224758e-06</td>\n",
       "      <td>-2879-06-08</td>\n",
       "      <td>-2878.564956</td>\n",
       "      <td>[-2876.0, -2677.0]</td>\n",
       "      <td>[4897.0, 4698.0]</td>\n",
       "      <td>1122.168941</td>\n",
       "      <td>1122.268941</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-2776.500000</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>100</td>\n",
       "      <td>84.9</td>\n",
       "      <td>1.34224e-05</td>\n",
       "      <td>black</td>\n",
       "      <td>0.00100796</td>\n",
       "      <td>3.15625</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.322885e-08</td>\n",
       "      <td>1.460770</td>\n",
       "      <td>7.959900e-06</td>\n",
       "      <td>-3372-09-28</td>\n",
       "      <td>-3371.257753</td>\n",
       "      <td>[-3386.8297780304565, -3135.38189068595]</td>\n",
       "      <td>[5407.8297780304565, 5156.381890685951]</td>\n",
       "      <td>629.476144</td>\n",
       "      <td>629.576144</td>\n",
       "      <td>3.156250</td>\n",
       "      <td>-3371.257753</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233046</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.233e-06</td>\n",
       "      <td>grey</td>\n",
       "      <td>0.00101011</td>\n",
       "      <td>2</td>\n",
       "      <td>Gyvakarai1</td>\n",
       "      <td>[-2621:-2472]</td>\n",
       "      <td>[-4642:-4493]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Panevezys County</td>\n",
       "      <td>55.35</td>\n",
       "      <td>23.75</td>\n",
       "      <td>55.9156</td>\n",
       "      <td>25.0312</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.PRE1</td>\n",
       "      <td>SAMEA104233046</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>0</td>\n",
       "      <td>2.489749e-09</td>\n",
       "      <td>0.274926</td>\n",
       "      <td>2.150063e-06</td>\n",
       "      <td>-2600-02-18</td>\n",
       "      <td>-2599.866006</td>\n",
       "      <td>[-2621.0, -2480.907503502162]</td>\n",
       "      <td>[4642.0, 4501.9075035021615]</td>\n",
       "      <td>771.391746</td>\n",
       "      <td>1400.967891</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>-2546.500000</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>100</td>\n",
       "      <td>94.7</td>\n",
       "      <td>1.34762e-05</td>\n",
       "      <td>black</td>\n",
       "      <td>0.00102217</td>\n",
       "      <td>4.3125</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.959170e-08</td>\n",
       "      <td>2.163375</td>\n",
       "      <td>1.421271e-05</td>\n",
       "      <td>-2644-06-12</td>\n",
       "      <td>-2643.555870</td>\n",
       "      <td>[-2758.1911839442546, -2599.428548849007]</td>\n",
       "      <td>[4779.191183944255, 4620.428548849007]</td>\n",
       "      <td>727.701883</td>\n",
       "      <td>1357.278028</td>\n",
       "      <td>4.312500</td>\n",
       "      <td>-2643.555870</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_0000408</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.00106841</td>\n",
       "      <td>593.344</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>9.056080e-09</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.068425e-22</td>\n",
       "      <td>1978-02-16</td>\n",
       "      <td>1978.128239</td>\n",
       "      <td>[1968.4891921919855, 1983.692453807525]</td>\n",
       "      <td>[52.51080780801453, 37.307546192475]</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5978.962136</td>\n",
       "      <td>593.343750</td>\n",
       "      <td>1978.128239</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_0000411</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.00106841</td>\n",
       "      <td>594.938</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.510844e-09</td>\n",
       "      <td>0.829370</td>\n",
       "      <td>1.068425e-22</td>\n",
       "      <td>1978-05-16</td>\n",
       "      <td>1978.371715</td>\n",
       "      <td>[1968.5283949205668, 1984.0160167471524]</td>\n",
       "      <td>[52.471605079433175, 36.983983252847565]</td>\n",
       "      <td>0.243476</td>\n",
       "      <td>5979.205612</td>\n",
       "      <td>594.937500</td>\n",
       "      <td>1978.371715</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_0000412</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.00106841</td>\n",
       "      <td>595.875</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.947457e-09</td>\n",
       "      <td>0.546313</td>\n",
       "      <td>1.068425e-22</td>\n",
       "      <td>1986-11-09</td>\n",
       "      <td>1986.856321</td>\n",
       "      <td>[1968.5469333253366, 2000.6367394245203]</td>\n",
       "      <td>[52.45306667466343, 20.36326057547967]</td>\n",
       "      <td>8.484607</td>\n",
       "      <td>5987.690219</td>\n",
       "      <td>595.875000</td>\n",
       "      <td>1986.856321</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_0000413</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.00106841</td>\n",
       "      <td>596.75</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>8.903422e-09</td>\n",
       "      <td>0.983143</td>\n",
       "      <td>1.068425e-22</td>\n",
       "      <td>1986-11-09</td>\n",
       "      <td>1986.856321</td>\n",
       "      <td>[1968.6420434068189, 2000.4761805051378]</td>\n",
       "      <td>[52.35795659318114, 20.52381949486221]</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5987.690219</td>\n",
       "      <td>596.750000</td>\n",
       "      <td>1986.856321</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_0000414</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.00106841</td>\n",
       "      <td>597.5</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>9.056080e-09</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.068425e-22</td>\n",
       "      <td>1986-11-09</td>\n",
       "      <td>1986.856321</td>\n",
       "      <td>[1968.633783825002, 2000.397427341175]</td>\n",
       "      <td>[52.366216174998044, 20.60257265882501]</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5987.690219</td>\n",
       "      <td>597.500000</td>\n",
       "      <td>1986.856321</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1050 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               UFboot   sCF Branch_Length Branch_Support_Color     coord_x  \\\n",
       "Name                                                                         \n",
       "NODE0             100    NA             0                black       0.001   \n",
       "SAMEA3541827       NA    NA    3.8408e-06                 grey  0.00100322   \n",
       "NODE1             100  84.9   1.34224e-05                black  0.00100796   \n",
       "SAMEA104233046     NA    NA     2.233e-06                 grey  0.00101011   \n",
       "NODE2             100  94.7   1.34762e-05                black  0.00102217   \n",
       "...               ...   ...           ...                  ...         ...   \n",
       "NODE_0000408       NA    NA            NA                   NA  0.00106841   \n",
       "NODE_0000411       NA    NA            NA                   NA  0.00106841   \n",
       "NODE_0000412       NA    NA            NA                   NA  0.00106841   \n",
       "NODE_0000413       NA    NA            NA                   NA  0.00106841   \n",
       "NODE_0000414       NA    NA            NA                   NA  0.00106841   \n",
       "\n",
       "                coord_y      Strain           Date         DateBP    Country  \\\n",
       "Name                                                                           \n",
       "NODE0            15.542          NA             NA             NA         NA   \n",
       "SAMEA3541827          1     RISE509  [-2876:-2677]  [-4897:-4698]     Russia   \n",
       "NODE1           3.15625          NA             NA             NA         NA   \n",
       "SAMEA104233046        2  Gyvakarai1  [-2621:-2472]  [-4642:-4493]  Lithuania   \n",
       "NODE2            4.3125          NA             NA             NA         NA   \n",
       "...                 ...         ...            ...            ...        ...   \n",
       "NODE_0000408    593.344          NA             NA             NA         NA   \n",
       "NODE_0000411    594.938          NA             NA             NA         NA   \n",
       "NODE_0000412    595.875          NA             NA             NA         NA   \n",
       "NODE_0000413     596.75          NA             NA             NA         NA   \n",
       "NODE_0000414      597.5          NA             NA             NA         NA   \n",
       "\n",
       "                        Province CountryLat CountryLon ProvinceLat  \\\n",
       "Name                                                                 \n",
       "NODE0                         NA         NA         NA          NA   \n",
       "SAMEA3541827    Krasnoyarsk Krai    64.6863    97.7453     63.3234   \n",
       "NODE1                         NA         NA         NA          NA   \n",
       "SAMEA104233046  Panevezys County      55.35      23.75     55.9156   \n",
       "NODE2                         NA         NA         NA          NA   \n",
       "...                          ...        ...        ...         ...   \n",
       "NODE_0000408                  NA         NA         NA          NA   \n",
       "NODE_0000411                  NA         NA         NA          NA   \n",
       "NODE_0000412                  NA         NA         NA          NA   \n",
       "NODE_0000413                  NA         NA         NA          NA   \n",
       "NODE_0000414                  NA         NA         NA          NA   \n",
       "\n",
       "               ProvinceLon      Biovar Branch_Major Branch_Minor  \\\n",
       "Name                                                               \n",
       "NODE0                   NA          NA           NA           NA   \n",
       "SAMEA3541827        97.098  Bronze Age        0.PRE       0.PRE2   \n",
       "NODE1                   NA          NA           NA           NA   \n",
       "SAMEA104233046     25.0312  Bronze Age        0.PRE       0.PRE1   \n",
       "NODE2                   NA          NA           NA           NA   \n",
       "...                    ...         ...          ...          ...   \n",
       "NODE_0000408            NA          NA           NA           NA   \n",
       "NODE_0000411            NA          NA           NA           NA   \n",
       "NODE_0000412            NA          NA           NA           NA   \n",
       "NODE_0000413            NA          NA           NA           NA   \n",
       "NODE_0000414            NA          NA           NA           NA   \n",
       "\n",
       "                     BioSample   BioSampleComment Branch_Number  \\\n",
       "Name                                                              \n",
       "NODE0                       NA                 NA            NA   \n",
       "SAMEA3541827      SAMEA3541827  KEEP: SRA Ancient             0   \n",
       "NODE1                       NA                 NA            NA   \n",
       "SAMEA104233046  SAMEA104233046  KEEP: SRA Ancient             0   \n",
       "NODE2                       NA                 NA            NA   \n",
       "...                        ...                ...           ...   \n",
       "NODE_0000408                NA                 NA            NA   \n",
       "NODE_0000411                NA                 NA            NA   \n",
       "NODE_0000412                NA                 NA            NA   \n",
       "NODE_0000413                NA                 NA            NA   \n",
       "NODE_0000414                NA                 NA            NA   \n",
       "\n",
       "                timetree_rate  timetree_rate_fold_change  \\\n",
       "Name                                                       \n",
       "NODE0            9.056080e-09                   1.000000   \n",
       "SAMEA3541827     2.775061e-09                   0.306431   \n",
       "NODE1            1.322885e-08                   1.460770   \n",
       "SAMEA104233046   2.489749e-09                   0.274926   \n",
       "NODE2            1.959170e-08                   2.163375   \n",
       "...                       ...                        ...   \n",
       "NODE_0000408     9.056080e-09                   1.000000   \n",
       "NODE_0000411     7.510844e-09                   0.829370   \n",
       "NODE_0000412     4.947457e-09                   0.546313   \n",
       "NODE_0000413     8.903422e-09                   0.983143   \n",
       "NODE_0000414     9.056080e-09                   1.000000   \n",
       "\n",
       "                timetree_mutation_length timetree_raw_date  timetree_num_date  \\\n",
       "Name                                                                            \n",
       "NODE0                       1.000000e-03       -4001-04-08       -4000.733897   \n",
       "SAMEA3541827                3.224758e-06       -2879-06-08       -2878.564956   \n",
       "NODE1                       7.959900e-06       -3372-09-28       -3371.257753   \n",
       "SAMEA104233046              2.150063e-06       -2600-02-18       -2599.866006   \n",
       "NODE2                       1.421271e-05       -2644-06-12       -2643.555870   \n",
       "...                                  ...               ...                ...   \n",
       "NODE_0000408                1.068425e-22        1978-02-16        1978.128239   \n",
       "NODE_0000411                1.068425e-22        1978-05-16        1978.371715   \n",
       "NODE_0000412                1.068425e-22        1986-11-09        1986.856321   \n",
       "NODE_0000413                1.068425e-22        1986-11-09        1986.856321   \n",
       "NODE_0000414                1.068425e-22        1986-11-09        1986.856321   \n",
       "\n",
       "                              timetree_num_date_confidence  \\\n",
       "Name                                                         \n",
       "NODE0           [-4028.4379593546546, -3773.0741326132847]   \n",
       "SAMEA3541827                            [-2876.0, -2677.0]   \n",
       "NODE1             [-3386.8297780304565, -3135.38189068595]   \n",
       "SAMEA104233046               [-2621.0, -2480.907503502162]   \n",
       "NODE2            [-2758.1911839442546, -2599.428548849007]   \n",
       "...                                                    ...   \n",
       "NODE_0000408       [1968.4891921919855, 1983.692453807525]   \n",
       "NODE_0000411      [1968.5283949205668, 1984.0160167471524]   \n",
       "NODE_0000412      [1968.5469333253366, 2000.6367394245203]   \n",
       "NODE_0000413      [1968.6420434068189, 2000.4761805051378]   \n",
       "NODE_0000414        [1968.633783825002, 2000.397427341175]   \n",
       "\n",
       "                                   timetree_num_date_bar  \\\n",
       "Name                                                       \n",
       "NODE0             [6049.437959354655, 5794.074132613285]   \n",
       "SAMEA3541827                            [4897.0, 4698.0]   \n",
       "NODE1            [5407.8297780304565, 5156.381890685951]   \n",
       "SAMEA104233046              [4642.0, 4501.9075035021615]   \n",
       "NODE2             [4779.191183944255, 4620.428548849007]   \n",
       "...                                                  ...   \n",
       "NODE_0000408        [52.51080780801453, 37.307546192475]   \n",
       "NODE_0000411    [52.471605079433175, 36.983983252847565]   \n",
       "NODE_0000412      [52.45306667466343, 20.36326057547967]   \n",
       "NODE_0000413      [52.35795659318114, 20.52381949486221]   \n",
       "NODE_0000414     [52.366216174998044, 20.60257265882501]   \n",
       "\n",
       "                timetree_clock_length  timetree_coord_x  timetree_coord_y  \\\n",
       "Name                                                                        \n",
       "NODE0                        0.100000          0.100000         15.541999   \n",
       "SAMEA3541827              1122.168941       1122.268941          1.000000   \n",
       "NODE1                      629.476144        629.576144          3.156250   \n",
       "SAMEA104233046             771.391746       1400.967891          2.000000   \n",
       "NODE2                      727.701883       1357.278028          4.312500   \n",
       "...                               ...               ...               ...   \n",
       "NODE_0000408                 0.000000       5978.962136        593.343750   \n",
       "NODE_0000411                 0.243476       5979.205612        594.937500   \n",
       "NODE_0000412                 8.484607       5987.690219        595.875000   \n",
       "NODE_0000413                 0.000000       5987.690219        596.750000   \n",
       "NODE_0000414                 0.000000       5987.690219        597.500000   \n",
       "\n",
       "                timetree_reg_x  timetree_reg_y  timetree_reg_bad  \n",
       "Name                                                              \n",
       "NODE0             -4000.733897        0.000000             False  \n",
       "SAMEA3541827      -2776.500000        0.000003             False  \n",
       "NODE1             -3371.257753        0.000008             False  \n",
       "SAMEA104233046    -2546.500000        0.000010             False  \n",
       "NODE2             -2643.555870        0.000022             False  \n",
       "...                        ...             ...               ...  \n",
       "NODE_0000408       1978.128239        0.000068             False  \n",
       "NODE_0000411       1978.371715        0.000068             False  \n",
       "NODE_0000412       1986.856321        0.000068             False  \n",
       "NODE_0000413       1986.856321        0.000068             False  \n",
       "NODE_0000414       1986.856321        0.000068             False  \n",
       "\n",
       "[1050 rows x 34 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(tree_df_dill, \"rb\") as infile:\n",
    "    tree_df = dill.load(infile)\n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import lat lon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geo</th>\n",
       "      <th>Name</th>\n",
       "      <th>Lat</th>\n",
       "      <th>Lon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>country</td>\n",
       "      <td>Russia</td>\n",
       "      <td>64.686314</td>\n",
       "      <td>97.745306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>country</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>55.350000</td>\n",
       "      <td>23.750000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>country</td>\n",
       "      <td>Estonia</td>\n",
       "      <td>58.752378</td>\n",
       "      <td>25.331908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>country</td>\n",
       "      <td>Germany</td>\n",
       "      <td>51.083420</td>\n",
       "      <td>10.423447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>country</td>\n",
       "      <td>China</td>\n",
       "      <td>35.000074</td>\n",
       "      <td>104.999927</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>province</td>\n",
       "      <td>MSila</td>\n",
       "      <td>35.130021</td>\n",
       "      <td>4.200311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>province</td>\n",
       "      <td>State of Mato Grosso</td>\n",
       "      <td>-12.211501</td>\n",
       "      <td>-55.571655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>province</td>\n",
       "      <td>Midlands</td>\n",
       "      <td>-19.278504</td>\n",
       "      <td>29.879085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>province</td>\n",
       "      <td>La Paz</td>\n",
       "      <td>-16.495545</td>\n",
       "      <td>-68.133623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>province</td>\n",
       "      <td>Cajamarca</td>\n",
       "      <td>-6.250000</td>\n",
       "      <td>-78.833333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>154 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Geo                  Name        Lat         Lon\n",
       "0     country                Russia  64.686314   97.745306\n",
       "1     country             Lithuania  55.350000   23.750000\n",
       "2     country               Estonia  58.752378   25.331908\n",
       "3     country               Germany  51.083420   10.423447\n",
       "4     country                 China  35.000074  104.999927\n",
       "..        ...                   ...        ...         ...\n",
       "149  province                 MSila  35.130021    4.200311\n",
       "150  province  State of Mato Grosso -12.211501  -55.571655\n",
       "151  province              Midlands -19.278504   29.879085\n",
       "152  province                La Paz -16.495545  -68.133623\n",
       "153  province             Cajamarca  -6.250000  -78.833333\n",
       "\n",
       "[154 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "latlon_df = pd.read_csv(auspice_latlon_path, sep='\\t', header=None)\n",
    "latlon_df.columns = [\"Geo\",\"Name\",\"Lat\",\"Lon\"]\n",
    "# Fix the problem with multiple forms of NA in the table\n",
    "# Consolidate missing data to the NO_DATA_CHAR\n",
    "latlon_df.fillna(NO_DATA_CHAR, inplace=True)\n",
    "latlon_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 1. Run mugration analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup Mugration Dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "mug_dict = {}\n",
    "\n",
    "\n",
    "tt_copy = copy.deepcopy(tt)\n",
    "# Set branch length to mutations\n",
    "for n in tt_copy.tree.find_clades():\n",
    "    n.branch_length=n.mutation_length\n",
    "tree_div = tt_copy.tree\n",
    "\n",
    "for attr in ATTRIBUTE_LIST:\n",
    "    mug_dict[attr] = {}\n",
    "    # Make a copy of the input tree\n",
    "    mug_dict[attr][\"tree_div\"] = copy.deepcopy(tree_div)\n",
    "    # Map the taxon name to the attribute for mugration analysis\n",
    "    #mug_dict[attr][\"leaf_to_attr\"] = {sample:str(data[attr]) for sample,data in tree_df.iterrows()\n",
    "    #                if data[attr]!=NO_DATA_CHAR and data[attr]}    \n",
    "    mug_dict[attr][\"leaf_to_attr\"] = {sample:str(data[attr]) for sample,data in tree_df.iterrows()\n",
    "                    if data[attr]!=NO_DATA_CHAR}     \n",
    "    # Make blank dicts for the mugration output\n",
    "    mug_dict[attr][\"tree_mug\"] = {}\n",
    "    mug_dict[attr][\"letter_to_state\"] = {}\n",
    "    mug_dict[attr][\"reverse_alphabet\"] = {}\n",
    "    mug_dict[attr][\"unique_states\"] = {}\n",
    "\n",
    "    \n",
    "# Perform necessary type conversions\n",
    "\n",
    "if \"Branch_Number\" in mug_dict.keys():\n",
    "    for sample,attr_val in mug_dict[\"Branch_Number\"][\"leaf_to_attr\"].items():\n",
    "        float_to_str = str(math.ceil(float(attr_val)))\n",
    "        mug_dict[\"Branch_Number\"][\"leaf_to_attr\"][sample] = float_to_str"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run mugration to capture log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running mugration for attribute: Branch_Number\n",
      "Standard output restored, logging to file disabled.\n",
      "Running mugration for attribute: Branch_Major\n",
      "Standard output restored, logging to file disabled.\n",
      "Running mugration for attribute: Branch_Minor\n",
      "Standard output restored, logging to file disabled.\n",
      "Running mugration for attribute: Country\n",
      "Standard output restored, logging to file disabled.\n",
      "Running mugration for attribute: Province\n",
      "Standard output restored, logging to file disabled.\n"
     ]
    }
   ],
   "source": [
    "for attr in ATTRIBUTE_LIST:\n",
    "    # Initialize stdout capture\n",
    "    print(\"Running mugration for attribute: {}\".format(attr))    \n",
    "    old_stdout = sys.stdout\n",
    "    new_stdout = io.StringIO()\n",
    "    sys.stdout = new_stdout\n",
    "    \n",
    "    # Run mugration\n",
    "    mug, letter_to_state, reverse_alphabet = wrappers.reconstruct_discrete_traits(mug_dict[attr][\"tree_div\"], \n",
    "                                                                     traits=mug_dict[attr][\"leaf_to_attr\"], \n",
    "                                                                     missing_data=NO_DATA_CHAR,\n",
    "                                                                     #missing_data=\"nan\",\n",
    "                                                                     #pc=pc, \n",
    "                                                                     #sampling_bias_correction=sampling_bias_correction, \n",
    "                                                                     verbose=4, \n",
    "                                                                     #weights=params.weights\n",
    "                                                                    )\n",
    "    mug_dict[attr][\"tree_mug\"] = mug\n",
    "    mug_dict[attr][\"letter_to_state\"] = letter_to_state\n",
    "    mug_dict[attr][\"reverse_alphabet\"] = reverse_alphabet\n",
    "    mug_dict[attr][\"unique_states\"] = sorted(letter_to_state.values())\n",
    "    \n",
    "    # Save stdout to file\n",
    "    output = new_stdout.getvalue()\n",
    "    out_path = os.path.join(outdir, SCRIPT_NAME + \"_{}.log\".format(attr.lower())) \n",
    "    with open(out_path, \"w\") as file:\n",
    "        file.write(output)\n",
    "    # Restore stdout\n",
    "    sys.stdout = old_stdout\n",
    "    print(\"Standard output restored, logging to file disabled.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 2. Add mugration to dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UFboot</th>\n",
       "      <th>sCF</th>\n",
       "      <th>Branch_Length</th>\n",
       "      <th>Branch_Support_Color</th>\n",
       "      <th>coord_x</th>\n",
       "      <th>coord_y</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Date</th>\n",
       "      <th>DateBP</th>\n",
       "      <th>Country</th>\n",
       "      <th>Province</th>\n",
       "      <th>CountryLat</th>\n",
       "      <th>CountryLon</th>\n",
       "      <th>ProvinceLat</th>\n",
       "      <th>ProvinceLon</th>\n",
       "      <th>Biovar</th>\n",
       "      <th>Branch_Major</th>\n",
       "      <th>Branch_Minor</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleComment</th>\n",
       "      <th>Branch_Number</th>\n",
       "      <th>timetree_rate</th>\n",
       "      <th>timetree_rate_fold_change</th>\n",
       "      <th>timetree_mutation_length</th>\n",
       "      <th>timetree_raw_date</th>\n",
       "      <th>timetree_num_date</th>\n",
       "      <th>timetree_num_date_confidence</th>\n",
       "      <th>timetree_num_date_bar</th>\n",
       "      <th>timetree_clock_length</th>\n",
       "      <th>timetree_coord_x</th>\n",
       "      <th>timetree_coord_y</th>\n",
       "      <th>timetree_reg_x</th>\n",
       "      <th>timetree_reg_y</th>\n",
       "      <th>timetree_reg_bad</th>\n",
       "      <th>Mugration_Branch_Number</th>\n",
       "      <th>Mugration_Branch_Number_Confidence</th>\n",
       "      <th>Mugration_Branch_Number_Entropy</th>\n",
       "      <th>Mugration_Branch_Major</th>\n",
       "      <th>Mugration_Branch_Major_Confidence</th>\n",
       "      <th>Mugration_Branch_Major_Entropy</th>\n",
       "      <th>Mugration_Branch_Minor</th>\n",
       "      <th>Mugration_Branch_Minor_Confidence</th>\n",
       "      <th>Mugration_Branch_Minor_Entropy</th>\n",
       "      <th>Mugration_Country</th>\n",
       "      <th>Mugration_Country_Confidence</th>\n",
       "      <th>Mugration_Country_Entropy</th>\n",
       "      <th>Mugration_Country_Lat</th>\n",
       "      <th>Mugration_Country_Lon</th>\n",
       "      <th>Mugration_Province</th>\n",
       "      <th>Mugration_Province_Confidence</th>\n",
       "      <th>Mugration_Province_Entropy</th>\n",
       "      <th>Mugration_Province_Lat</th>\n",
       "      <th>Mugration_Province_Lon</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NODE0</th>\n",
       "      <td>100</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>black</td>\n",
       "      <td>0.001</td>\n",
       "      <td>15.542</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>9.056080e-09</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000e-03</td>\n",
       "      <td>-4001-04-08</td>\n",
       "      <td>-4000.733897</td>\n",
       "      <td>[-4028.4379593546546, -3773.0741326132847]</td>\n",
       "      <td>[6049.437959354655, 5794.074132613285]</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>15.541999</td>\n",
       "      <td>-4000.733897</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>8.85652e-08</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.998709</td>\n",
       "      <td>0.00991765</td>\n",
       "      <td>0.PRE2</td>\n",
       "      <td>0.392407</td>\n",
       "      <td>1.32976</td>\n",
       "      <td>Russia</td>\n",
       "      <td>0.993164</td>\n",
       "      <td>0.0442662</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>Krasnoyarsk Krai</td>\n",
       "      <td>0.382595</td>\n",
       "      <td>1.64309</td>\n",
       "      <td>63.3234</td>\n",
       "      <td>97.098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3541827</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.8408e-06</td>\n",
       "      <td>grey</td>\n",
       "      <td>0.00100322</td>\n",
       "      <td>1</td>\n",
       "      <td>RISE509</td>\n",
       "      <td>[-2876:-2677]</td>\n",
       "      <td>[-4897:-4698]</td>\n",
       "      <td>Russia</td>\n",
       "      <td>Krasnoyarsk Krai</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>63.3234</td>\n",
       "      <td>97.098</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.PRE2</td>\n",
       "      <td>SAMEA3541827</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>0</td>\n",
       "      <td>2.775061e-09</td>\n",
       "      <td>0.306431</td>\n",
       "      <td>3.224758e-06</td>\n",
       "      <td>-2879-06-08</td>\n",
       "      <td>-2878.564956</td>\n",
       "      <td>[-2876.0, -2677.0]</td>\n",
       "      <td>[4897.0, 4698.0]</td>\n",
       "      <td>1122.168941</td>\n",
       "      <td>1122.268941</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-2776.500000</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>1</td>\n",
       "      <td>-9.99867e-13</td>\n",
       "      <td>0.PRE2</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>Russia</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>Krasnoyarsk Krai</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>63.3234</td>\n",
       "      <td>97.098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE1</th>\n",
       "      <td>100</td>\n",
       "      <td>84.9</td>\n",
       "      <td>1.34224e-05</td>\n",
       "      <td>black</td>\n",
       "      <td>0.00100796</td>\n",
       "      <td>3.15625</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.322885e-08</td>\n",
       "      <td>1.460770</td>\n",
       "      <td>7.959900e-06</td>\n",
       "      <td>-3372-09-28</td>\n",
       "      <td>-3371.257753</td>\n",
       "      <td>[-3386.8297780304565, -3135.38189068595]</td>\n",
       "      <td>[5407.8297780304565, 5156.381890685951]</td>\n",
       "      <td>629.476144</td>\n",
       "      <td>629.576144</td>\n",
       "      <td>3.156250</td>\n",
       "      <td>-3371.257753</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>8.85651e-08</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>2.40643e-05</td>\n",
       "      <td>0.PRE1</td>\n",
       "      <td>0.999201</td>\n",
       "      <td>0.00737495</td>\n",
       "      <td>Russia</td>\n",
       "      <td>0.979419</td>\n",
       "      <td>0.127114</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>Panevezys County</td>\n",
       "      <td>0.443092</td>\n",
       "      <td>1.91578</td>\n",
       "      <td>55.9156</td>\n",
       "      <td>25.0312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA104233046</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.233e-06</td>\n",
       "      <td>grey</td>\n",
       "      <td>0.00101011</td>\n",
       "      <td>2</td>\n",
       "      <td>Gyvakarai1</td>\n",
       "      <td>[-2621:-2472]</td>\n",
       "      <td>[-4642:-4493]</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>Panevezys County</td>\n",
       "      <td>55.35</td>\n",
       "      <td>23.75</td>\n",
       "      <td>55.9156</td>\n",
       "      <td>25.0312</td>\n",
       "      <td>Bronze Age</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>0.PRE1</td>\n",
       "      <td>SAMEA104233046</td>\n",
       "      <td>KEEP: SRA Ancient</td>\n",
       "      <td>0</td>\n",
       "      <td>2.489749e-09</td>\n",
       "      <td>0.274926</td>\n",
       "      <td>2.150063e-06</td>\n",
       "      <td>-2600-02-18</td>\n",
       "      <td>-2599.866006</td>\n",
       "      <td>[-2621.0, -2480.907503502162]</td>\n",
       "      <td>[4642.0, 4501.9075035021615]</td>\n",
       "      <td>771.391746</td>\n",
       "      <td>1400.967891</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>-2546.500000</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>0.PRE1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>Lithuania</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>55.35</td>\n",
       "      <td>23.75</td>\n",
       "      <td>Panevezys County</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.00009e-12</td>\n",
       "      <td>55.9156</td>\n",
       "      <td>25.0312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE2</th>\n",
       "      <td>100</td>\n",
       "      <td>94.7</td>\n",
       "      <td>1.34762e-05</td>\n",
       "      <td>black</td>\n",
       "      <td>0.00102217</td>\n",
       "      <td>4.3125</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.959170e-08</td>\n",
       "      <td>2.163375</td>\n",
       "      <td>1.421271e-05</td>\n",
       "      <td>-2644-06-12</td>\n",
       "      <td>-2643.555870</td>\n",
       "      <td>[-2758.1911839442546, -2599.428548849007]</td>\n",
       "      <td>[4779.191183944255, 4620.428548849007]</td>\n",
       "      <td>727.701883</td>\n",
       "      <td>1357.278028</td>\n",
       "      <td>4.312500</td>\n",
       "      <td>-2643.555870</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>8.85652e-08</td>\n",
       "      <td>0.PRE</td>\n",
       "      <td>1</td>\n",
       "      <td>2.20342e-07</td>\n",
       "      <td>0.PRE1</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.74031e-05</td>\n",
       "      <td>Russia</td>\n",
       "      <td>0.978827</td>\n",
       "      <td>0.133127</td>\n",
       "      <td>64.6863</td>\n",
       "      <td>97.7453</td>\n",
       "      <td>Pärnu maakond</td>\n",
       "      <td>0.443136</td>\n",
       "      <td>1.91185</td>\n",
       "      <td>58.3195</td>\n",
       "      <td>24.3026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_0000408</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.00106841</td>\n",
       "      <td>593.344</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>9.056080e-09</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.068425e-22</td>\n",
       "      <td>1978-02-16</td>\n",
       "      <td>1978.128239</td>\n",
       "      <td>[1968.4891921919855, 1983.692453807525]</td>\n",
       "      <td>[52.51080780801453, 37.307546192475]</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5978.962136</td>\n",
       "      <td>593.343750</td>\n",
       "      <td>1978.128239</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>8.72422e-08</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1</td>\n",
       "      <td>2.11807e-07</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.01253e-06</td>\n",
       "      <td>Peru</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.7239e-05</td>\n",
       "      <td>-6.86997</td>\n",
       "      <td>-75.0459</td>\n",
       "      <td>La Libertad</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>5.21332e-05</td>\n",
       "      <td>-8</td>\n",
       "      <td>-78.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_0000411</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.00106841</td>\n",
       "      <td>594.938</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.510844e-09</td>\n",
       "      <td>0.829370</td>\n",
       "      <td>1.068425e-22</td>\n",
       "      <td>1978-05-16</td>\n",
       "      <td>1978.371715</td>\n",
       "      <td>[1968.5283949205668, 1984.0160167471524]</td>\n",
       "      <td>[52.471605079433175, 36.983983252847565]</td>\n",
       "      <td>0.243476</td>\n",
       "      <td>5979.205612</td>\n",
       "      <td>594.937500</td>\n",
       "      <td>1978.371715</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>8.71566e-08</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1</td>\n",
       "      <td>2.11589e-07</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.01137e-06</td>\n",
       "      <td>Peru</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.71807e-05</td>\n",
       "      <td>-6.86997</td>\n",
       "      <td>-75.0459</td>\n",
       "      <td>La Libertad</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>3.43986e-05</td>\n",
       "      <td>-8</td>\n",
       "      <td>-78.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_0000412</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.00106841</td>\n",
       "      <td>595.875</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.947457e-09</td>\n",
       "      <td>0.546313</td>\n",
       "      <td>1.068425e-22</td>\n",
       "      <td>1986-11-09</td>\n",
       "      <td>1986.856321</td>\n",
       "      <td>[1968.5469333253366, 2000.6367394245203]</td>\n",
       "      <td>[52.45306667466343, 20.36326057547967]</td>\n",
       "      <td>8.484607</td>\n",
       "      <td>5987.690219</td>\n",
       "      <td>595.875000</td>\n",
       "      <td>1986.856321</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>8.71565e-08</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1</td>\n",
       "      <td>2.11588e-07</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.01136e-06</td>\n",
       "      <td>Peru</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.71805e-05</td>\n",
       "      <td>-6.86997</td>\n",
       "      <td>-75.0459</td>\n",
       "      <td>La Libertad</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>3.41987e-05</td>\n",
       "      <td>-8</td>\n",
       "      <td>-78.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_0000413</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.00106841</td>\n",
       "      <td>596.75</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>8.903422e-09</td>\n",
       "      <td>0.983143</td>\n",
       "      <td>1.068425e-22</td>\n",
       "      <td>1986-11-09</td>\n",
       "      <td>1986.856321</td>\n",
       "      <td>[1968.6420434068189, 2000.4761805051378]</td>\n",
       "      <td>[52.35795659318114, 20.52381949486221]</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5987.690219</td>\n",
       "      <td>596.750000</td>\n",
       "      <td>1986.856321</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>8.71564e-08</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1</td>\n",
       "      <td>2.11588e-07</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.01136e-06</td>\n",
       "      <td>Peru</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.71802e-05</td>\n",
       "      <td>-6.86997</td>\n",
       "      <td>-75.0459</td>\n",
       "      <td>La Libertad</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>3.41946e-05</td>\n",
       "      <td>-8</td>\n",
       "      <td>-78.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NODE_0000414</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.00106841</td>\n",
       "      <td>597.5</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>9.056080e-09</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.068425e-22</td>\n",
       "      <td>1986-11-09</td>\n",
       "      <td>1986.856321</td>\n",
       "      <td>[1968.633783825002, 2000.397427341175]</td>\n",
       "      <td>[52.366216174998044, 20.60257265882501]</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5987.690219</td>\n",
       "      <td>597.500000</td>\n",
       "      <td>1986.856321</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>8.70708e-08</td>\n",
       "      <td>1.ORI</td>\n",
       "      <td>1</td>\n",
       "      <td>2.1137e-07</td>\n",
       "      <td>1.ORI1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0102e-06</td>\n",
       "      <td>Peru</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.71221e-05</td>\n",
       "      <td>-6.86997</td>\n",
       "      <td>-75.0459</td>\n",
       "      <td>La Libertad</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>3.39832e-05</td>\n",
       "      <td>-8</td>\n",
       "      <td>-78.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1050 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               UFboot   sCF Branch_Length Branch_Support_Color     coord_x  \\\n",
       "Name                                                                         \n",
       "NODE0             100    NA             0                black       0.001   \n",
       "SAMEA3541827       NA    NA    3.8408e-06                 grey  0.00100322   \n",
       "NODE1             100  84.9   1.34224e-05                black  0.00100796   \n",
       "SAMEA104233046     NA    NA     2.233e-06                 grey  0.00101011   \n",
       "NODE2             100  94.7   1.34762e-05                black  0.00102217   \n",
       "...               ...   ...           ...                  ...         ...   \n",
       "NODE_0000408       NA    NA            NA                   NA  0.00106841   \n",
       "NODE_0000411       NA    NA            NA                   NA  0.00106841   \n",
       "NODE_0000412       NA    NA            NA                   NA  0.00106841   \n",
       "NODE_0000413       NA    NA            NA                   NA  0.00106841   \n",
       "NODE_0000414       NA    NA            NA                   NA  0.00106841   \n",
       "\n",
       "                coord_y      Strain           Date         DateBP    Country  \\\n",
       "Name                                                                           \n",
       "NODE0            15.542          NA             NA             NA         NA   \n",
       "SAMEA3541827          1     RISE509  [-2876:-2677]  [-4897:-4698]     Russia   \n",
       "NODE1           3.15625          NA             NA             NA         NA   \n",
       "SAMEA104233046        2  Gyvakarai1  [-2621:-2472]  [-4642:-4493]  Lithuania   \n",
       "NODE2            4.3125          NA             NA             NA         NA   \n",
       "...                 ...         ...            ...            ...        ...   \n",
       "NODE_0000408    593.344          NA             NA             NA         NA   \n",
       "NODE_0000411    594.938          NA             NA             NA         NA   \n",
       "NODE_0000412    595.875          NA             NA             NA         NA   \n",
       "NODE_0000413     596.75          NA             NA             NA         NA   \n",
       "NODE_0000414      597.5          NA             NA             NA         NA   \n",
       "\n",
       "                        Province CountryLat CountryLon ProvinceLat  \\\n",
       "Name                                                                 \n",
       "NODE0                         NA         NA         NA          NA   \n",
       "SAMEA3541827    Krasnoyarsk Krai    64.6863    97.7453     63.3234   \n",
       "NODE1                         NA         NA         NA          NA   \n",
       "SAMEA104233046  Panevezys County      55.35      23.75     55.9156   \n",
       "NODE2                         NA         NA         NA          NA   \n",
       "...                          ...        ...        ...         ...   \n",
       "NODE_0000408                  NA         NA         NA          NA   \n",
       "NODE_0000411                  NA         NA         NA          NA   \n",
       "NODE_0000412                  NA         NA         NA          NA   \n",
       "NODE_0000413                  NA         NA         NA          NA   \n",
       "NODE_0000414                  NA         NA         NA          NA   \n",
       "\n",
       "               ProvinceLon      Biovar Branch_Major Branch_Minor  \\\n",
       "Name                                                               \n",
       "NODE0                   NA          NA           NA           NA   \n",
       "SAMEA3541827        97.098  Bronze Age        0.PRE       0.PRE2   \n",
       "NODE1                   NA          NA           NA           NA   \n",
       "SAMEA104233046     25.0312  Bronze Age        0.PRE       0.PRE1   \n",
       "NODE2                   NA          NA           NA           NA   \n",
       "...                    ...         ...          ...          ...   \n",
       "NODE_0000408            NA          NA           NA           NA   \n",
       "NODE_0000411            NA          NA           NA           NA   \n",
       "NODE_0000412            NA          NA           NA           NA   \n",
       "NODE_0000413            NA          NA           NA           NA   \n",
       "NODE_0000414            NA          NA           NA           NA   \n",
       "\n",
       "                     BioSample   BioSampleComment Branch_Number  \\\n",
       "Name                                                              \n",
       "NODE0                       NA                 NA            NA   \n",
       "SAMEA3541827      SAMEA3541827  KEEP: SRA Ancient             0   \n",
       "NODE1                       NA                 NA            NA   \n",
       "SAMEA104233046  SAMEA104233046  KEEP: SRA Ancient             0   \n",
       "NODE2                       NA                 NA            NA   \n",
       "...                        ...                ...           ...   \n",
       "NODE_0000408                NA                 NA            NA   \n",
       "NODE_0000411                NA                 NA            NA   \n",
       "NODE_0000412                NA                 NA            NA   \n",
       "NODE_0000413                NA                 NA            NA   \n",
       "NODE_0000414                NA                 NA            NA   \n",
       "\n",
       "                timetree_rate  timetree_rate_fold_change  \\\n",
       "Name                                                       \n",
       "NODE0            9.056080e-09                   1.000000   \n",
       "SAMEA3541827     2.775061e-09                   0.306431   \n",
       "NODE1            1.322885e-08                   1.460770   \n",
       "SAMEA104233046   2.489749e-09                   0.274926   \n",
       "NODE2            1.959170e-08                   2.163375   \n",
       "...                       ...                        ...   \n",
       "NODE_0000408     9.056080e-09                   1.000000   \n",
       "NODE_0000411     7.510844e-09                   0.829370   \n",
       "NODE_0000412     4.947457e-09                   0.546313   \n",
       "NODE_0000413     8.903422e-09                   0.983143   \n",
       "NODE_0000414     9.056080e-09                   1.000000   \n",
       "\n",
       "                timetree_mutation_length timetree_raw_date  timetree_num_date  \\\n",
       "Name                                                                            \n",
       "NODE0                       1.000000e-03       -4001-04-08       -4000.733897   \n",
       "SAMEA3541827                3.224758e-06       -2879-06-08       -2878.564956   \n",
       "NODE1                       7.959900e-06       -3372-09-28       -3371.257753   \n",
       "SAMEA104233046              2.150063e-06       -2600-02-18       -2599.866006   \n",
       "NODE2                       1.421271e-05       -2644-06-12       -2643.555870   \n",
       "...                                  ...               ...                ...   \n",
       "NODE_0000408                1.068425e-22        1978-02-16        1978.128239   \n",
       "NODE_0000411                1.068425e-22        1978-05-16        1978.371715   \n",
       "NODE_0000412                1.068425e-22        1986-11-09        1986.856321   \n",
       "NODE_0000413                1.068425e-22        1986-11-09        1986.856321   \n",
       "NODE_0000414                1.068425e-22        1986-11-09        1986.856321   \n",
       "\n",
       "                              timetree_num_date_confidence  \\\n",
       "Name                                                         \n",
       "NODE0           [-4028.4379593546546, -3773.0741326132847]   \n",
       "SAMEA3541827                            [-2876.0, -2677.0]   \n",
       "NODE1             [-3386.8297780304565, -3135.38189068595]   \n",
       "SAMEA104233046               [-2621.0, -2480.907503502162]   \n",
       "NODE2            [-2758.1911839442546, -2599.428548849007]   \n",
       "...                                                    ...   \n",
       "NODE_0000408       [1968.4891921919855, 1983.692453807525]   \n",
       "NODE_0000411      [1968.5283949205668, 1984.0160167471524]   \n",
       "NODE_0000412      [1968.5469333253366, 2000.6367394245203]   \n",
       "NODE_0000413      [1968.6420434068189, 2000.4761805051378]   \n",
       "NODE_0000414        [1968.633783825002, 2000.397427341175]   \n",
       "\n",
       "                                   timetree_num_date_bar  \\\n",
       "Name                                                       \n",
       "NODE0             [6049.437959354655, 5794.074132613285]   \n",
       "SAMEA3541827                            [4897.0, 4698.0]   \n",
       "NODE1            [5407.8297780304565, 5156.381890685951]   \n",
       "SAMEA104233046              [4642.0, 4501.9075035021615]   \n",
       "NODE2             [4779.191183944255, 4620.428548849007]   \n",
       "...                                                  ...   \n",
       "NODE_0000408        [52.51080780801453, 37.307546192475]   \n",
       "NODE_0000411    [52.471605079433175, 36.983983252847565]   \n",
       "NODE_0000412      [52.45306667466343, 20.36326057547967]   \n",
       "NODE_0000413      [52.35795659318114, 20.52381949486221]   \n",
       "NODE_0000414     [52.366216174998044, 20.60257265882501]   \n",
       "\n",
       "                timetree_clock_length  timetree_coord_x  timetree_coord_y  \\\n",
       "Name                                                                        \n",
       "NODE0                        0.100000          0.100000         15.541999   \n",
       "SAMEA3541827              1122.168941       1122.268941          1.000000   \n",
       "NODE1                      629.476144        629.576144          3.156250   \n",
       "SAMEA104233046             771.391746       1400.967891          2.000000   \n",
       "NODE2                      727.701883       1357.278028          4.312500   \n",
       "...                               ...               ...               ...   \n",
       "NODE_0000408                 0.000000       5978.962136        593.343750   \n",
       "NODE_0000411                 0.243476       5979.205612        594.937500   \n",
       "NODE_0000412                 8.484607       5987.690219        595.875000   \n",
       "NODE_0000413                 0.000000       5987.690219        596.750000   \n",
       "NODE_0000414                 0.000000       5987.690219        597.500000   \n",
       "\n",
       "                timetree_reg_x  timetree_reg_y  timetree_reg_bad  \\\n",
       "Name                                                               \n",
       "NODE0             -4000.733897        0.000000             False   \n",
       "SAMEA3541827      -2776.500000        0.000003             False   \n",
       "NODE1             -3371.257753        0.000008             False   \n",
       "SAMEA104233046    -2546.500000        0.000010             False   \n",
       "NODE2             -2643.555870        0.000022             False   \n",
       "...                        ...             ...               ...   \n",
       "NODE_0000408       1978.128239        0.000068             False   \n",
       "NODE_0000411       1978.371715        0.000068             False   \n",
       "NODE_0000412       1986.856321        0.000068             False   \n",
       "NODE_0000413       1986.856321        0.000068             False   \n",
       "NODE_0000414       1986.856321        0.000068             False   \n",
       "\n",
       "               Mugration_Branch_Number Mugration_Branch_Number_Confidence  \\\n",
       "Name                                                                        \n",
       "NODE0                                0                                  1   \n",
       "SAMEA3541827                         0                                  1   \n",
       "NODE1                                0                                  1   \n",
       "SAMEA104233046                       0                                  1   \n",
       "NODE2                                0                                  1   \n",
       "...                                ...                                ...   \n",
       "NODE_0000408                         1                                  1   \n",
       "NODE_0000411                         1                                  1   \n",
       "NODE_0000412                         1                                  1   \n",
       "NODE_0000413                         1                                  1   \n",
       "NODE_0000414                         1                                  1   \n",
       "\n",
       "               Mugration_Branch_Number_Entropy Mugration_Branch_Major  \\\n",
       "Name                                                                    \n",
       "NODE0                              8.85652e-08                  0.PRE   \n",
       "SAMEA3541827                      -1.00009e-12                  0.PRE   \n",
       "NODE1                              8.85651e-08                  0.PRE   \n",
       "SAMEA104233046                    -1.00009e-12                  0.PRE   \n",
       "NODE2                              8.85652e-08                  0.PRE   \n",
       "...                                        ...                    ...   \n",
       "NODE_0000408                       8.72422e-08                  1.ORI   \n",
       "NODE_0000411                       8.71566e-08                  1.ORI   \n",
       "NODE_0000412                       8.71565e-08                  1.ORI   \n",
       "NODE_0000413                       8.71564e-08                  1.ORI   \n",
       "NODE_0000414                       8.70708e-08                  1.ORI   \n",
       "\n",
       "               Mugration_Branch_Major_Confidence  \\\n",
       "Name                                               \n",
       "NODE0                                   0.998709   \n",
       "SAMEA3541827                                   1   \n",
       "NODE1                                   0.999998   \n",
       "SAMEA104233046                                 1   \n",
       "NODE2                                          1   \n",
       "...                                          ...   \n",
       "NODE_0000408                                   1   \n",
       "NODE_0000411                                   1   \n",
       "NODE_0000412                                   1   \n",
       "NODE_0000413                                   1   \n",
       "NODE_0000414                                   1   \n",
       "\n",
       "               Mugration_Branch_Major_Entropy Mugration_Branch_Minor  \\\n",
       "Name                                                                   \n",
       "NODE0                              0.00991765                 0.PRE2   \n",
       "SAMEA3541827                     -9.99867e-13                 0.PRE2   \n",
       "NODE1                             2.40643e-05                 0.PRE1   \n",
       "SAMEA104233046                   -1.00009e-12                 0.PRE1   \n",
       "NODE2                             2.20342e-07                 0.PRE1   \n",
       "...                                       ...                    ...   \n",
       "NODE_0000408                      2.11807e-07                 1.ORI1   \n",
       "NODE_0000411                      2.11589e-07                 1.ORI1   \n",
       "NODE_0000412                      2.11588e-07                 1.ORI1   \n",
       "NODE_0000413                      2.11588e-07                 1.ORI1   \n",
       "NODE_0000414                       2.1137e-07                 1.ORI1   \n",
       "\n",
       "               Mugration_Branch_Minor_Confidence  \\\n",
       "Name                                               \n",
       "NODE0                                   0.392407   \n",
       "SAMEA3541827                                   1   \n",
       "NODE1                                   0.999201   \n",
       "SAMEA104233046                                 1   \n",
       "NODE2                                   0.999999   \n",
       "...                                          ...   \n",
       "NODE_0000408                                   1   \n",
       "NODE_0000411                                   1   \n",
       "NODE_0000412                                   1   \n",
       "NODE_0000413                                   1   \n",
       "NODE_0000414                                   1   \n",
       "\n",
       "               Mugration_Branch_Minor_Entropy Mugration_Country  \\\n",
       "Name                                                              \n",
       "NODE0                                 1.32976            Russia   \n",
       "SAMEA3541827                     -1.00009e-12            Russia   \n",
       "NODE1                              0.00737495            Russia   \n",
       "SAMEA104233046                   -1.00009e-12         Lithuania   \n",
       "NODE2                             1.74031e-05            Russia   \n",
       "...                                       ...               ...   \n",
       "NODE_0000408                      1.01253e-06              Peru   \n",
       "NODE_0000411                      1.01137e-06              Peru   \n",
       "NODE_0000412                      1.01136e-06              Peru   \n",
       "NODE_0000413                      1.01136e-06              Peru   \n",
       "NODE_0000414                       1.0102e-06              Peru   \n",
       "\n",
       "               Mugration_Country_Confidence Mugration_Country_Entropy  \\\n",
       "Name                                                                    \n",
       "NODE0                              0.993164                 0.0442662   \n",
       "SAMEA3541827                              1              -1.00009e-12   \n",
       "NODE1                              0.979419                  0.127114   \n",
       "SAMEA104233046                            1              -1.00009e-12   \n",
       "NODE2                              0.978827                  0.133127   \n",
       "...                                     ...                       ...   \n",
       "NODE_0000408                       0.999999                1.7239e-05   \n",
       "NODE_0000411                       0.999999               1.71807e-05   \n",
       "NODE_0000412                       0.999999               1.71805e-05   \n",
       "NODE_0000413                       0.999999               1.71802e-05   \n",
       "NODE_0000414                       0.999999               1.71221e-05   \n",
       "\n",
       "               Mugration_Country_Lat Mugration_Country_Lon Mugration_Province  \\\n",
       "Name                                                                            \n",
       "NODE0                        64.6863               97.7453   Krasnoyarsk Krai   \n",
       "SAMEA3541827                 64.6863               97.7453   Krasnoyarsk Krai   \n",
       "NODE1                        64.6863               97.7453   Panevezys County   \n",
       "SAMEA104233046                 55.35                 23.75   Panevezys County   \n",
       "NODE2                        64.6863               97.7453      Pärnu maakond   \n",
       "...                              ...                   ...                ...   \n",
       "NODE_0000408                -6.86997              -75.0459        La Libertad   \n",
       "NODE_0000411                -6.86997              -75.0459        La Libertad   \n",
       "NODE_0000412                -6.86997              -75.0459        La Libertad   \n",
       "NODE_0000413                -6.86997              -75.0459        La Libertad   \n",
       "NODE_0000414                -6.86997              -75.0459        La Libertad   \n",
       "\n",
       "               Mugration_Province_Confidence Mugration_Province_Entropy  \\\n",
       "Name                                                                      \n",
       "NODE0                               0.382595                    1.64309   \n",
       "SAMEA3541827                               1               -1.00009e-12   \n",
       "NODE1                               0.443092                    1.91578   \n",
       "SAMEA104233046                             1               -1.00009e-12   \n",
       "NODE2                               0.443136                    1.91185   \n",
       "...                                      ...                        ...   \n",
       "NODE_0000408                        0.999997                5.21332e-05   \n",
       "NODE_0000411                        0.999998                3.43986e-05   \n",
       "NODE_0000412                        0.999998                3.41987e-05   \n",
       "NODE_0000413                        0.999998                3.41946e-05   \n",
       "NODE_0000414                        0.999998                3.39832e-05   \n",
       "\n",
       "               Mugration_Province_Lat Mugration_Province_Lon  \n",
       "Name                                                          \n",
       "NODE0                         63.3234                 97.098  \n",
       "SAMEA3541827                  63.3234                 97.098  \n",
       "NODE1                         55.9156                25.0312  \n",
       "SAMEA104233046                55.9156                25.0312  \n",
       "NODE2                         58.3195                24.3026  \n",
       "...                               ...                    ...  \n",
       "NODE_0000408                       -8                  -78.5  \n",
       "NODE_0000411                       -8                  -78.5  \n",
       "NODE_0000412                       -8                  -78.5  \n",
       "NODE_0000413                       -8                  -78.5  \n",
       "NODE_0000414                       -8                  -78.5  \n",
       "\n",
       "[1050 rows x 53 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for attr in ATTRIBUTE_LIST:\n",
    "    # Initialize empty values for the new mugration attribute and its confidence\n",
    "    tree_df[\"Mugration_\" + attr] = [NO_DATA_CHAR for row in range(0,len(tree_df))]\n",
    "    tree_df[\"Mugration_\" + attr + \"_Confidence\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]    \n",
    "    tree_df[\"Mugration_\" + attr + \"_Entropy\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]  \n",
    "    \n",
    "    tree_mug = mug_dict[attr][\"tree_mug\"].tree\n",
    "    unique_states = mug_dict[attr][\"unique_states\"]\n",
    "    \n",
    "    # If this attribute is associated with lat,lon\n",
    "    if attr + \"Lat\" in tree_df.columns:\n",
    "        tree_df[\"Mugration_\" + attr + \"_Lat\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]\n",
    "        tree_df[\"Mugration_\" + attr + \"_Lon\"] = [NO_DATA_CHAR for row in range(0,len(tree_df))]       \n",
    "\n",
    "    # Iterate through the nodes in the tree\n",
    "    for c in tree_mug.find_clades():\n",
    "        # Store the name and confidence for the state with the most support\n",
    "        state_max_name = mug_dict[attr][\"letter_to_state\"][c.cseq[0]]\n",
    "        state_conf_list = c.marginal_profile[0]\n",
    "        state_max_conf = max(state_conf_list)\n",
    "        c.other = []\n",
    "\n",
    "        # Store all the states and confidence values\n",
    "        for state_name,state_conf in zip(unique_states, state_conf_list):\n",
    "            attr_other = Phylo.PhyloXML.Other(tag=state_name, value=\"{:0.4f}\".format(state_conf), namespace=attr)\n",
    "            c.other.append(attr_other)\n",
    "\n",
    "        # Store the Maximum assigned value\n",
    "        attr_other = Phylo.PhyloXML.Other(tag=state_max_name, value=\"{:0.4f}\".format(state_max_conf), namespace=attr + \"_assign\")\n",
    "        c.other.append(attr_other)\n",
    "    \n",
    "        if c.name in tree_df.index:\n",
    "            tree_df.at[c.name, \"Mugration_\" + attr] = state_max_name\n",
    "            tree_df.at[c.name, \"Mugration_\" + attr + \"_Confidence\"] = state_max_conf\n",
    "            \n",
    "            prob_dist = c.marginal_profile[0]\n",
    "            S = -np.sum(prob_dist * np.log(prob_dist + MUG_TINY))\n",
    "            tree_df.at[c.name, \"Mugration_\" + attr + \"_Entropy\"] = S\n",
    "        \n",
    "        # Add mugration lat lon\n",
    "        if attr + \"Lat\" in tree_df.columns:\n",
    "            # Check for lat and lon\n",
    "            c_geo = tree_df[\"Mugration_\" + attr][c.name]\n",
    "            c_geo_match = latlon_df[(latlon_df[\"Geo\"] == attr.lower()) &\n",
    "                            (latlon_df[\"Name\"] == c_geo)\n",
    "                           ]\n",
    "            c_geo_lat = c_geo_match[\"Lat\"].values[0]\n",
    "            c_geo_lon = c_geo_match[\"Lon\"].values[0]\n",
    "            tree_df.at[c.name, \"Mugration_\" + attr + \"_Lat\"] = c_geo_lat\n",
    "            tree_df.at[c.name, \"Mugration_\" + attr + \"_Lon\"] = c_geo_lon\n",
    "            \n",
    "tree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add Metadata as Comments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_to_comment(tt.tree, tree_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 3. Plot confidence boxplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'for attr in ATTRIBUTE_LIST:\\n    fig, (ax1, ax2) = plt.subplots(2, \\n                               sharex=False, \\n                               gridspec_kw={\\'hspace\\': 0},\\n                               figsize=figsize, \\n                               dpi=dpi,\\n                               constrained_layout=True,\\n                              )\\n    \\n    # --------------------------\\n    # Axis 1 - Number of tips per state\\n    # Exclude samples with no attribute recorded\\n    data = tree_df[tree_df[attr] != NO_DATA_CHAR]\\n    label_order = list(data[attr].value_counts().index)\\n    label_order = [lab for lab in label_order if lab != \"nan\"] + [\"nan\"]\\n    \\n    sns.countplot(data=data, \\n                  x=attr, \\n                  #color=\"blue\", \\n                  edgecolor=\"black\",\\n                  ax=ax1, \\n                  order=label_order\\n                 )\\n    plt.setp(ax1.get_xticklabels(), visible=False)\\n    xticklabels = [item.get_text() for item in ax1.get_xticklabels()]\\n    ax1.set_xticklabels(xticklabels, rotation = 90, ha=\"right\")\\n    ax1.set_xlabel(\"\")\\n    ax1.set_ylabel(\"Number of Samples (Tips)\")\\n    ax1.set_xlim(-1,len(label_order)) \\n    ax1.set_title(attr.replace(\"_\",\" \"))\\n    \\n    # --------------------------\\n    # Axis 2 - Mugration Confidence\\n    # Exclude nodes that are terminals (ie. branch support is grey)\\n    data = tree_df[tree_df[\"Branch_Support_Color\"] != \"grey\"]\\n    \\n    # Fix typing\\n    if attr == \"Branch_Number\":\\n        label_order = [str(math.ceil(lab)) for lab in label_order if lab != \"nan\"] + [\"nan\"]\\n\\n    # Customize outlier style\\n    flierprops = dict(marker=\\'o\\', markerfacecolor=\\'white\\', markersize=2,\\n                      linestyle=\\'none\\', markeredgecolor=\\'black\\')\\n    # Create a boxplot\\n    sns.boxplot(data=data, \\n                  x=\"Mugration_\" + attr, \\n                  y=\"Mugration_\" + attr + \"_Confidence\",\\n                  #color=\"blue\", \\n                  ax=ax2, \\n                  order=label_order,\\n                  flierprops=flierprops)\\n    #plt.setp(ax2.get_xticklabels(), visible=False)\\n    xticklabels = [item.get_text() for item in ax2.get_xticklabels()]\\n    ax2.set_xticklabels(xticklabels, rotation = 90, ha=\"center\")\\n    #ax2.axhline(y=MUG_CONF_THRESH, color=THRESH_COL, linewidth=0.5, linestyle=\\'--\\')\\n    ax2.set_xlabel(\"\")\\n    ax2.set_ylabel(\"Mugration Confidence (Internal Nodes)\")\\n    ax2.set_xlim(-1,len(label_order))\\n    ax2.set_ylim(0, 1.1)\\n    \\n    fig.suptitle(\"Sampling Distribution and Mugration Analysis Confidence\")\\n    \\n    out_path = os.path.join(outdir, SCRIPT_NAME + \"_boxplot_{}.{}\".format(attr.lower(), FMT)) \\n    plt.savefig(out_path, \\n            dpi=dpi, \\n            bbox_inches = \"tight\")'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"for attr in ATTRIBUTE_LIST:\n",
    "    fig, (ax1, ax2) = plt.subplots(2, \n",
    "                               sharex=False, \n",
    "                               gridspec_kw={'hspace': 0},\n",
    "                               figsize=figsize, \n",
    "                               dpi=dpi,\n",
    "                               constrained_layout=True,\n",
    "                              )\n",
    "    \n",
    "    # --------------------------\n",
    "    # Axis 1 - Number of tips per state\n",
    "    # Exclude samples with no attribute recorded\n",
    "    data = tree_df[tree_df[attr] != NO_DATA_CHAR]\n",
    "    label_order = list(data[attr].value_counts().index)\n",
    "    label_order = [lab for lab in label_order if lab != \"nan\"] + [\"nan\"]\n",
    "    \n",
    "    sns.countplot(data=data, \n",
    "                  x=attr, \n",
    "                  #color=\"blue\", \n",
    "                  edgecolor=\"black\",\n",
    "                  ax=ax1, \n",
    "                  order=label_order\n",
    "                 )\n",
    "    plt.setp(ax1.get_xticklabels(), visible=False)\n",
    "    xticklabels = [item.get_text() for item in ax1.get_xticklabels()]\n",
    "    ax1.set_xticklabels(xticklabels, rotation = 90, ha=\"right\")\n",
    "    ax1.set_xlabel(\"\")\n",
    "    ax1.set_ylabel(\"Number of Samples (Tips)\")\n",
    "    ax1.set_xlim(-1,len(label_order)) \n",
    "    ax1.set_title(attr.replace(\"_\",\" \"))\n",
    "    \n",
    "    # --------------------------\n",
    "    # Axis 2 - Mugration Confidence\n",
    "    # Exclude nodes that are terminals (ie. branch support is grey)\n",
    "    data = tree_df[tree_df[\"Branch_Support_Color\"] != \"grey\"]\n",
    "    \n",
    "    # Fix typing\n",
    "    if attr == \"Branch_Number\":\n",
    "        label_order = [str(math.ceil(lab)) for lab in label_order if lab != \"nan\"] + [\"nan\"]\n",
    "\n",
    "    # Customize outlier style\n",
    "    flierprops = dict(marker='o', markerfacecolor='white', markersize=2,\n",
    "                      linestyle='none', markeredgecolor='black')\n",
    "    # Create a boxplot\n",
    "    sns.boxplot(data=data, \n",
    "                  x=\"Mugration_\" + attr, \n",
    "                  y=\"Mugration_\" + attr + \"_Confidence\",\n",
    "                  #color=\"blue\", \n",
    "                  ax=ax2, \n",
    "                  order=label_order,\n",
    "                  flierprops=flierprops)\n",
    "    #plt.setp(ax2.get_xticklabels(), visible=False)\n",
    "    xticklabels = [item.get_text() for item in ax2.get_xticklabels()]\n",
    "    ax2.set_xticklabels(xticklabels, rotation = 90, ha=\"center\")\n",
    "    #ax2.axhline(y=MUG_CONF_THRESH, color=THRESH_COL, linewidth=0.5, linestyle='--')\n",
    "    ax2.set_xlabel(\"\")\n",
    "    ax2.set_ylabel(\"Mugration Confidence (Internal Nodes)\")\n",
    "    ax2.set_xlim(-1,len(label_order))\n",
    "    ax2.set_ylim(0, 1.1)\n",
    "    \n",
    "    fig.suptitle(\"Sampling Distribution and Mugration Analysis Confidence\")\n",
    "    \n",
    "    out_path = os.path.join(outdir, SCRIPT_NAME + \"_boxplot_{}.{}\".format(attr.lower(), FMT)) \n",
    "    plt.savefig(out_path, \n",
    "            dpi=dpi, \n",
    "            bbox_inches = \"tight\")\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 4. Plot colored trees"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Color branches on mug/div tree according to state"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'out_path_colors = os.path.join(outdir, SCRIPT_NAME + \"_colors.tsv\")\\nfile_colors = open(out_path_colors, \"w\")\\n\\nfor attr in ATTRIBUTE_LIST:  \\n    if attr != \"Branch_Major\" and attr != \"Branch_Number\":\\n        continue\\n    \\n    # --------------------------------------------   \\n    # Canvas\\n    fig, ax1 = plt.subplots(1, dpi=dpi, figsize=figsize, constrained_layout=True,)\\n\\n    # --------------------------------------------\\n    # Color Tree\\n    hex_dict = color_tree(tree=tree_div, \\n                          df=tree_df, \\n                          attribute=\"Mugration_\" + attr,\\n                          attribute_confidence=\"Mugration_\" + attr + \"_Confidence\",\\n                          threshold_confidence=MUG_CONF_THRESH,                          \\n                          color_pal=CONT_COLOR_PAL)\\n    \\n    # Add the hex color dict to the dict\\n    mug_dict[attr][\"hex_color\"] = hex_dict  \\n\\n    # Write to color file\\n    for state,color in hex_dict.items():\\n        file_colors.write(attr.lower() + \"\\t\" + state + \"\\t\" + color + \"\\n\")    \\n        \\n    # Draw tree\\n    Phylo.draw(tree_div,\\n               axes=ax1, \\n               show_confidence=False, \\n               label_func = lambda x:\\'\\', \\n               do_show=False)\\n    # --------------------------------------------\\n    # Draw tips\\n    colors = [mug_dict[attr][\"hex_color\"][state] for state in tree_df[\"Mugration_\" + attr]]    \\n    ax1.scatter(data=tree_df, \\n                x=\"coord_x\", \\n                y=\"coord_y\", \\n                s=0.5, \\n                c=colors,\\n               )\\n    # --------------------------------------------\\n    # Ticks\\n    x_buffer = max(tree_df[\"coord_x\"]) * 0.01\\n    y_buffer = math.ceil(len(tree_div.get_terminals()) * 0.01)   \\n    # --------------------------------------------\\n    # Limits\\n    ax1.set_xlim(0 - x_buffer, max(tree_df[\"coord_x\"]) + x_buffer)    \\n    ax1.set_ylim(len(tree_div.get_terminals()) + y_buffer, 0 - y_buffer)\\n    ax1.set_yticks([])    \\n    # --------------------------------------------\\n    # Labels\\n    ax1.set_xlabel(\"Branch Length\")\\n    ax1.set_ylabel(\"\")\\n    fig.suptitle(\"{} Mugration\".format(attr.replace(\"_\",\" \")))    \\n    # --------------------------------------------\\n    # Legend\\n    legend_elements = [patches.Patch(facecolor=value, edgecolor=value,) for value in mug_dict[attr][\"hex_color\"].values()]\\n    legend_labels = list(mug_dict[attr][\"hex_color\"].keys())\\n    ax1.legend(legend_elements, \\n              legend_labels,\\n              bbox_to_anchor=(1.0, 1.0), \\n              loc=\\'upper left\\',\\n              frameon=False,\\n             )\\n    # --------------------------------------------\\n    # Export\\n    out_path = os.path.join(outdir, SCRIPT_NAME + \"_tree_{}.{}\".format(attr.lower(), FMT))  \\n    plt.savefig(out_path, dpi=dpi, bbox_inches = \"tight\")\\n    \\n    # Save Tree!\\n    #out_path_dill_tree = os.path.join(outdir,  SCRIPT_NAME + \"_{}.phylo.obj\".format(attr.lower()))\\n    #with open(out_path_dill_tree,\"wb\") as outfile:\\n    #    dill.dump(tree_div, outfile)\\n    \\nfile_colors.close()'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"out_path_colors = os.path.join(outdir, SCRIPT_NAME + \"_colors.tsv\")\n",
    "file_colors = open(out_path_colors, \"w\")\n",
    "\n",
    "for attr in ATTRIBUTE_LIST:  \n",
    "    if attr != \"Branch_Major\" and attr != \"Branch_Number\":\n",
    "        continue\n",
    "    \n",
    "    # --------------------------------------------   \n",
    "    # Canvas\n",
    "    fig, ax1 = plt.subplots(1, dpi=dpi, figsize=figsize, constrained_layout=True,)\n",
    "\n",
    "    # --------------------------------------------\n",
    "    # Color Tree\n",
    "    hex_dict = color_tree(tree=tree_div, \n",
    "                          df=tree_df, \n",
    "                          attribute=\"Mugration_\" + attr,\n",
    "                          attribute_confidence=\"Mugration_\" + attr + \"_Confidence\",\n",
    "                          threshold_confidence=MUG_CONF_THRESH,                          \n",
    "                          color_pal=CONT_COLOR_PAL)\n",
    "    \n",
    "    # Add the hex color dict to the dict\n",
    "    mug_dict[attr][\"hex_color\"] = hex_dict  \n",
    "\n",
    "    # Write to color file\n",
    "    for state,color in hex_dict.items():\n",
    "        file_colors.write(attr.lower() + \"\\t\" + state + \"\\t\" + color + \"\\n\")    \n",
    "        \n",
    "    # Draw tree\n",
    "    Phylo.draw(tree_div,\n",
    "               axes=ax1, \n",
    "               show_confidence=False, \n",
    "               label_func = lambda x:'', \n",
    "               do_show=False)\n",
    "    # --------------------------------------------\n",
    "    # Draw tips\n",
    "    colors = [mug_dict[attr][\"hex_color\"][state] for state in tree_df[\"Mugration_\" + attr]]    \n",
    "    ax1.scatter(data=tree_df, \n",
    "                x=\"coord_x\", \n",
    "                y=\"coord_y\", \n",
    "                s=0.5, \n",
    "                c=colors,\n",
    "               )\n",
    "    # --------------------------------------------\n",
    "    # Ticks\n",
    "    x_buffer = max(tree_df[\"coord_x\"]) * 0.01\n",
    "    y_buffer = math.ceil(len(tree_div.get_terminals()) * 0.01)   \n",
    "    # --------------------------------------------\n",
    "    # Limits\n",
    "    ax1.set_xlim(0 - x_buffer, max(tree_df[\"coord_x\"]) + x_buffer)    \n",
    "    ax1.set_ylim(len(tree_div.get_terminals()) + y_buffer, 0 - y_buffer)\n",
    "    ax1.set_yticks([])    \n",
    "    # --------------------------------------------\n",
    "    # Labels\n",
    "    ax1.set_xlabel(\"Branch Length\")\n",
    "    ax1.set_ylabel(\"\")\n",
    "    fig.suptitle(\"{} Mugration\".format(attr.replace(\"_\",\" \")))    \n",
    "    # --------------------------------------------\n",
    "    # Legend\n",
    "    legend_elements = [patches.Patch(facecolor=value, edgecolor=value,) for value in mug_dict[attr][\"hex_color\"].values()]\n",
    "    legend_labels = list(mug_dict[attr][\"hex_color\"].keys())\n",
    "    ax1.legend(legend_elements, \n",
    "              legend_labels,\n",
    "              bbox_to_anchor=(1.0, 1.0), \n",
    "              loc='upper left',\n",
    "              frameon=False,\n",
    "             )\n",
    "    # --------------------------------------------\n",
    "    # Export\n",
    "    out_path = os.path.join(outdir, SCRIPT_NAME + \"_tree_{}.{}\".format(attr.lower(), FMT))  \n",
    "    plt.savefig(out_path, dpi=dpi, bbox_inches = \"tight\")\n",
    "    \n",
    "    # Save Tree!\n",
    "    #out_path_dill_tree = os.path.join(outdir,  SCRIPT_NAME + \"_{}.phylo.obj\".format(attr.lower()))\n",
    "    #with open(out_path_dill_tree,\"wb\") as outfile:\n",
    "    #    dill.dump(tree_div, outfile)\n",
    "    \n",
    "file_colors.close()\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# 5. Export"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dataframe\n",
    "out_path_df = os.path.join(outdir, SCRIPT_NAME + \".tsv\" )\n",
    "out_path_pickle_df = os.path.join(outdir,  SCRIPT_NAME + \".df.obj\" )\n",
    "\n",
    "tree_df.to_csv(out_path_df, sep=\"\\t\")\n",
    "with open(out_path_pickle_df,\"wb\") as outfile:\n",
    "    dill.dump(tree_df, outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Timetrees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tt_copy = copy.deepcopy(tt)\n",
    "\n",
    "# Nexus\n",
    "out_path_nexus = os.path.join(outdir, SCRIPT_NAME + \"_timetree.nexus\" )\n",
    "Phylo.write(tt_copy.tree, out_path_nexus, 'nexus', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))\n",
    "\n",
    "# Dill object\n",
    "out_path_dill_tree = os.path.join(outdir,  SCRIPT_NAME + \"_timetree.treetime.obj\" )\n",
    "with open(out_path_dill_tree,\"wb\") as outfile:\n",
    "    dill.dump(tt_copy, outfile)\n",
    "    \n",
    "# Newick (remove comments)\n",
    "for c in tt_copy.tree.find_clades(): c.comment = None\n",
    "out_path_nwk = os.path.join(outdir, SCRIPT_NAME + \"_timetree.nwk\" )\n",
    "Phylo.write(tt_copy.tree, out_path_nwk, 'newick', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Divtrees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tt_copy = copy.deepcopy(tt)\n",
    "# Convert to divtree\n",
    "for n in tt_copy.tree.find_clades():\n",
    "    n.branch_length=n.mutation_length\n",
    "\n",
    "# Nexus\n",
    "out_path_nexus = os.path.join(outdir, SCRIPT_NAME + \"_divtree.nexus\" )\n",
    "Phylo.write(tt_copy.tree, out_path_nexus, 'nexus', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))\n",
    "\n",
    "# Dill object\n",
    "out_path_dill_tree = os.path.join(outdir,  SCRIPT_NAME + \"_divtree.treetime.obj\" )\n",
    "with open(out_path_dill_tree,\"wb\") as outfile:\n",
    "    dill.dump(tt_copy, outfile)\n",
    "    \n",
    "# Newick (remove comments)\n",
    "for c in tt_copy.tree.find_clades(): c.comment = None\n",
    "out_path_nwk = os.path.join(outdir, SCRIPT_NAME + \"_divtree.nwk\" )\n",
    "Phylo.write(tt_copy.tree, out_path_nwk, 'newick', format_branch_length='%1.{}f'.format(BRANCH_LEN_SIG_DIG))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Augur JSON\n",
    "\n",
    "  - alignment (empty)\n",
    "  - input_tree (tree_path)\n",
    "  - nodes (node_dict)\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ufboot': 100.0, 'scf': 'NA', 'branch_length': 0.0, 'strain': 'NA', 'date': 'NA', 'datebp': 'NA', 'country': 'Russia', 'province': 'Krasnoyarsk Krai', 'biovar': 'NA', 'branch_major': '0.PRE', 'branch_minor': '0.PRE2', 'biosample': 'NA', 'biosamplecomment': 'NA', 'branch_number': '0', 'rate': 9.056079828112778e-09, 'rate_fold_change': 1.0, 'mutation_length': 0.001, 'raw_date': '-4001-04-08', 'num_date': -4000.7338973552987, 'num_date_confidence': [-4028.4379593546546, -3773.0741326132847], 'num_date_bar': '6049.437959354655:5794.074132613285', 'clock_length': 0.1, 'branch_number_confidence': {'0': 0.9999999959154534}, 'branch_number_entropy': 8.856520735862206e-08, 'branch_major_confidence': {'0.PRE': 0.9987094648969286}, 'branch_major_entropy': 0.009917649653245002, 'branch_minor_confidence': {'0.PRE2': 0.3924074548210794}, 'branch_minor_entropy': 1.329759845199042, 'country_confidence': {'Russia': 0.9931643504765991}, 'country_entropy': 0.04426619316105468, 'province_confidence': {'Krasnoyarsk Krai': 0.3825949094902657}, 'province_entropy': 1.643089389423311}\n"
     ]
    }
   ],
   "source": [
    "augur_dict = augur_export(\n",
    "    tree_path=out_path_nwk, \n",
    "    aln_path=aln_path, \n",
    "    tree=tt.tree,     \n",
    "    tree_df=tree_df, \n",
    "    color_keyword_exclude=[\"color\", \"coord\", \"reg\", \"lat\", \"lon\"],\n",
    "    type_convert = {\n",
    "        \"Branch_Number\" : (lambda x : str(x))\n",
    "    },\n",
    ")\n",
    "\n",
    "\n",
    "print(augur_dict[\"nodes\"][\"NODE0\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add Mugration Model to JSON"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "mug_models_dict = {}\n",
    "\n",
    "for attr in ATTRIBUTE_LIST:\n",
    "    mug_models_dict[attr] = {}\n",
    "    \n",
    "    # Mugration GTR model\n",
    "    gtr = mug_dict[attr][\"tree_mug\"].gtr\n",
    "    \n",
    "    # Mugration Rate\n",
    "    mug_models_dict[attr][\"rate\"] = gtr.mu\n",
    "    \n",
    "    # Mugration states\n",
    "    alphabet = mug_dict[attr][\"letter_to_state\"]\n",
    "    mug_models_dict[attr][\"alphabet\"] = [alphabet[k] for k in sorted(alphabet.keys())]\n",
    "    \n",
    "    # Mugration probabilities\n",
    "    prob =  list(mug_dict[attr][\"tree_mug\"].gtr.Pi)\n",
    "    mug_models_dict[attr][\"equilibrium_probabilities\"] = list(gtr.Pi)\n",
    "    \n",
    "    # Mugration transition matrix\n",
    "    mat = [list(x) for x in gtr.W]\n",
    "    mug_models_dict[attr][\"equilibrium_probabilities\"] = mat\n",
    "    \n",
    "augur_dict[\"models\"] = mug_models_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ufboot': 'NA', 'scf': 'NA', 'branch_length': 1.8758e-06, 'strain': 'NA', 'date': 1992.0, 'datebp': -29.0, 'country': 'United States of America', 'province': 'Colorado', 'biovar': 'Orientalis', 'branch_major': '1.ORI', 'branch_minor': '1.ORI1', 'biosample': 'NA', 'biosamplecomment': 'KEEP: Assembly Modern Reference', 'branch_number': '1', 'rate': 4.137969101655365e-08, 'rate_fold_change': 4.569271892690116, 'mutation_length': 1.72000214266788e-06, 'raw_date': '1991-12-26', 'num_date': 1991.9846805945695, 'num_date_confidence': [1992.0, 1992.0], 'num_date_bar': '29.0:29.0', 'clock_length': 26.18326154138572, 'branch_number_confidence': {'1': 0.9999999999999999}, 'branch_number_entropy': -9.99866855976916e-13, 'branch_major_confidence': {'1.ORI': 1.0}, 'branch_major_entropy': -1.000088900581841e-12, 'branch_minor_confidence': {'1.ORI1': 1.0}, 'branch_minor_entropy': -1.000088900581841e-12, 'country_confidence': {'United States of America': 0.9999999999999999}, 'country_entropy': -9.99866855976916e-13, 'province_confidence': {'Colorado': 1.0}, 'province_entropy': -1.000088900581841e-12}\n"
     ]
    }
   ],
   "source": [
    "out_path_augur_json = os.path.join(outdir, SCRIPT_NAME + \"_augur.json\" )\n",
    "utils.write_json(data=augur_dict, file_name=out_path_augur_json, indent=JSON_INDENT)\n",
    "\n",
    "print(augur_dict[\"nodes\"][\"Reference\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Auspice JSON\n",
    "\n",
    "Manual edits of https://github.com/nextstrain/augur/blob/master/augur/export_v2.py\n",
    "\n",
    "This can then be used for auspice via:\n",
    "\n",
    "```\n",
    "HOST=\"localhost\" auspice view --datasetDir .\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'dict'>\n",
      "Validating schema of '/2/scratch/keaton/plague-phylogeography/results/config/auspice_config.json'...\n",
      "Validation success.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: You asked for a color-by for trait 'chrom_cov', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'pcd1_cov', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'pmt1_cov', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'ppcp1_cov', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'pla_cov', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'chrom_dep', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'pcd1_dep', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'pmt1_dep', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'ppcp1_dep', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'pla_dep', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: You asked for a color-by for trait 'ppcp1_pla_dep_ratio', but it has no values on the tree. It has been ignored.\n",
      "\n",
      "WARNING: [config file] Trait 'num_date_bar' is missing type information. We've guessed 'categorical'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating produced JSON\n",
      "Validating schema of '/2/scratch/keaton/plague-phylogeography/results/mugration/all/chromosome_filter5/mugration_model_auspice.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/2/scratch/keaton/plague-phylogeography/results/mugration/all/chromosome_filter5/mugration_model_auspice.json' succeeded.\n",
      "Validation successful for local JSON.\n",
      "Validating produced JSON\n",
      "Validating schema of '/2/scratch/keaton/plague-phylogeography/auspice/plague-phylogeography_mugration-model.json'...\n",
      "Validating that the JSON is internally consistent...\n",
      "Validation of '/2/scratch/keaton/plague-phylogeography/auspice/plague-phylogeography_mugration-model.json' succeeded.\n",
      "Validation successful for remote JSON.\n"
     ]
    }
   ],
   "source": [
    "auspice_dict = auspice_export(\n",
    "    tree=tree_div, \n",
    "    augur_json_paths=[out_path_augur_json], \n",
    "    auspice_config_path=auspice_config_path, \n",
    "    auspice_colors_path=auspice_colors_path,\n",
    "    auspice_latlons_path=auspice_latlon_path, \n",
    "    )\n",
    "\n",
    "# Write outputs - For Local Rendering\n",
    "out_path_auspice_local_json = os.path.join(outdir, SCRIPT_NAME + \"_auspice.json\" )\n",
    "utils.write_json(data=auspice_dict, file_name=out_path_auspice_local_json, indent=JSON_INDENT, include_version=False)\n",
    "export_v2.validate_data_json(out_path_auspice_local_json)\n",
    "print(\"Validation successful for local JSON.\")\n",
    "\n",
    "# Write outputs - For Remote Rendering\n",
    "out_path_auspice_remote_json = os.path.join(auspice_remote_dir_path, AUSPICE_PREFIX + SCRIPT_NAME.replace(\"_\",\"-\") + \".json\" )\n",
    "utils.write_json(data=auspice_dict, file_name=out_path_auspice_remote_json, indent=JSON_INDENT, include_version=False)\n",
    "export_v2.validate_data_json(out_path_auspice_remote_json)\n",
    "print(\"Validation successful for remote JSON.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
